MSC4140: Cancellable delayed events (#4140)
matrix-org/matrix-doc: proposals/4140-delayed-events-futures.md

Delayed events are returned in chronological order of their intended send time, which is `running_since` + `delay`.

To filter results to include only delayed events with a scheduled send time within a specific time range,
the endpoint supports query parameters of `from_ts` and `to_ts`, which are Unix timestamps that specify this time range.
If either parameter is unspecified, its respective end of the time range is unbounded.

To return results in reverse chronological order, the endpoint supports a query parameter of `dir=b`.
Providing a query parameter of `dir=f` uses the default of increasing chronological order.

The endpoint accepts a query parameter `from` which is a token that can be used to paginate the list of delayed events
as per the [pagination convention](https://spec.matrix.org/v1.18/appendices/#pagination).
The homeserver can choose a suitable page size.

To filter results on delayed events with certain properties, the endpoint accepts any of the following query parameters:
- `room_id` - Return only delayed events that were scheduled to be sent into the room with this ID.
- `type` - Return only delayed events of the specified event type.

> - `status`: `"scheduled"|"send"|"cancel"|"error"` - Return only delayed events that are still scheduled to be sent, or

Reviewer: Cosmetic: One of the actions is a past participle, two are imperatives and the last is a noun. Would there be any point in harmonizing this? For instance, we could make all of them participles:

```suggestion
- `status`: `"scheduled"|"sent"|"cancelled"|"failed"` - Return only delayed events that are still scheduled to be sent, or
```

Reviewer 2: The grammatical inconsistency was a means to reuse existing terms from elsewhere in the proposal, i.e. `"send"` & `"cancel"` matches `/send` and `/cancel`; and `"error"` matches the `"error"` field in the `GET` payloads.

But given your feedback, and at least one positive reaction from someone else, it looks like this wasn't worth the cost :sweat_smile:  I have no problems with going for the cosmetic route.

  timeout reset it to `{}`. If we have two separate request sending `A` could work
  but the event with content `{}` could fail. The state would not automatically
  reset to `{}`.

For this usecase an optional `m.send_now` field can be added to the body.

## Usecase specific considerations

### MatrixRTC

We want can use the actions and the timeout for matrix rtc for the following situations

- If the client takes care of its membership, we use a short timeout value (around 5-20 seconds)
  The client will have to ping the refresh endpoint approx every 2-19 seconds.
- When the SFU is capable of taking care of managing our connection state and we trust the SFU to
  not disconnect a really long value can be chosen (approx. 2-10hours). The SFU will then only send
  an action once the user disconnects or looses connection (it could even be a different action for both cases
  handling them differently on the client)
  This significantly reduces the amount of calls for the `/future` endpoint since the sfu only needs to ping
  once per session (per user) and every 2-5hours (instead of every `X` seconds.)

### Self destructing messages

> This MSC also allows to implement self destructing messages:

Reviewer 3: It seems possible to use that also to schedule messages for later I believe? I am unsure if a timeout is enough or if a timestamp would be useful. I think timeout seems enough since the computation would be simple enough to do client side from a timestamp to a timeout.

A way to retrieve futures would be useful however, so we can list them in a client if needed.

Reviewer 3: Regarding timeout vs timestamp it would probably incur some small delay however, because between the computation client side and the request being registered server side, it may take some seconds. Not sure if it is really important however.
Replacing timeout by timestamp (instead of allowing both) may be the way to go for simplicity and accuracy.

Author: Per room? so you want others room members to see the messages that are going to be sent?
I was thinking more about a per user api where I can see a list of events i have scheduled.

Reviewer 3: I was thinking about per room for the current user. But then it's tedious to get it for all rooms if needed so a global one is probably better, and the client can filter. Perhaps we want to be able to specify a message `type`.

This leads me to think... Do we want this in sync in the end ? We could have a `futures` section there.

Author: That makes sense. I really dont like the idea of sharing this information with other users however. I really think the concept of: "Everything that is shown/shared by other room members should be in the DAG" is super important. To continue with this model a future that is not yet sent into the room should also only be known  by the sender. Otherwise it would make much more sense to just put the data into the room state. If that is desired no-one is stopping the sender to also mention in the room state what futures they have scheduled.

Author: > It seems possible to use that also to schedule messages for later I believe

I noticed I have not yet commented on this. But yes this is totally possible and could be another section in the use cases!

Author: Since there now is a proposed enpoint for the list of futures I mark this as resolved.

      - new event -> timeout: the new event will invalidate the future. No

- After the homeservers sends a timeout or action future event, the associated
  timer and `future_token` is canceled/invalidated.

So for each Future the client sends, the homeserver will send one event
conditionally at an unknown time that can trigger logic on the client.
This allows for any generic timeout logic.

Timed messages/reminders or ephemeral events could be implemented using this where
clients send a redact as a future or a room event with intentional mentions.

In some scenarios it is important to allow to send an event with an associated
future at the same time.

- One example would be redacting an event. It only makes sense to redact the event
  if it exists.
  It might be important to have the guarantee, that the redact is received
  by the server at the time where the original message is sent.
- In the case of a state event we might want to set the state to `A` and after a
  timeout reset it to `{}`. If we have two separate request sending `A` could work
  but the event with content `{}` could fail. The state would not automatically
  reset to `{}`.

> For this usecase an optional `m.send_now` field can be added to the body.

Author: If we have a generlized way to batch sent matrix events this could be leverged on this and a future itself is JUST the future. And the batch send event would define the semantics for sending the future and guaranteeing that the send_now event is also sent

Author: [MSC2716](https://github.com/matrix-org/matrix-spec-proposals/pull/2716) allows bulk sending events.

Reviewer 4: Marking as resolved because batch sending is outside of the scope of the proposal and is discussed in the alternatives section.

Reviewer 4: It is explicitly mentioned in [`84b8dc0`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/84b8dc05ca89f0fd52c58afade2071caca068c46)

# MSC4140: Delayed events (Futures)

Allowing to schdule/delay events would solve numerous issues in
matrix.

- Updating call member events after the user disconnected.
- Sending scheduled messages (send at a specific time)
- Creating self destructing events (By sending a delayed redact)

Currently there is no mechanism for a client to reliably that an event is still valid.
The only way to update an event is to post a new one.

> In some situations the client just looses connection or fails to sent the expired

Reviewer 2: ```suggestion
In some situations the client just loses connection or fails to sent the expired
```

version of the event. This proposal also includes a expiration/timeout
system so that those scenarios are also covered.

We want to send an event in advance
to the homeserver but let the homeserver decide the time when its actually added to the
DAG.
The condition for actually sending the delayed event would could be a timeout or a external trigger via a synapse endpoint.

## Proposal

To make this as generic as possible, the proposed solution is to allow sending
multiple presigned events and delegate the control of when to actually send these
events to an external services. This allows to a very flexible way to mark events as expired,

> since the sender can choose what event will be sent once expired.

Reviewer 2: Is it entire events that need to be signed, or just their content?  If it's the former, then `/send/future` should behave more like #4080's [`/send_pdus`](https://github.com/devonh/matrix-spec-proposals/blob/cryptoIDs/proposals/4080-cryptographic-identities.md#post-_matrixclientv1send_pdustxnid).  This could be done by having `/send/future`'s `send_*` fields accept fully-signed events instead of signed content + other fields.

For this to work, there'd also need to be the [modified PUT `/send` & PUT `/state` endpoints](https://github.com/devonh/matrix-spec-proposals/blob/cryptoIDs/proposals/4080-cryptographic-identities.md#put-_matrixclientv4roomsroomidsendeventtypetxnid--put-_matrixclientv4roomsroomidstateeventtypestatekey) for retrieving the PDUs that need signing.

That would make the client flow of sending a Future as follows:
- call PUT `/send` / PUT `/state` for each event that's to be sent in a Future (ideally, this could be batched)
- sign each retrieved PDU
- put each signed PDU in a request to `/send/future`, placing them in `"send_*"` fields as desired

Author: > call PUT /send / PUT /state for each event that's to be sent in a Future (ideally, this could be batched)
> sign each retrieved PDU
> put each signed PDU in a request to /send/future, placing them in "send_*" fields as desired

What should be batched is mostly the put PDU's right. Since creating the signed PDU's is okay to not be batched. The client then just needs to be sure that they have created all the events before sending the signed PDU's. if creating the PDU's fails the client just retires until it has the full list of events that need to sent (because they rely on each other)

So what really would need to be batched is the last step. Sending the signed PDU's.

Author: If there are multiple futures in one send_pdus call multiple future tokens need to be issued. In the response a dictionary with future_id-> token would use the id to map all the tokens.

Reviewer 4: I've added an explicit section on https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/expiring-events-keep-alive/proposals/4140-delayed-events-futures.md#compatibility-with-cryptographic-identities

Furthermore, the proposal has changed a lot since these comments.

If there are any clarifications needed in relation to cryptographic identities I suggest adding them to https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/expiring-events-keep-alive/proposals/4140-delayed-events-futures.md#compatibility-with-cryptographic-identities directly.

We call those delayed events `Futures`.

A new endpoint is introduced:
`PUT /_matrix/client/v3/rooms/{roomId}/send/future/{txnId}`
It behaves exactly like the normal send endpoint except that that it allows

> to send a list of event contents. The body looks as following:

Reviewer 2: How do power levels interact with Futures?  Should PLs be enforced on the events sent in `/send/future`, or on the event that the homeserver adds to the DAG on timeout / action time?

At least the `send_now` event should be subject to PLs.

Author: (As discussed)
Power levels should always be applied right before the event is distributed (added into the DAG)
The homeserver treats the event as if it was sent at the point in time where it is resolved (either by the timeout or an action)
The homeserver runs all the normal checks (power level, room memberhsip...) at the time of resolving the event. It is possible for a future to be rejected on timeout and not be send into the room in case the permissions have changed.

```json
{
  "m.timeout": 10,
  "m.send_on_timeout": {
    "content": sendEventBody0,
    "type": "m.room.message",
  },

  "m.send_on_action:${actionName}": {
    "content": sendEventBody1,
    "type": "m.room.message"
  },

  // optional
  "m.send_now": {
    "content": sendEventBody2,
    "type": "m.room.message"
    },
}

> ```

Author: I removed them.

Each of the `sendEventBody` objects are exactly the same as sending a normal
event.

> There can be an arbitrary amount of `actionName`s.

Author: I like this.

> There can be an arbitrary amount of `actionName`s.

Author: and added it now

All of the fields are optional except the `timeout` and the `send_on_timeout`.
This guarantees that all tokens will expire eventually.

The homeserver can set a limit to the timeout and return an error if the limit
is exceeded.

### Response

The response will mimic the request:

```json
{
  "m.send_on_timeout": { "eventId": "id_hash" },
  "m.send_on_action:${actionName}": { "eventId": "id_hash" },

  "future_token": "token",

  // optional
  "m.send_now": { "eventId": "id_hash" }
}
```

### Delegating futures

The `token` can be used to call another future related endpoint:

> `PUT /_matrix/client/v3/futures/refresh` and `PUT /_matrix/client/v3/futures/action/${actionName}`.

Reviewer 2: Should there be a `/futures/cancel` endpoint as well?  Without it, the only way to cancel a Future is to send a state event with the same state key & content as the Future's `send_on_timeout` event (which, of course, is applicable only when the `send_on_timeout` event is a state event).  That also prevents external services from being able to cancel, as they can only hit the `future_token`-accepting endpoints & can't send events.

> `PUT /_matrix/client/v3/futures/refresh` and `PUT /_matrix/client/v3/futures/action/${actionName}`.

Reviewer 2: ```suggestion
`POST /_matrix/client/v3/futures/refresh` and `POST /_matrix/client/v3/futures/action/${actionName}`.
```
because these don't accept transaction IDs and aren't idempotent.  For example, multiple `POST /refresh`s for the same token should repeatedly refresh the timeout, and any `POST /action/${actionName}` after the first should fail (as the action event would have been posted already).

{
  "future_token": "token"
}
```

The information required to call this endpoint is very limited so that almost
no metadata is leaked. This allows to share a refresh link to a different
service. This allows to delegate the send time. An SFU for instance, that tracks the current client connection state,
and pings the HS to refresh and call a dedicated action to communicate
that the user has intentionally left the conference.

The homeserver does the following when receiving a Future.

- It **sends** the optional `m.send_now` event.
- It **generates** a `future_token` and stores it alongside with the time
  of retrieval, the event list and the timeout duration.
- **Starts a timer** for the stored `future_token`.

  - If a `PUT /_matrix/client/v3/futures/refresh` is received, it
    **restarts the timer** with the stored timeout duration.
  - If a `PUT /_matrix/client/v3/futures/action/${actionName}` is received, it **sends the associated action event**
    `m.action:${actionName}`.
  - If the timer times out, **it sends the timeout event** `m.send_timeout`.
  - If the future is a state event and includes a `m.send_now` event
    the future is only valid while the `m.send_now`

>     is still the current state:

Reviewer 2: `send_now` is out of the main proposal, so resolving this.

This allows for any generic timeout logic.

Timed messages/reminders or ephemeral events could be implemented using this where
clients send a redact as a future or a room event with intentional mentions.

In some scenarios it is important to allow to send an event with an associated
future at the same time.

- One example would be redacting an event. It only makes sense to redact the event
  if it exists.
  It might be important to have the guarantee, that the redact is received
  by the server at the time where the original message is sent.
- In the case of a state event we might want to set the state to `A` and after a
  timeout reset it to `{}`. If we have two separate request sending `A` could work
  but the event with content `{}` could fail. The state would not automatically
  reset to `{}`.

For this usecase an optional `m.send_now` field can be added to the body.

### Getting running futures

Using `GET /_matrix/client/v3/futures` it is possible to get the list of all running futures.
This is an authenticated endpoint. It sends back the json
of the final events how they will end up in the DAG with the associated `future_token`.

```json
[
  {

>     "m.send_now": finalEvent_0,

Reviewer 2: Is there a point in returning the `send_now` event if it would have been sent already?

- If the client takes care of its membership, we use a short timeout value (around 5-20 seconds)
  The client will have to ping the refresh endpoint approx every 2-19 seconds.
- When the SFU is capable of taking care of managing our connection state and we trust the SFU to
  not disconnect a really long value can be chosen (approx. 2-10hours). The SFU will then only send
  an action once the user disconnects or looses connection (it could even be a different action for both cases
  handling them differently on the client)
  This significantly reduces the amount of calls for the `/future` endpoint since the sfu only needs to ping
  once per session (per user) and every 2-5hours (instead of every `X` seconds.)

### Self destructing messages

This MSC also allows to implement self destructing messages:

`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`

```json
{
  "m.text": "my msg"
}
```

`PUT /_matrix/client/v3/rooms/{roomId}/send/future/{txnId}`

```json
{
  "m.timeout": 10*60,
  "m.send_on_timeout": {
    "type":"m.room.readact",
    "content":{
      "redacts": "EvId"
    }
  }
}
```

> ## EventId template variable

Reviewer 2: If pre-signing events requires them to have event IDs to be included in the signature, then all event IDs will be known by the time `/send/future` is called & it won't be necessary to have this template variable.

I asked about signing with event IDs here: https://github.com/matrix-org/matrix-spec-proposals/pull/4080/files#r1610545596

Author: Oh this is really great actually.
It will force us to have multiple request but what is really good is, that it is really easy to send the current state and the timout state at the same time!

Author: Added a note so that it is clear that this is only a stop gap and that there are more elegant solutions.
I still would like to keep that section since its part of the explantion why we would like to send events at the same time and
why its better to solve it in a different way.

# MSC4140: Delayed events (Futures)

> Allowing to schdule/delay events would solve numerous issues in

Reviewer 2: ```suggestion
Allowing to schedule/delay events would solve numerous issues in
```

matrix.

- Updating call member events after the user disconnected.
- Sending scheduled messages (send at a specific time)
- Creating self destructing events (By sending a delayed redact)

Currently there is no mechanism for a client to reliably that an event is still valid.
The only way to update an event is to post a new one.
In some situations the client just loses connection or fails to sent the expired
version of the event. This proposal also includes a expiration/timeout
system so that those scenarios are also covered.

We want to send an event in advance
to the homeserver but let the homeserver decide the time when its actually added to the
DAG.
The condition for actually sending the delayed event would could be a timeout or a external trigger via a synapse endpoint.

## Proposal

To make this as generic as possible, the proposed solution is to allow sending
multiple presigned events and delegate the control of when to actually send these
events to an external services. This allows to a very flexible way to mark events as expired,
since the sender can choose what event will be sent once expired.

We call those delayed events `Futures`.

A new endpoint is introduced:

> `PUT /_matrix/client/v3/rooms/{roomId}/send/future/{txnId}`

Reviewer 2: ```suggestion
`PUT /_matrix/client/v3/futures/send/{txnId}`
```
because:
1) `/send/future` has the same path prefix as the regular `/send/{eventType}`, and would conflict with it unless servers treat `future` as a special-case value that can never be an event type
1) The other future endpoints all start with `/futures`, and it would be consistent if this one did too
1) Removing `roomId` from the path opens up the possibility for each event in the future to refer to a different room

Author: changed it to `/_matrix/client/v3/rooms/{roomId}/send_future/{txnId}` now since this mimics the current `send` endpoint. In the alternatives there is a section to consider using the `send` endpoint directly as discussed.

It behaves exactly like the normal send endpoint except that that it allows
to send a list of event contents. The body looks as following:

```json
{
  "timeout": 10,
  "send_on_timeout": {...fullSignedEvent},

  "send_on_action":{
    "${action1}": {...fullSignedEvent},
    "${action2}": {...fullSignedEvent},
    ...
  }

  // optional
  "send_now": {...fullSignedEvent},
}
```

Each of the `sendEventBody` objects are exactly the same as sending a normal
event.

Power levels are evaluated once one of the events will be distributed/inserted into the DAG.
This implies a future can fail if it violates power levels at the time it resolves.
(Its also possible to successfylly send a future the user has no permission to at the time of sending
if the power level situation has changed at the time the future resolves.)

There can be an arbitrary amount of `actionName`s.

All of the fields are optional except the `timeout` and the `send_on_timeout`.

> This guarantees that all tokens will expire eventually.

Reviewer 4: ```suggestion
This guarantees that all tokens will eventually expire.
```

In some scenarios it is important to allow to send an event with an associated
future at the same time.

- One example would be redacting an event. It only makes sense to redact the event
  if it exists.
  It might be important to have the guarantee, that the redact is received
  by the server at the time where the original message is sent.
- In the case of a state event we might want to set the state to `A` and after a
  timeout reset it to `{}`. If we have two separate request sending `A` could work
  but the event with content `{}` could fail. The state would not automatically
  reset to `{}`.

For this usecase an optional `m.send_now` field can be added to the body.

### Getting running futures

Using `GET /_matrix/client/v3/futures` it is possible to get the list of all running futures.
This is an authenticated endpoint. It sends back the json
of the final events how they will end up in the DAG with the associated `future_token`.

```json
[
  {
    "m.send_now": finalEvent_0,
    "m.send_on_timeout": finalEvent_1,
    ...,

    "future_token":"token"
  },
]
```

This can be used so clients can optionally display events
that will be send in the future.

> For self destructing messages it is recommanded to include

Reviewer 4: ```suggestion
For self-destructing messages it is recommended to include
```

this information in the event itself so that the usage of
this endpoint can be minimized.

## Usecase specific considerations

### MatrixRTC

We want can use the actions and the timeout for matrix rtc for the following situations

- If the client takes care of its membership, we use a short timeout value (around 5-20 seconds)
  The client will have to ping the refresh endpoint approx every 2-19 seconds.
- When the SFU is capable of taking care of managing our connection state and we trust the SFU to
  not disconnect a really long value can be chosen (approx. 2-10hours). The SFU will then only send
  an action once the user disconnects or looses connection (it could even be a different action for both cases
  handling them differently on the client)
  This significantly reduces the amount of calls for the `/future` endpoint since the sfu only needs to ping
  once per session (per user) and every 2-5hours (instead of every `X` seconds.)

> ### Self destructing messages

Reviewer 4: ```suggestion
### Self-destructing messages
```

This MSC also allows to implement self destructing messages:

`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`

```json
{
  "m.text": "my msg"
}
```

`PUT /_matrix/client/v3/rooms/{roomId}/send/future/{txnId}`

```json
{
  "m.timeout": 10*60,
  "m.send_on_timeout": {
    "type":"m.room.readact",
    "content":{
      "redacts": "EvId"
    }
  }
}
```

## EventId template variable

> [!IMPORTANT]  
> This proposes a stop gap solution. It is highly preferred to have a general batch sending solution.
> Also see the **Alternatives** section

It would be useful to be able to send redactions and edits as one http request.
This would make sure that the client cannot loose connection after sending the first event.
For instance sending a self destructing message without the redaction.

The optional proposal is to introduce template variables that are only valid in `Future` events.
`$m.send_now.event_id` in the content of one of the `m.send_on_action:${actionName}` and
`m.send_on_timeout` contents this template variable can be used.

> The **Self destructing messages** example would simplify to:

Reviewer 4: ```suggestion
The **Self-destructing messages** example would simplify to:
```

`PUT /_matrix/client/v3/rooms/{roomId}/send/future/{txnId}`

```json
{
  "m.send_now":{
    "type":"m.room.message",
    "content":{
      "m.text": "my msg"
    }
  },
  "m.timeout": 10*60,
  "m.send_on_timeout": {
    "type":"m.room.readact",
    "content":{
      "redacts": "$m.send_now.event_id"
    }
  }
}
```

With cryptographic identities events would be presigned.

> The server will first send the finilized event to the client.

Reviewer 2: ```suggestion
The server will first send the finalized event to the client.
```

At this point the client has the id but the event is not in the DAG.
So it would be trivial to sign both the event and the redaction/related event
and then send them.

## Potential issues

## Alternatives

[MSC4018](https://github.com/matrix-org/matrix-spec-proposals/pull/4018) also
proposes a way to make call memberships reliable. It uses the client sync loop as
an indicator to determine if the event is expired. Instead of letting the SFU
inform about the call termination or using the call app ping loop like we propose
here.

---

The following names for the endpoint are considered

- Future
- DelayedEvents

> - RetardedEvents

Reviewer 2: Let's not consider this one.

How about `PostponedEvents`?

New endpoints are introduced:

`PUT /_matrix/client/v3/rooms/{roomId}/send_future/{txnId}?timeout={timeout_duration}&future_group_id={group_id}`

`PUT /_matrix/client/v3/rooms/{roomId}/state_future/{eventType}/{stateKey}?timeout={timeout_duration}&future_group_id={group_id}`

Those behave like the normal `send`/`state` endpoints except that that they allow
to define a `timeout` and a `future_group_id` in their query parameters.

The **`future_group_id`** is an identifier defined by the client.
The purpose of this identifier is to group multiple futures in one mutually exclusive group.
Only one of the events in such a group can ever reach the DAG/will be distributed by the homeserver.
All other events will be discarded.
One group can only contain one event with a `timeout` (timeout futures). The other events do not have a timeout (action futures) and are send
as an mutually exclusive alternative to the event send with `timeout`.
We call these timeout futures and action futures.
The server will respond with a [`409`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409)
(`Conflict` This response is sent when a request conflicts with the current state of the server.) if

> the client tries to send an action future without there being a timeout future with the same `group_id`

Reviewer 5: ```suggestion
the client tries to send an action future without there being a timeout future with the same `future_group_id`
```

Author: I called the value `{group_id}` but here it makes more sense to refer to the parameter name.

The body is the same as sending a normal event.

Power levels are evaluated for each event only once the trigger has occurred and it will be distributed/inserted into the DAG.
This implies a future can fail if it violates power levels at the time it resolves.
(It's also possible to successfully send a future the user has no permission to at the time of sending
if the power level situation has changed at the time the future resolves.)

### Response

The response will include a `send_token` and a optional `refresh_token` but no `event_id` since the `event_id` depends on the `origin_server_ts` which is not yet determined. A timeout future will contain both, `send_token` and `refresh_token` but an action future will only have a `action_token` in its body.

```json
{
  // always present

>   "send_token": "token",

Reviewer 5: What about adding the `future_group_id` as well?

### Getting running futures

Using `GET /_matrix/client/v3/futures` it is possible to get the list of all running futures issues by the authenticated user.
This is an authenticated endpoint. It sends back the json
of the final event content with the associated tokens.

```json
[
  {
    "url":"https://domain/_matrix/client/v3/futures/{refresh_token}",
    "body":{
      ...event_body
    },
    "response":{
      // always present
      "send_token": "token",
      // optional if there is a timeout
      "refresh_token": "token",
      "cancel_token": "token"
    }
  },
]
```

This can be used so clients can optionally display events
that will be send in the future.
And to optionally cancel tokens for them.

For all usecases where the existence of a running future is also of interest for other room members,
(like self-destructing messages) it is recommended to include
this information in the effected event itself.

## Usecase specific considerations

### MatrixRTC

In this section an overview is given how this MSC is used in [MSC4143: MatrixRTC](https://github.com/matrix-org/matrix-spec-proposals/pull/4143)
and alternative expiration systems are evaluated.

#### Background

> MatrixRTC makes it necessary to have real time information about the current matrixRTC session.

Reviewer 5: ```suggestion
MatrixRTC makes it necessary to have real time information about the current MatrixRTC session.
```

There are numerous approaches to solve such a situation. They split into two categories:

- Polling based
  - Ask the users if they are still connected.
  - Ask an RTC backend (SFU) who is connected.
- Timeout based

  - Update the room state every x seconds. This allows clients to check how long an event has not been updated and ignore it if its expired.
  - Use Future events with a 10s timeout to send the disconnected from call in less then 10s after the user is not anymore pingin the `/refresh` endpoint. (or delegate the disconnect action to a service attached to the SFU)
  - Use the client sync loop as a special case timeout for call member events. (See [Alternatives/MSC4018 (use client sync loop))](#msc4018-use-client-sync-loop))

Polling based solution have a big overhead in complexity and network requests on the clients.
Example:

> A room list with 100 rooms where there has been a call before in every room (or there is an ongoing call) would require the client to send a to-device message (or a request to the SFU) to every user that has an active state event to check if they are still online. Just to display the room tile properly.

> For displaying the room list timeout based approaches are much more reasonable because this allows computing matrixRTC metadata for a room to be synchronous.

Reviewer 5: ```suggestion
For displaying the room list timeout based approaches are much more reasonable because this allows computing MatrixRTC metadata for a room to be synchronous.
```

the users has left the call.

This proposal also includes a way to refresh the timeout. Allowing to delay the event multiple times.
A periodic ping of the refreshing can be used as a heardbeat mechanism. Once the refresh ping is not send
anymore the timeout condition is met and the homerver sends the event with the expired content information.

This translate to: _"only send the event when the client is not running the its program anymore (not sending the heartbeat anymore)"_
We call those delayed events `Futures`.

New endpoints are introduced:

`PUT /_matrix/client/v3/rooms/{roomId}/send_future/{eventType}/{txnId}?future_timeout={timeout_duration}&future_group_id={group_id}`

`PUT /_matrix/client/v3/rooms/{roomId}/state_future/{eventType}/{stateKey}?future_timeout={timeout_duration}&future_group_id={group_id}`

Those behave like the normal `send`/`state` endpoints except that that they allow
to define `future_timeout` and `future_group_id` in their query parameters.

- `future_timeout: number | "none"` is a required parameter that defines how long the homeserver will wait before sending

>   the event into the room. Since the timeout can be refreshed and sending the future can be triggered via an endpoint (see: [Proposal/Delegating futures](#delegating-futures)) this value is not enough to predict the time this event will arrive in the room.

Reviewer 5: ```suggestion
  the event into the room. **Note** Since the timeout can be refreshed and sending the future can be triggered via an endpoint (see: [Proposal/Delegating futures](#delegating-futures)) this value is not enough to predict the time this event will arrive in the room.
```

- The server can optionally configure a maximum `timeout_duration`
  (In the order of one week dependent on how long they want to track futures)
  The server will respond with a [`400`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) (`Bad Request`, with a message
  containing the maximum allowed `timeout_duration`) if the
  client tries to send a timeout future with a larger `timeout_duration`.

The body is the same as sending a normal event.

Power levels are evaluated for each event only once the trigger has occurred and it will be distributed/inserted into the DAG.
This implies a future can fail if it violates power levels at the time it resolves.
(It's also possible to successfully send a future the user has no permission to at the time of sending
if the power level situation has changed at the time the future resolves.)

### Response

The response will include a `send_token` and a optional `refresh_token` but no `event_id` since the `event_id` depends on the `origin_server_ts` which is not yet determined. A timeout future will contain both, `send_token` and `refresh_token` but an action future will only have a `action_token` in its body.

> ```json

Reviewer 2: Maybe we can specify different response types for timeout & action futures, to make it easier to describe/refer to their exclusive fields.

On that note, do only timeout futures need a `cancel_token`?  And could action futures use a `send_token` instead of a specially-named `action_token`?  (The example JSON below suggests that is the case)

Since the refresh and send tokens are of the same format it is not even possible to evaluate
what that token is for when reading the https request log.
This unauthenticated endpoint allows to delegate resolving the future.
An SFU for instance, that tracks the current client connection state, could get a url that it
needs to call every X hours while a user is connected and a url it has to call once the user disconnects.
This way the SFU can be used as the source of truth for the call member room state even if the client
gets closed or looses connection and without knowing anything about the Matrix call.

The homeserver does the following when receiving a Future:

- It checks for the validity of the request (based on the `future_timeout` and the `future_group_id` query parameters)
  and returns a `409` or `400` if necessary.
- It **generates** a `send_token` and optionally a `future_group_id`, a `refresh_token` and a `cancel_token` and stores them alongside the time
  of retrieval and the `timeout_duration`.
- If `future_timeout` was present, it **Starts a timer** for the `refresh_token`.

  - If a `PUT /_matrix/client/v3/futures/{refresh_token}` is received, it

>     **restarts the timer** with the stored `timeout_duration` for the associated timeout future.

Reviewer 2: Should this respond with a 4xx for action futures, or respond successfully to not leak whether the future is a timeout or action future?

Author: I think this would not happen because the refresh token is different to the send token. So a refresh token for a action future (that did not send a refresh token in its response) is equivalent to an arbitrary token unknown to the homeserver.
I think responding with a `410` (Gone) in case the homeserver does not know the token (or does not know it anymore since it got invalidated) would make sense to let the service know this token is invalidated.

  reset to `{}`.

For this usecase batch sending of multiple futures would be desired.

We do not include batch sending in the proposal of this MSC however since batch sending should
become a generic Matrix concept as proposed with `/send_pdus`. (see: [MSC4080: Cryptographic Identities](https://github.com/matrix-org/matrix-spec-proposals/pull/4080))

There is a [batch sending version](#batch-sending-futures-with-custom-endpoint) in the Alternatives section
that proposes a future specific group sending endpoint in case this is required sooner then its realistic to implement [MSC4080: Cryptographic Identities](https://github.com/matrix-org/matrix-spec-proposals/pull/4080).

### Getting running futures

Using `GET /_matrix/client/v3/futures` it is possible to get the list of all running futures issues by the authenticated user.
This is an authenticated endpoint. It sends back the json
of the final event content with the associated tokens.

```json
[
  {
    "url":"/_matrix/client/v3/rooms/{roomId}/send_future/{eventType}/{txnId}?timeout={timeout_duration}&future_group_id={group_id}",
    "body":{
      ...event_body

>     },

Reviewer 2: Could this instead return the path & query parameters (excluding `txnId`) as fields of the returned JSON items?

Also, since the payload of `send_future` and `state_future` is event _content_, we could return the `"content"` that was set in the original request instead of the entire event "body"` (which won't be a complete body anyhow).

Author: The advantage here is that this is similar to what the homeserver has saved. The homeserver is basically delaying the event so it might just cache the request until one of the futures is triggered.
So this is very lightwait for the homeserver to compose. (Its just the cached http request + response basically.)

How exactly do you think the json should look like?
maybe it makes sense to compose the properties in a simpler manner.

Author: ```
"path":"/rooms/{roomId}/send_future/{eventType}",
"content": eventContent,
// always present
"send_token": "token",
"cancel_token": "token",
"refresh_token": "token"|undefined,
"future_group_id": "group_id",
"future_timeout": number | undefined,
```
?

Reviewer 4: I've clarified what the body/content field contains in [`114da1e`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/114da1e31e753722a7a0643b478b9f63dbdd5389).

The `GET` response format has also evolved significantly and I believe incorporates all the feedback here.

n.b. it no longer exposes the full `path` and instead does the constituent fields.

    "response":{
      // always present
      "send_token": "token",
      // optional if there is a timeout
      "future_group_id": "group_id",
      "refresh_token": "token",
      "cancel_token": "token",
    }
  },
]
```

This can be used so clients can optionally display events
that will be send in the future.
And to acquire cancel_tokens for then.

For all usecases where the existence of a running future is also of interest for other room members,
(like self-destructing messages) it is recommended to include
this information in the effected event itself.

## Usecase specific considerations

### MatrixRTC

In this section an overview is given how this MSC is used in [MSC4143: MatrixRTC](https://github.com/matrix-org/matrix-spec-proposals/pull/4143)
and alternative expiration systems are evaluated.

#### Background

MatrixRTC makes it necessary to have real time information about the current MatrixRTC session.
To properly display room tiles and header in the room list (or compute a list of ongoing calls) need to know:

- If there is a running session.
- What type that session has.

> - Who and how many perople are currently participating.
> 
> A particular delicat situation is that clients are not able to inform others if they loose connection.

Reviewer 2: ```suggestion
- Who and how many people are currently participating.

A particular delicate situation is that clients are not able to inform others if they lose connection.
```

There are numerous approaches to solve such a situation. They split into two categories:

- Polling based
  - Ask the users if they are still connected.
  - Ask an RTC backend (SFU) who is connected.
- Timeout based

>   - Update the room state every x seconds. This allows clients to check how long an event has not been updated and ignore it if its expired.
>   - Use Future events with a 10s timeout to send the disconnected from call in less then 10s after the user is not anymore pingin the `/refresh` endpoint. (or delegate the disconnect action to a service attached to the SFU)

Reviewer 2: ```suggestion
  - Update the room state every x seconds. This allows clients to check how long an event has not been updated and ignore it if it's expired.
  - Use Future events with a 10s timeout to send the disconnected from call in less then 10s after the user is not anymore pinging the `/refresh` endpoint. (or delegate the disconnect action to a service attached to the SFU)
```

  This significantly reduces the amount of calls for the `/future` endpoint since the sfu only needs to ping
  once per session (per user) and every 2-5hours (instead of every `X` seconds.)

### Self-destructing messages

This MSC also allows to implement self-destructing messages:

First send (or generate the pdu when [MSC4080: Cryptographic Identities](https://github.com/matrix-org/matrix-spec-proposals/pull/4080) is available):
`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{eventType}/{txnId}`

```json
{
  "m.text": "my msg"
}
```

then send:
`PUT /_matrix/client/v3/rooms/{roomId}/send_future/{eventType}/{txnId}?timeout={10*60}&future_group_id={XYZ}`

```json
{
  "redacts": "{event_id}"
}
```

This would redact the message with content: `"m.text": "my msg"` after 10minutes.

## Potential issues

## Alternatives

### Reusing the `send`/`state` endpoint

Since the `send_future` and `state_future` endpoints are almost identical to the
normal `send` and `state` endpoints it comes to mind, that one could reuse them and allow adding the
query parameters `?future_timeout={timeout_duration}&future_group_id={group_id}` directly to the
`send` and `state` endpoint.

> This would be elegant but since those two endpoint are core to Matrix changes to them might
> be controversion if their return value is altered.

Reviewer 2: ```suggestion
This would be elegant but since those two endpoint are core to Matrix, changes to them might
be controversial if their return value is altered.
```

Currently they always return

```json
{
  "event_id":string
}
```

as a non optional field.

When sending a future the `event_id` would not be available:

- The `event_id` is using the [reference hash](https://spec.matrix.org/v1.10/rooms/v11/#event-ids) which is
  [calculated via the essential fields](https://spec.matrix.org/v1.10/server-server-api/#calculating-the-reference-hash-for-an-event)
  of an event including the `origin_server_timestamp` as defined in [this list](https://spec.matrix.org/v1.10/rooms/v11/#client-considerations)
- Since the `origin_server_timestamp` should be the timestamp the event has when entering the DAG (required for call duration computation)
  we cannot compute the `event_id` when using the send endpoint when the future has not yet resolved.

As a result the return type would change to:

```json
{
  "event_id": string | undefined,
  "future_group_id": string | undefined,
  "send_token": string | undefined,
  "refresh_token": string | undefined,
  "cancel_token": "string | undefined
}
```

> dependent on the query parameters.

Reviewer 2: It would be nice if we could express the response type as a union between the "just an event ID" response type & the response to future requests.  But since all the fields are top-level, I don't think there's a good way to do that.

Reviewer 2: True, the language of the spec has more flexibility in how to describe response types & optional fields.

### Batch sending futures with custom endpoint

> The proposed solution does not allow to send multiple events/futures that belong to each other with one

Reviewer 2: ```suggestion
The proposed solution does not allow to send events together with futures that reference them with one
```

HTTPS request. This is desired for self-destructing events and for MatrixRTC room state events, where
we want the guarantee, that the event itself and the future removing the event both reach the homeserver

> with one request. Otherwise there is a rist for the client to loose connecting or crash between sending the
> event and the future which results in never expiring call memberhsip or never destructing self-destructing messages.

Reviewer 2: ```suggestion
with one request. Otherwise there is a risk for the client to lose connection or crash between sending the
event and the future which results in never expiring call membership or never destructing self-destructing messages.
```

This would be solved once [MSC4080](https://github.com/matrix-org/matrix-spec-proposals/pull/4080) and the `/send_pdus` endpoint is implemented.
(Then the `future_timeout` and `future_group_id` could be added
to the `PDUInfo` instead of the query parameters and everything could be send at once.)

> This would be the preferred solution since we currently don't have any other batch sending mechanism.
> before

Reviewer 2: ```suggestion
This would be the preferred solution since we currently don't have any other batch sending mechanism.
```

It would however require lots of changes since a new widget action for futures would be needed.
With the current main proposal it is enough to add a `future_timeout` and `future_group_id` parameter to the send message widget action.
The widget driver would then take care of calling `send` or `send_future` based on the presence of those fields.

An alternative to the proposed solution that allows this kind of batch sending would be to
introduce this endpoint:
`PUT /_matrix/client/v3/rooms/{roomId}/send/future/{txnId}`
It allows to send a list of event contents. The body looks as following:

```json
{
  "timeout": 10,

>   "group_id": group_id,

Reviewer 2: Since the timeout & action events are already batched, could the group ID be removed from the batch request?

Author: Yes this is how it was intended. Good find!

here.

### Naming

The following alternative names for this concept are considered

- Future
- DelayedEvents
- RetardedEvents
- PostponedEvents

## Security considerations

We are using an unauthenticated endpoint to refresh the expirations. Since we use
generated tokens it is hard to guess a correct request and force sending one
of the Futures. (The homeserver has them but they can always send events in your name
as long as we do not have [MSC4080: Cryptographic Identities](https://github.com/matrix-org/matrix-spec-proposals/pull/4080))

It is an intentional decision to not provide an endpoint like
`PUT /_matrix/client/v3/futures/room/{roomId}/event/{eventId}`
where any client with access to the room could also `end` or `refresh`
the expiration. With the token the client creating the future has ownership
over the expiration and only intentional delegation of that ownership
(sharing the token) is possible.

On the other hand the token makes sure that the instance gets as little
information about the Matrix metadata of the associated `future` event. It cannot
even tell with which room or user it is interacting or what the token does (refresh vs send).

## Unstable prefix

> Use `io.element.` instead of `m.` as long as the MSC is not stable.

Reviewer 2: I think there's also a policy of prepending all endpoint paths with `/org.matrix.msc###/` (and setting the path's version string to `unstable`, instead of `v#`)

We call those delayed events `Futures`.

New endpoints are introduced:

`PUT /_matrix/client/v3/rooms/{roomId}/send_future/{eventType}/{txnId}?future_timeout={timeout_duration}&future_group_id={group_id}`

`PUT /_matrix/client/v3/rooms/{roomId}/state_future/{eventType}/{stateKey}?future_timeout={timeout_duration}&future_group_id={group_id}`

Those behave like the normal `send`/`state` endpoints except that that they allow
to define `future_timeout` and `future_group_id` in their query parameters.

- `future_timeout: number` defines how long (in milliseconds) the homeserver will wait before sending
  the event into the room. **Note**, since the timeout can be refreshed and sending the future can be triggered via an endpoint (see: [Proposal/Delegating futures](#delegating-futures)) this value is not enough to predict the time this event will arrive in the room.
  - If this query parameter is not added the future will never expire and can only be send by the [external delegation endpoint](#delegating-futures).
    We call such a future **action future**.
  - If set to a `number` (ms) we call the future **timeout future**

> - `future_group_id: string` is optional if a `future_timeout` is a `number`. The purpose of this identifier is to group

Reviewer 5: ```suggestion
- `future_group_id: string` is optional if a `future_timeout` parameter is set. The purpose of this identifier is to group
```

>   **multiple futures in one mutually exclusive group**.

Reviewer 2: These aren't currently scoped per user, so it'd be possible to share the group ID to another user who could then add their own futures to that group.  Is that something we want to allow?  Given that the group ID must be explicitly shared, this shouldn't be a security issue, and might even allow for interesting use-cases; but maybe it's an overreach of what future groups are meant to be.

Author: Since the HS is issuing the group ids it s the homeservers job to make then unique.

I think that is mentioned somewhere. That the homeserver needs to issue unique ids

Reviewer 2: As we discussed out-of-band, we do want to scope the `group_id`s to users, at least for these reasons:
- Not scoping `group_id`s raises the question of whether they should be available to all federated users or only local users, which adds complexity either way.  Without scoping, future management can avoid needing federation.
- Having a `group_id` effectively grants full control over the future group (as one could hijack the group by adding a short timeout future to it), so allowing users to share them would give each user full control over them, which would subvert the spirit of the guarantees we want futures & future groups to have.

    An `unknown_token` either means that the service is making something up or that the service is using a
    token that is invalidated by now.
  - If a timer times out, **it sends the timeout future**.
  - If the homeserver receives a _new state event_ with the same state key as existing futures the
    **futures get invalidated and the associated timers are stopped**.

    - There is no race condition here since a possible race between timeout and
      the _new state event_ will always converge to the _new state event_:
      - Timeout -> _new state event_: the room state will be updated twice. once by
        the content of the future but later with the content of _new state event_.
      - _new state event_ -> timeout: the _new state event_ will invalidate the future.

- After the homeservers sends a timeout future or action future, the associated
  timer and tokens is canceled/deleted.

So for each `future_group_id`, the homeserver will at most send one timeline event.

- No timeline event will be send in case all of the timeout futures in a future group are cancelled via `/_matrix/client/v3/futures/{cancel_token}`.

> - Otherwise one of the timeout or action futures will be send.

Reviewer 2: Do we want to associate futures with the device that created them?  Namely, if a future was created by a particular device, and that device later logs out / is invalidated, should the future be cancelled?

Reviewer 2: FTR, my intuition is that device invalidation should _not_ cancel a future, as tying future lifetime to device lifetime makes it much more important for users not to lose their device sessions.  It would also lessen the utility of futures if all it takes to render them ineffective is to log out, which would heavily impact some use-cases of futures (like scheduled messages).

Author: FTR: we as described by @AndrewFerr we do not want futures to expire if a device logs out.

Both of the query parameters are optional but one of them has to be present.
This gives us the following options:

```
?future_timeout=10 - a timeout future in a new future group
?future_timeout=10&future_group_id="groupA" - a timeout future added to groupA
?future_group_id="groupA" - an action future added to groupA
```

Possible error responses are all error responses that can occur when using the `send` and `state` endpoint accordingly and:

- The server will respond with a [`409`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409)
  (`Conflict` This response is sent when a request conflicts with the current state of the server.) if
  the client tries to send an action future without there being a timeout future with the same `future_group_id`
- The server can optionally configure a maximum `timeout_duration`
  (In the order of one week dependent on how long they want to track futures)
  The server will respond with a [`400`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) (`Bad Request`, with a message
  containing the maximum allowed `timeout_duration`) if the
  client tries to send a timeout future with a larger `timeout_duration`.

> - The future is using a group_id that belongs to a future group from another user. In this case the homeserver sends a [`405`] (`Not Allowed`).

Reviewer 2: Even better (and what I've been implementing) is for each user to get their own namespace of group IDs, i.e. for the effective group ID to be the tuple of (user ID, group ID).  Benefits include:
- Users cannot guess at other users' group IDs by spamming future requests and waiting to receive an error response.
- More group IDs are available for each user, which is especially useful if we want to allow user-defined group IDs that could otherwise clash.

Reviewer 4: Group IDs are not in the latest proposal.

Since the refresh and send tokens are of the same format it is not even possible to evaluate
what that token is for when reading the https request log.
This unauthenticated endpoint allows to delegate resolving the future.
An SFU for instance, that tracks the current client connection state, could get a url that it
needs to call every X hours while a user is connected and a url it has to call once the user disconnects.
This way the SFU can be used as the source of truth for the call member room state even if the client
gets closed or looses connection and without knowing anything about the Matrix call.

The homeserver does the following when receiving a Future:

- It checks for the validity of the request (based on the `future_timeout` and the `future_group_id` query parameters)
  and returns a `409` or `400` if necessary.
- It **generates** a `send_token`, a `cancel_token` and if not provided in the request a `future_group_id` and a optionally `refresh_token` and stores them alongside the time
  of retrieval and the `timeout_duration`.
- If `future_timeout` was present, it **Starts a timer** for the `refresh_token`.

>   - If a `PUT /_matrix/client/v3/futures/{refresh_token}` is received, it
>     **restarts the timer** with the stored `timeout_duration` for the associated timeout future.
>   - If a `PUT /_matrix/client/v3/futures/{send_token}` is received, it **sends the associated action or timeout future**
>     and deletes any stored futures with the `group_id` associated with that token.
>   - If a `PUT /_matrix/client/v3/futures/{cancel_token}` is received, it **does NOT send any future**
>     and deletes/invalidates the associated stored future. This can mean that a whole future group gets deleted (see below).
>   - If a `PUT /_matrix/client/v3/futures/{unknown_token}` is received the server responds with a `410` (Gone).
>     An `unknown_token` either means that the service is making something up or that the service is using a
>     token that is invalidated by now.

Reviewer 2: ```suggestion
  - If a `POST /_matrix/client/v3/futures/{refresh_token}` is received, it
    **restarts the timer** with the stored `timeout_duration` for the associated timeout future.
  - If a `POST /_matrix/client/v3/futures/{send_token}` is received, it **sends the associated action or timeout future**
    and deletes any stored futures with the `group_id` associated with that token.
  - If a `POST /_matrix/client/v3/futures/{cancel_token}` is received, it **does NOT send any future**
    and deletes/invalidates the associated stored future. This can mean that a whole future group gets deleted (see below).
  - If a `POST /_matrix/client/v3/futures/{unknown_token}` is received the server responds with a `410` (Gone).
    An `unknown_token` either means that the service is making something up or that the service is using a
    token that is invalidated by now.
```
to match the change from `PUT` to `POST` due to non-idempotency.

  - [Usecase specific considerations](#usecase-specific-considerations)
    - [MatrixRTC](#matrixrtc)
      - [Background](#background)
      - [How this MSC would be used for MatrixRTC](#how-this-msc-would-be-used-for-matrixrtc)
    - [Self-destructing messages](#self-destructing-messages)
  - [Potential issues](#potential-issues)
  - [Alternatives](#alternatives)
    - [Reusing the send/state endpoint](#reusing-the-sendstate-endpoint)
    - [Batch sending futures with custom endpoint](#batch-sending-futures-with-custom-endpoint)
    - [MSC4018 use client sync loop](#msc4018-use-client-sync-loop)
    - [Naming](#naming)
  - [Security considerations](#security-considerations)
  - [Unstable prefix](#unstable-prefix)
  - [Dependencies](#dependencies)

<!-- /TOC -->

The motivation for this MSC is: Updating call member events after the user disconnected by allowing to schedule/delay/timeout/expire events in a generic way.
It turns out that there is a big overlap to other usecases in Matrix which can also be implemented using the proposed concept:

- Updating call member events after the user disconnected.
- Sending scheduled messages (send at a specific time).

> - Creating self-destructing events (by sending a delayed redact).

Author: Thread for highlight 1:
I wasn't aware that it is on the table that self-destructing events would use an entirely different concept to redactions.
I was coming from: https://github.com/matrix-org/matrix-spec-proposals/blob/matthew/msc2228/proposals/2228-self-destructing-events.md
Which basically does the same thing. It redacts the event based on conditions.

What I like about this proposal, is that instead of making a custom event and add logic to create synthetic redaction we generalize the concept of event delays and everything else is a completely normal redaction.

Also
> erase the event from the DAG entirely

Sounds like an anti-pattern for a distributed system. Can you guide me on where I can find information about erasing whole events from the DAG including all there metadata without federation conflicts?

> - Creating self-destructing events (by sending a delayed redact).

Reviewer 6: Erasing events isn't currently possible, but we need to either fix that or the metadata issue.

(*resolving thread as it no longer applies to the current MSC*)

## Proposal

To make this as generic as possible, the proposed solution is to allow sending events and delegate
the control of when to actually send these events to an external service or a timeout condition.
This allows a very flexible way to mark events as expired.
The sender can define what event will be sent once the timeout condition is met. For state
events the timed out version of the event would be an event where the content communicates, that
the users has left the call.

This proposal also includes a way to refresh the timeout. Allowing to delay the event multiple times.
A periodic ping of the refreshing can be used as a heartbeat mechanism. Once the refresh ping is not send
anymore the timeout condition is met and the homeserver sends the event with the expired content information.

This translate to: _"only send the event when the client is not running its program anymore (not sending the heartbeat anymore)"_
We call those delayed events `Futures`.

New endpoints are introduced:

`PUT /_matrix/client/v3/rooms/{roomId}/send_future/{eventType}/{txnId}?future_timeout={timeout_duration}&future_group_id={group_id}`

> `PUT /_matrix/client/v3/rooms/{roomId}/state_future/{eventType}/{stateKey}?future_timeout={timeout_duration}&future_group_id={group_id}`

Reviewer 7: Is there a reason that they use `v3` instead of `v1`? Those being new endpoints I would assume they use the latter.

  the event into the room. **Note**, since the timeout can be refreshed and sending the future can be triggered via an endpoint (see: [Proposal/Delegating futures](#delegating-futures)) this value is not enough to predict the time this event will arrive in the room.
  - If this query parameter is not added the future will never expire and can only be send by the [external delegation endpoint](#delegating-futures).
    We call such a future **action future**.
  - If set to a `number` (ms) we call the future **timeout future**
- `future_group_id: string` is optional if a `future_timeout` is a `number`. The purpose of this identifier is to group
  **multiple futures in one mutually exclusive group**.
  - Only one of the events in such a group can ever reach the DAG/will be distributed by the homeserver.
    All other futures will be discarded.
  - Every future group needs at least one timeout future to guarantee that all future expire eventually.
  - If a timeout future is sent without a `future_group_id` a unique identifier will be generated by the
    homeserver and is part of the `send_future` response.

>   - Group id's can only be used by one user. Reasons for this are, that this would basically allow full control over a future group once another matrix user knows the group id. It would also require to federate futures if the users are not on the same homeserver.

Author: The `group_id`'s are server generated UUID's, so I am not sure how this could be phrased like transaction ID's.

>   - Group id's can only be used by one user. Reasons for this are, that this would basically allow full control over a future group once another matrix user knows the group id. It would also require to federate futures if the users are not on the same homeserver.

Reviewer 4: Groups do not exist in the latest proposal, so marking this conversation as resolved.

Both of the query parameters are optional but one of them has to be present.
This gives us the following options:

```
?future_timeout=10 - a timeout future in a new future group
?future_timeout=10&future_group_id="groupA" - a timeout future added to groupA
?future_group_id="groupA" - an action future added to groupA
```

Possible error responses are all error responses that can occur when using the `send` and `state` endpoint accordingly and:

- The server will respond with a [`409`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409)
  (`Conflict` This response is sent when a request conflicts with the current state of the server.) if
  the client tries to send an action future without there being a timeout future with the same `future_group_id`
- The server can optionally configure a maximum `timeout_duration`
  (In the order of one week dependent on how long they want to track futures)
  The server will respond with a [`400`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) (`Bad Request`, with a message
  containing the maximum allowed `timeout_duration`) if the
  client tries to send a timeout future with a larger `timeout_duration`.

> - The future is using a group_id that belongs to a future group from another user. In this case the homeserver sends a [`405`] (`Not Allowed`).

Author: If we make all the error codes 400 we would need a body to differenciate the different kind of errors right?
Bad Request definitly also feels appropriate.

> - The future is using a group_id that belongs to a future group from another user. In this case the homeserver sends a [`405`] (`Not Allowed`).

Reviewer 8: You need a body in any case unless there's a strong justification for non-standard errors. The MSC should say errors produce [standard error responses](https://spec.matrix.org/v1.11/client-server-api/#standard-error-response) and define the `errcode` in addition to the HTTP status code. Old errors are not properly defined, but any new MSCs should not make the same mistakes as the spec did in the past and should define a separate code for each error.

> - The future is using a group_id that belongs to a future group from another user. In this case the homeserver sends a [`405`] (`Not Allowed`).

Author: Working on that right now.

> - The future is using a group_id that belongs to a future group from another user. In this case the homeserver sends a [`405`] (`Not Allowed`).

Reviewer 4: This has been incorporated into the latest draft.

```json
{
  // always present
  "send_token": "send_token",
  "cancel_token": "cancel_token",
  "future_group_id": "group_id",
  // optional, only present if its a a timeout future response
  "refresh_token": "refresh_token"
}
```

### Delegating futures

This MSC also proposes a `futures` endpoint.
The `token` can be used to call this public `futures` endpoint:
`POST /_matrix/client/v1/future/{token}`

The information required to call this endpoint is minimal so that
no metadata is leaked when sharing the refresh/send url with a third party.
Since the refresh and send tokens are of the same format it is not even possible to evaluate
what that token is for when reading the https request log.
This unauthenticated endpoint allows to delegate resolving the future.
An SFU for instance, that tracks the current client connection state, could get a url that it
needs to call every X hours while a user is connected and a url it has to call once the user disconnects.
This way the SFU can be used as the source of truth for the call member room state even if the client

> gets closed or looses connection and without knowing anything about the Matrix call.

Reviewer 4: The ability to delegate "heartbeats" to an SFU is no longer in the latest proposal, so marking this conversation as resolved.

The homeserver does the following when receiving a Future:

- It checks for the validity of the request (based on the `future_timeout` and the `future_group_id` query parameters)
  and returns a `409`, `400` or `405` if necessary.
- It **generates** a `send_token`, a `cancel_token` and if not provided in the request a `future_group_id` and a optionally `refresh_token` and stores them alongside the time

>   of retrieval and the `timeout_duration`.

Author: Added the format/security of those tokens.

We might not want them to be part of the url.
What do you think about moving those tokens into the body?

>   of retrieval and the `timeout_duration`.

Reviewer 4: "Delegation" is no longer part of this proposal so marking this conversation as resolved.

- If `future_timeout` was present, it **Starts a timer** for the `refresh_token`.

  - If a `POST /_matrix/client/v1/future/{refresh_token}` is received, it
    **restarts the timer** with the stored `timeout_duration` for the associated timeout future.
  - If a `POST /_matrix/client/v1/future/{send_token}` is received, it **sends the associated action or timeout future**
    and deletes any stored futures with the `group_id` associated with that token.
  - If a `POST /_matrix/client/v1/future/{cancel_token}` is received, it **does NOT send any future**
    and deletes/invalidates the associated stored future. This can mean that a whole future group gets deleted (see below).
  - If a `POST /_matrix/client/v1/future/{unknown_token}` is received the server responds with a `410` (Gone).
    An `unknown_token` either means that the service is making something up or that the service is using a
    token that is invalidated by now.
  - If a timer times out, **it sends the timeout future**.
  - If the homeserver receives a _new state event_ with the same state key as existing futures the

>     **futures get invalidated and the associated timers are stopped**.

Author: > to jump ahead and send it early

This is done by using the `future_send_token`. It can be sent to immediately send the future.

> Why is this limited to state events?

This additional rule for state events makes sure, that if a users want to update state A with C that they dont overwrite it if the state changed to B before the future resolves.

For room events there is no concept of a (type, key) pair that could be overwritten.

>     **futures get invalidated and the associated timers are stopped**.

Reviewer 4: I've incorporated the suggestion of not having a `send` action and instead doing a `cancel` followed by new send in [`e7d0986`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/e7d09863e92538b891f5152ab02cc955f8f419e8).

I've added a clarification on why this only applies to state events in [`e1b460a`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/e1b460a2663ed14c8efbedca26b297c48cbf0030).

There are numerous approaches to solve such a situation. They split into two categories:

- Polling based
  - Ask the users if they are still connected.
  - Ask an RTC backend (SFU) who is connected.
- Timeout based

  - Update the room state every x seconds.
    This allows clients to check how long an event has not been updated and ignore it if it's expired.
  - Use Future events with a 10s timeout to send the disconnected from call
    in less then 10s after the user is not anymore pinging the `/refresh` endpoint.
    (or delegate the disconnect action to a service attached to the SFU)
  - Use the client sync loop as a special case timeout for call member events.
    (See [Alternatives/MSC4018 (use client sync loop))](#msc4018-use-client-sync-loop))

Polling based solution have a big overhead in complexity and network requests on the clients.
Example:

> A room list with 100 rooms where there has been a call before in every room
> (or there is an ongoing call) would require the client to send a to-device message
> (or a request to the SFU) to every user that has an active state event to check if

> > they are still online. Just to display the room tile properly.

Author: Thread for highlight 3:

This is the most important reason for why we need a heardbeat like expiration system to make matrixRTC performant an realibale!

The main problem we solve here is that we DONT want the client to connect to each SFU all the time and we dont talk about calls this client is connected to. This might sound like a possbile thing todo but is more overhead that one would think (polling or connecting to a socket is not really making a difference here.)

Maybe a more detailed description of the current situation is required here:
Without MatrixRTC your client has no idea about any ongoing call when it starts up.
We introduced `call.member` state events so now we can easily read who is connected in a session in each room.
But due to the nature of state events there is no guarantee that the client will not forget to remove (set them to `{}`) after they disconnect. Each event a client misses to remove results in a room that looks like there is an ongoing call. So for each of these rooms the client has to:
 - Authenticate with the sfu (jwt service token)
 - Connect to the sfu wbsocket.
 - Check the current participants.
 - Invalidate all the member event locally. (Since they cannot write to those member events because its owned by another member they have to do this on every client.)
 It is very easy to have such a invalid event: a user pressing ctrl + w while in call in EW is enough.
 The commented section of the MSC gives an exmaple. If you have multiple rooms where there are left over member state events you need do the 3 steps for each of them individually.
 
 Workarounds we tried:
 Storing a timestamp in the state event and updating the event every 30min. This allows to compute an invalidate event without connecting to an sfu. (but only after a 30min time where you see a call that is not happening)

- When the SFU is capable of taking care of managing our connection state and we trust the SFU to
  not disconnect a really long value can be chosen (approx. 2-10hours). The SFU will then only send
  an action once the user disconnects or looses connection (it could even be a different action for both cases
  handling them differently on the client)
  This significantly reduces the amount of calls for the `/future` endpoint since the sfu only needs to ping
  once per session (per user) and every 2-5hours (instead of every `X` seconds.)

### Self-destructing messages

This MSC also allows to implement self-destructing messages:

First send (or generate the pdu when
[MSC4080: Cryptographic Identities](https://github.com/matrix-org/matrix-spec-proposals/pull/4080)
is available):
`PUT /_matrix/client/v1/rooms/{roomId}/send/{eventType}/{eventType}/{txnId}`

```json
{
  "m.text": "my msg"
}
```

then send:
`PUT /_matrix/client/v1/rooms/{roomId}/send_future/{eventType}/{txnId}?timeout={10*60}&future_group_id={XYZ}`

```json
{
  "redacts": "{event_id}"
}
```

This would redact the message with content: `"m.text": "my msg"` after 10minutes.

> ## Potential issues

Reviewer 6: The keep alive approach appears to reverse the expectation of scheduled messages a bit, which I think may be one of the core concerns with this proposal. With [MSC3277](https://github.com/matrix-org/matrix-spec-proposals/pull/3277), senders expect that their event will go out at the time they schedule it. This proposal moves the expectation to much later in the send process, and creates an assumption that any event can be cancelled or "undone" at any point. This behaviour leads to "unexpected" consequences, because the sender was expecting to be an opportunity for their event to never send.

It's a bit subtle, but that expectation changing I think makes MSC3277 more favourable. MSC3277 doesn't really help with the call participant count problem though. For that, I think the SFU can likely count its connections more reliably than a keep alive (a single network partition leads to a false count). This is discussed in more detail in the 'MatrixRTC use case' thread I've started.

Author: The core assumption of the voip writing this MSC is that it is of essence that syncing room state is enough to compute all ongoing MatrixRTC sessions. With the reasons described here: https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r1650573427

So It seems we have the following two scenarios:
 - We go with a SFU based state event validation approach (each client logs requests a token and logs into each SFU mentioned in any RTC member event it encounters and verifies its validity (if the user is still connected to the session))
  - We want the memberships/sessions to be reliably represented in the room state.
  
 If we decide for 1. we really don't need any of the MSC's and I can see how a static timeout as described in [MSC3277](https://github.com/matrix-org/matrix-doc/pull/3277) could be a solution. Even though no interaction to the scheduling definitively limits its use-cases.
 
 But the experience we made during a whole year concluded that we should make the state events reliable.
 (Other logic: historic session computation, changes in call realted UX, building matrixRTC sdk's all are super hard to implement and involve a a lot of duct tape if we cannot trust member state events but have to check the SFU all the time. It also means we can only validate state events in real time but never tell if a state event was invalid in the past.)

Reviewer 4: I believe this thread can be resolved because:

1. MSC3277 is now included in the alternatives section
2. The MSC has been renamed and wording clarified that delayed events can be cancelled

> ## Alternatives

Reviewer 6: This proposal's main competition appears to be [MSC3277](https://github.com/matrix-org/matrix-spec-proposals/pull/3277), where events are scheduled by placing them in the DAG. The homeserver then forwards the previously-created event at a set time, but it is part of the room until that point.

The DAG approach has a few advantages which I think make it more appealing:
* Message delivery is not subject to a third party being online. Specifically, an appservice, SFU, or similar does not need to be looped in to ensure a message gets sent - it's the homeserver's own responsibility to stay online. We may even be able to federate the event outwards with a `soft_fail_until` timestamp for maximum reliability (at the cost of eager delivery and harder cancellation).
* Events are known to be authorized and able to be sent because they've been given an event ID (and thus run through the auth rules).
* The tracking overhead is extremely minimal: servers at most need to remember to un-soft-fail an event after a set time, but it can do that with rough precision. They do not need to track different tokens for different actions - if a user redacts a scheduled event while it's still soft failed, the send is cancelled. It knows this intrinsically.

Author: We are very open to insert the events into the DAG when sent. This was discussed and even has a bunch of advantages (mostly that the event_id is available immediately)

The biggest reason against it was the assumption that this might be too big of a change: It adds the complexity of a state where an event is in the dag but not yet (and maybe never) is distributed to clients.

> Message delivery is not subject to a third party being online. Specifically, an appservice, SFU, or similar does not need to be looped in to ensure a message gets sent - it's the homeserver's own responsibility to stay online. We may even be able to federate the event outwards with a soft_fail_until timestamp for maximum reliability (at the cost of eager delivery and harder cancellation).

The delivery does not rely on a third party. It relies on the homeservers timeout computation. (that is enforced by the future group! The example with the client that pings every 10s describes this well. The client getting disconnected and not sending the ping anymore lets the homerserver know it can deliver the event now.

To me, it seems that the least intrusive change is to not add events to the DAG and instaed just queu them on the homeserver/pretend the http request was received later. But I really like it if it's possible to add it to the DAG right after receiving the event and immediately computing an `event_id`.
Because Future events can also be canceled, it needs to be valid to have non sent events in the dag forever. As described in the MSC, in most cases you want to schedule multiple events in one future group and only send one of them. Adding them all to the DAG seems unnecassary.
But other than that, this sounds like a very compatible and nice solution. Do I understand it correctly that the homeserver will not send the event to federating homeservers until the timeout condition is met?

Author: The VoIP team had a dedicated meeting to thoroughly investigate the option to use MSC3277's approach of adding the event to the DAG on retrieval.

We have created a summary document here: https://hackmd.io/h0z82KvKSaiW-jYlOnU69w?edit
(this will be posted below as well for visibility)

Author: # Reliable State events (Future MSC)

The Future MSC is not sending 

### Soft fail (DAG) vs Homeserver queuing (not in DAG)
MSC4140 and MSC3277

https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r1649563962
> Message delivery is not subject to a third party being online. Specifically, an appservice, SFU, or similar does not need to be looped in to ensure a message gets sent - it's the homeserver's own responsibility to stay online. We may even be able to federate the event outwards with a soft_fail_until timestamp for maximum reliability (at the cost of eager delivery and harder cancellation).

This is the case for both proposals. The delivery itself is in full control of the homeserver.

For both proposals the homserver is given additional information to the event and will make it real eventually based on those parameters.

The main differentiator is that with MSC4140 external interaction is optionally possible.
The `/future/{futureToken}` endpoint allows interacting with event scheduling.

**This is the main feature we require to get Reliable VoIP to work.**

It is not important if we add the event to the DAG immediately, but we need the scheduling interaction of `/future/{futureToken}`.

> Events are known to be authorized and able to be sent because they've been given an event ID (and thus run through the auth rules).

We explicitly only have one auth period: at send time. While scheduled, the auth conditions can change, so the proper time to do auth computation is at send time.

With MSC3277 we would need to rerun the auth rules when sending the event.

> The tracking overhead is extremely minimal: servers at most need to remember to un-soft-fail an event after a set time, but it can do that with rough precision. They do not need to track different tokens for different actions - if a user redacts a scheduled event while it's still soft failed, the send is cancelled. It knows this intrinsically.

Using this with the required interaction tokens would result in the following:
 - Send a scheduled event with 10s
 - On each refresh/reset token retrival the homserver would redact the event and send a new one (with the send_at updated by 10s)
     - This would become quite complicated with cryptographic identities, since the user would need sign all of them. Each reset/refresh needs a client server roundtrip for signing.
 - Whenever there is no reset/refresh for 10 seconds (e.g. the device has crashed) the homeserver would un-soft-fail (send) the event.
     - Automatically
The main issue we have with this is the large trafic that generates on the homeserver. For clients this would be equivalent (at least without cryptographic identities) but the homeserver needs to insert a new event (and send it over federation) for each refresh.

Comment:
 - If we delegate leaves to the SFU we would configure the timeout to a duration in the hour range. This would make the above much more bearable.
 - We still need to decide if we want to federate the delegation tokens. Otherwise we dont get any benefit from federating the scheduled event if we can only interact with it through the sending homeserver.

Comparison
||MSC3277|MSC4140|
|-|--|--|
Traffic|Lots of traffic (one evet per refresh) between federating homeservers. For the client it is the same|Since we do not add the event to the DAG until we know it will be sent, there is no traffic because of the scheduling.|
resilience (only scheduling)|Scheduled events are even sent if the sender's homeserver goes down.|The sender's homeserver is required at send time to send the event. If it is down at sent time the event will not be sent until it comes back up.|
resilience (with interaction)|This is just not possible because sending an event (based on an interaction) before the `at` time requires redacting the scheduled one and creating a new one, which cannot be done by a different homeserver then the one hosting the sending user.| Cannot be done. It would also require federting all the interaction tokens and the unsent events but we end up in the same scenario where a homeserver would need to send an event for a user on a different homerserver.
Federation (without interaction) | Works|Doesn't work
Federation (with interaction)|We need interaction tokens (`futureTokens`) and they cannot be federated. Otherwise federating homeserver admins could interact with the Futures/scheduled messages. **With interaction there is no way we can make federation work and we lose all the benefit of inserting the event into the DAG.**|The same, we cannot ever send `futureTokens` to other homeservers since they would be able to interact with the futures. 
Auth | Here interaction could work by redacting the scheduled event and sending a new scheduled event. This is authenticaed but only the client can send the events. If we want interaction that can be delegated to the SFU (which is the best source of truth) we would need to introduce tokens similar to `futureTokens`.| With the `futureToken` based authentication (that is different to the client auth token) we have an extremely scoped auth mechanism that can be send to third parties like the SFU. The SFU can notify the homeserver about a user disconnect. The SFU is the best source of truth for that.
Tracking overhead|All the schedule tracking information is stored in the DAG and a timer needs to run through all potential expired scheduled messages periadically. This has to happen on ALL homeservers.|Very similar logic to the scheduled events except that the tracking information is stored in a seperate Future database only on the sender's homeserver (less overhead).

## TLDR:
If we don't need interaction, there is a benefit in federating the scheduled message (adding it to the DAG immediately). It increases resiliance: the sender's homeserver can disconnect and the scheduled message still will enter non-soft-failed state (will be sent).

With delegatable interaction (which is the one property we need for reliable state events), we lose the possiblity to federate scheduled messages* and the two solutions would converge to one where we lose the property of the sender's homeserver being able to lose connection but the scheduled message still is sent on other homeservers.

_\* reason being that this would also require to federate (and hence leak) the interaction tokens and allowing other homeservers to interact with the future without the user explicitly giving consent by sending the tokens.)_

#### Conclusion
We are faced with the decision between:
 - Do not assume that the senders homeserver stays online:
    - Federate the scheduled message (reslience) but loose the interaction we need for MatrixRTC reliable call member events.
 - Assume that the senders homeserver stays online:
    - If the senders homeserver is expected to stay online there is no reason to federate the scheduled/future event and we can safely add interaction which allows us to implement reliable call member events

Reviewer 4: I've captured what I think is the essence of this thread in [`4a2ca48`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/4a2ca4800d6e1f0debad654a8812d2e7ffc1bbd8).

### Reusing the `send`/`state` endpoint

Since the `send_future` and `state_future` endpoints are almost identical to the
normal `send` and `state` endpoints it comes to mind, that one could reuse them and allow adding the
query parameters `?future_timeout={timeout_duration}&future_group_id={group_id}` directly to the
`send` and `state` endpoint.

This would be elegant but since those two endpoint are core to Matrix, changes to them might  

> be controversial if their return value is altered.

Reviewer 6: I think this would be fine. Clients would only get a different response if they use the parameters, meaning they should be expecting a different format. It would be different if *lack of* the future parameters meant a different response body.

Author: That is really nice!
It would be as you describe:
Without future parameters the response would be the same as it is now.
With the parameters the response would be different.

Reviewer 4: Resolving this conversation as it matches the latest proposal.

`PUT /_matrix/client/v1/rooms/{roomId}/send/future/{txnId}`

```json
{
  "m.send_now":{
    "type":"m.room.message",
    "content":{
      "m.text": "my msg"
    }
  },
  "m.timeout": 10*60,
  "m.send_on_timeout": {
    "type":"m.room.readact",
    "content":{
      "redacts": "$m.send_now.event_id"
    }
  }
}
```

With cryptographic identities events would be presigned.
The server will first send the finalized event to the client.
At this point the client has the id but the event is not in the DAG.
So it would be trivial to sign both the event and the redaction/related event
and then send them via `/send_pdus` (see: [MSC4080: Cryptographic Identities](https://github.com/matrix-org/matrix-spec-proposals/pull/4080)).

### MSC4018 (use client sync loop)

[MSC4018](https://github.com/matrix-org/matrix-spec-proposals/pull/4018) also
proposes a way to make call memberships reliable. It uses the client sync loop as
an indicator to determine if the event is expired. Instead of letting the SFU
inform about the call termination or using the call app ping loop like we propose
here.

### Naming

The following alternative names for this concept are considered

- Future
- DelayedEvents

> - RetardedEvents

Reviewer 6: ```suggestion
```

We do not use this word.

      "event_type": "m.room.message",
      "timeout": 15000,
      "running_since": 1721732853284,
      "content":{
        "msgtype": "m.text",
        "body": "I am now offline"
      }
    },
    {
      "future_id": "abcdefgh",
      "room_id": "!roomid:example.com",
      "event_type": "m.call.member",
      "state_key": "@user:example.com_DEVICEID",
      "timeout": 5000,
      "running_since": 1721732853284,
      "content":{
        "memberships": []
      }
    }
  ]
}
```

`running_since` is the timestamp (as unix time in milliseconds) when the delayed event was scheduled or last refreshed.
So, unless the delayed event is updated beforehand, the event will be sent after `running_since` + `timeout`.

This can be used by clients to display events that have been scheduled to be sent in the future.

For use cases where the existence of a delayed event is also of interest for other room members,
(e.g. self-destructing messages) it is recommended to include this information in the original/affected event itself.

### Managing delayed events

A new authenticated client-server API endpoint at `POST /_matrix/client/v1/futures/{future_id}` allows scheduled events

> to be managed.

Reviewer 2: How about putting the `future_id` in the request body (to avoid it appearing in web request logs)?

It also feels awkward to use both a path parameter & a body parameter (for the action).

Reviewer 2: I rescind this request: as we discussed, there's precedent in using the response of a `GET` endpoint as a path parameter in a `POST` method for the same endpoint path (which is already done for event/room IDs).

The recommended strategy is to retain finalised events for up to 7 days or 1000 events per user,
whichever occurs first.

There is no guarantee for a client that events will be available
if they exceed the limits of their homeserver.
Additionally, a homeserver MAY discard finalised delayed events that have been returned by a
`GET /_matrix/client/v1/delayed_events/{delay_id}` response.

### Additional homeserver behaviour

#### `delay_id` in `unsigned` event data
The `delay_id` of a sent delayed event MUST be included in the resulting room event's `unsigned` data
if, and only if, the client being given the event (e.g. over `/sync` or `/messages`) is authenticated as the event's sender.

#### No `transaction_id` in `unsigned` event data
As a delayed event is uniquely identified by its sender and `delay_id`,
it is redundant to further distinguish it with the `transaction_id` used in the request to schedule it.
As such, the resulting room event of any sent delayed event MUST NOT have a `transaction_id` in its `unsigned` data
or elsewhere, unlike non-delayed events which do include `transaction_id` in `unsigned` for their sender.

> #### Event content is evaluated at the point of sending

Reviewer 9: I think this is a copy and paste error?

```suggestion
#### Event content is evaluated at the point of scheduling
```

Reviewer 2: Whoops, thanks.  f71394f1

### Scheduling a delayed event

We propose extending the existing
[`PUT /_matrix/client/v3/rooms/{roomId}/state/{eventType}/{stateKey}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
and
[`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidstateeventtypestatekey)
endpoints by adding an optional `future_timeout` query parameter.

The new query parameter is used to configure the event scheduling:

- `future_timeout` - Optional number of milliseconds the homeserver should wait before sending the event. If no `future_timeout` is provided, the event is sent immediately as normal.

The body of the request is the same as currently.

If a `future_timeout` is provided, the homeserver schedules the event to be sent with the specified delay and returns the _future ID_ in the `future_id` field (omitting the `event_id` as it is not available):

```http
200 OK
Content-Type: application/json

{
  "future_id": "1234567890"
}
```

The homeserver can optionally enforce a maximum timeout duration. If the requested timeout exceeds the maximum the homeserver

> can respond with a [`400`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) and a Matrix error code `M_FUTURE_MAX_TIMEOUT_EXCEEDED` and the maximum allowed timeout (`timeout_duration` in milliseconds).

Reviewer 4: The field is now called `max_delay`. I think `M_INVALID_PARAM` could work as an alternative. I don't think it makes a difference at this stage to the proposal. It could be resolved during FCP if required.

> can respond with a [`400`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) and a Matrix error code `M_FUTURE_MAX_TIMEOUT_EXCEEDED` and the maximum allowed timeout (`timeout_duration` in milliseconds).

Reviewer 4: Actually, I've added this to the alternatives section.

For example the following specifies a maximum delay of 24 hours:

```http
400 Bad Request
Content-Type: application/json

{
  "errcode": "M_FUTURE_MAX_TIMEOUT_EXCEEDED",
  "error": "The requested timeout exceeds the allowed maximum.",
  "timeout_duration": 86400000
}
```

### Getting delayed events

We propose adding a new authenticated endpoint `GET /_matrix/client/v1/futures` to the client-server API to allow clients to get a list of all the events that have been scheduled to send in the future.

```http
HTTP 200 OK
Content-Type: application/json

{
  "futures": [
    {
      "future_id": "1234567890",
      "room_id": "!roomid:example.com",
      "event_type": "m.room.message",
      "timeout": 15000,
      "running_since": 1721732853284,
      "content":{
        "msgtype": "m.text",
        "body": "I am now offline"
      }
    },
    {
      "future_id": "abcdefgh",
      "room_id": "!roomid:example.com",

>       "event_type": "m.call.member",

Reviewer 4: I've incorporated this in to [`93c932a`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/93c932aa01faed18b81048fb30c2fb781b78a50e)

metadata for a room to be synchronous.

The current solution updates the room state every X minutes.
This is not elegant since we basically resend room state with the same content.
In large calls this could result in huge traffic/large DAGs (100 call members
implies 100 state events every X minutes.) X cannot be a long duration because
it is the duration after which we can consider the event as expired. Improper
disconnects would result in the user being displayed as "still in the call" for
X minutes (we want this to be as short as possible!)

Additionally this approach requires perfect server client time synchronization to compute the expiration.
This is currently not possible over federation since `unsigned.age` is not available over federation.

#### How this MSC would be used for MatrixRTC

With this proposal we can provide an elegant solution using actions and timeouts
to only send one event for joining and one for leaving (reliably)

- If the client takes care of its membership, we use a short timeout value (around 5-20 seconds)
  The client will have to ping the refresh endpoint approx every 2-19 seconds.

> - When the SFU is capable of taking care of managing our connection state, and we trust the SFU to

Author: I think since this is not proposing any change to the spec (leaving it out would not change the msc) and the delegation itself probably will not be an msc (there will be an msc for the scoping by endpoints but the usage for matrix rtc will just be on the implementation level) we can keep it in here for context. Except if someone suspects this makes the msc harder to understand and to review. I have the opinion that this is giving interesting context and answers all concerns how and if a sfu driven disconnect can be possible.

> - When the SFU is capable of taking care of managing our connection state, and we trust the SFU to

Reviewer 5: sorry I was ref to 

> When the SFU is capable of taking care of managing our connection state, and we trust the SFU to not disconnect, a really long value can be chosen (approx. 2-10hours). The SFU will then only send an action once the user disconnects or loses connection (it could even be a different action for both cases handling them differently on the client) This significantly reduces the amount of calls for the /update_future endpoint since the SFU only needs to ping once per session (per user) and every 2-5hours (instead of every X seconds.)

> - When the SFU is capable of taking care of managing our connection state, and we trust the SFU to

Author: That is the context I was commenting on.
With scoped tokens this is no additional MSC.  When a client can request a token the SFU can get this token and use it as any other client to send the request as specified in this MSC here.
Hence there might not even be an MSC for it. The MatrixRTC MSC will just make it clear that the state has to be updated once the user disconnects and to have a "guarantee" that the user will disconnect eventually a future has to be used. If a third party get scoped access to the the future sending this is just a more efficient client implemantation to fullfill the requirement that the state is updated on dicsonnect.
So mentioning it here might be good to give context that this is a possible usecase.

> - When the SFU is capable of taking care of managing our connection state, and we trust the SFU to

Reviewer 4: This reference to delegation has been removed in [`9098fea`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/9098fea5e5bc31d900c5ad4db92c604f6293df86).

with an `errcode` of `M_LIMIT_EXCEEDED`, and a `Retry-After` header whose value is set to the time of/until
the scheduled send time of the next of the user's delayed events to be sent,
rounded up to the nearest second.

```http
429 Too Many Requests
Content-Type: application/json
Retry-After: 1200

{
  "errcode": "M_LIMIT_EXCEEDED",
  "error": "The maximum number of delayed events has been reached.",
}
```

If the feature is disabled on the server via configuration, the homeserver SHOULD instead respond with HTTP 403 and
a [standard error response](https://spec.matrix.org/v1.19/client-server-api/#standard-error-response)
with an `errcode` of `M_FORBIDDEN`.

#### Delayed event limits as a capability

The values of both the maximum allowed delay and the maximum allowed number of scheduled events are advertised as a
[capability](https://spec.matrix.org/v1.19/client-server-api/#capabilities-negotiation) named `m.delayed_events`, via
the values of non-negative integer valued fields named `max_delay_ms` and `max_scheduled` respectively.
For any of these limits enforced by the server, its representative field MUST be present in the capability.

> If the server doesn't enforce one of these limits, its representative field MUST be absent from the capability.

Reviewer 10: This framing does mean it's impossible for a server to say that delayed events are disabled. I'm not sure that is something we want to support, but equally feels odd. I'm not sure if it's worth doing anything about it though

Reviewer 2: It can communicate disabled delayed events by responding with:
```json
{
  "capabilities": {
    "m.delayed_events": {
      "max_delay_ms": 0,
      "max_scheduled": 0
    }
  }
}
```
Even including only one of the `max_` fields would be sufficient, as it's enough for either one of them to be 0 to mean that delayed events are disabled.

Does that cover your concern?

Reviewer 10: Mmm, true I suppose. Though if we want clients to consider `0` to mean disabled then that's something we should be specific. While technically a limit of 0 is the same, it needs to be handled specially by clients (i.e. this operation will never work no longer how long we wait). 

OTOH I'm not sure if we *want* to allow delayed events to be disabled by the server? I guess that is the fundamental question, and we should either explicitly allow or deny `0` as the limit.

(This is a very low severity thing)

Reviewer 2: > While technically a limit of 0 is the same, it needs to be handled specially by clients (i.e. this operation will never work no longer how long we wait).

Aha, this is what that ["special case"](https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3658009434) I was banging on about was for.  It landed as ec7cbfd.

> OTOH I'm not sure if we _want_ to allow delayed events to be disabled by the server?

IMO there's no harm to allow disabling it, especially if server configs already allow you to constrain it (why not allow constraining it all the way to 0?).

As another point, this section should also specify what clients should assume if the `m.delayed_events` capability is absent entirely (not just its fields).

af63014d does so, and also explicitly specifies what 0 limits mean.

#### No `transaction_id` in `unsigned` event data
As a delayed event is uniquely identified by its sender and `delay_id`,
it is redundant to further distinguish it with the `transaction_id` used in the request to schedule it.
As such, the resulting room event of any sent delayed event SHOULD NOT have a `transaction_id` in its `unsigned` data
or elsewhere, unlike non-delayed events which do include `transaction_id` in `unsigned` for their sender.

#### Event content is evaluated at the point of scheduling

At the time of scheduling a delayed event, the homeserver MUST validate that the event is well-formed, lest the event
has no chance to be successfully sent into a room upon its scheduled send time.

#### Power levels are evaluated at the point of sending

Power levels are evaluated for each event only once the delay has occurred and it will be distributed/inserted into the
DAG. This implies a delayed event can fail if it violates power levels at the time the delay passes.

Conversely, it's also possible to successfully schedule an event that the user has no permission to send at the time of

> sending. If the power level situation has changed at the time the delay passes, the event can even reach the DAG.

Reviewer 10: I think we should allow servers to reject scheduled events if the user doesn't have permission to send them at schedule time. It feels somewhat perverse to a) not error immediately, and b) allow stacking up delayed events in arbitrary rooms.

We shouldn't scope this to just power levels, it applies to all auth rules. We're basically saying "handle the event as if the client manually sent it at the scheduled time". This is important as servers could  construct and validate  the event when either a) it's scheduled, or b) it's sent. We want the latter.

I think it's fine to not handle the edge case where the room state changes between schedule and send, i.e. if it the event was allowed initially then it's fine not to check auth again until it comes to send time.

Maybe something like:

```suggestion
#### Auth rule evaluation

When a server comes to send a scheduled event it MUST construct the final event based on the state
of the room at the time of sending (NOT when it is scheduled), as if the client itself sent the
event. In particular, the server MUST check the event is authorized to be sent based on the state at
the time the server sends the event.

The server SHOULD also check the event could be sent when it is scheduled and reject the request if
not. In particular, servers SHOULD check the sender is in the room (or otherwise authorized to send
the event) to avoid storing events for rooms the user is not in.

> [!NOTE]
>
> Servers do not need to reevaluate the auth checks for delayed events in a room when the state
> changes, it can defer checks until the event comes to be sent.
```

Reviewer 2: > It feels somewhat perverse to ... allow stacking up delayed events in arbitrary rooms.

Good point; that is indeed something worth guarding against.

> We're basically saying "handle the event as if the client manually sent it at the scheduled time".

An exception should be made for rate-limiting rules, i.e. rate limits on event sending should apply on delayed events only on their send time, not on their schedule time.

Or, does that distinction not need to be made here, since rate limits are a different topic than auth rule evaluation?  (At least I don't see rate limits mentioned anywhere in (one version of) the spec for [auth rules](https://spec.matrix.org/v1.19/rooms/v12/#authorisation-rules).)

Reviewer 2: Actually, how to apply rate limiting is already made clear in the very next section ([Rate-limiting at the point of sending](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/expiring-events-keep-alive/proposals/4140-delayed-events-futures.md#rate-limiting-at-the-point-of-sending)).  So between that & auth rules not being about rate limits, let's call this resolved.

Another possible use case are reliable "hang up" events in VoIP calls. [MSC4143], for instance,
communicates call membership through timeline events. Leaving a call requires sending a new
event. The leaving client may not be able to send this event on its own, however, as it may lose
connectivity due to network issues or because the application was shutdown unexpectedly.
In this situation it would be helpful, if the client could schedule its "hang up" event to be
sent by the server at a defined later time. While the client is still connected and in the call,
it could repeatedly push the scheduled time forward as a kind of "heartbeat" mechanism. If the
client then loses connectivity, the server would emit the "hang up" event at the scheduled time
resulting in reliable call membership status for other participants.

[MSC4143]: https://github.com/matrix-org/matrix-spec-proposals/pull/4143

This proposal caters to the use cases described above and introduces a mechanism by which a
Matrix client can schedule "delayed events" which will be sent into a room at a later time by
the homeserver. This includes APIs for scheduling delayed events and managing scheduled delayed
events.

> [Selective Forwarding Units (SFUs)]: https://trueconf.com/blog/wiki/sfu

Reviewer 11: nit: this link reference definition seems to be orphaned

Reviewer 2: bee5de6f

  Using [timestamp massaging](https://spec.matrix.org/v1.18/application-service-api/#timestamp-massaging)
  does not affect the value of this field.

Whether a delayed event is still scheduled, or has been sent, failed to be sent due to an error, or was cancelled
can be determined by examining which of the optional fields are present in the response object:

- If `finalised_ts` is absent, then the delayed event is still scheduled.
- Otherwise, if `event_id` is present, then the delayed event has been sent.
  - If `finalised_ts` < `scheduled_at` + `delay_ms`, then the event was sent manually by
    [the `/send` endpoint](#managing-scheduled-delayed-events); otherwise, it was sent on its scheduled send time.
- Otherwise, if `error` is present, then the delayed event failed to be sent (and was descheduled) due to an error.
- Otherwise, the delayed event was cancelled by [the `/cancel` endpoint](#managing-scheduled-delayed-events).

#### Getting a list of delayed events

A new authenticated Client-Server API endpoint at
`GET /_matrix/client/v1/delayed_events` responds with
a list of details about scheduled delayed events owned by the requesting user.

> Delayed events are returned in chronological order of their intended send time, which is `scheduled_at` + `delay_ms`.

Reviewer 11: I assume ascending order?

Reviewer 2: Yes; I thought "chronological" also implied ascending time.  If not, that can be clarified here.

Reviewer 2: 5a81974a

  These objects contain the same fields as the object returned by
  [the single-item lookup](#getting-a-single-delayed-event),
  except for the fields exclusive to finalised delayed events (`error`, `event_id`, and `finalised_ts`).

```http
200 OK
Content-Type: application/json

{
  "delayed_events": [
    {
      "delay_id": "...",
      "room_id": "!roomid:example.com",
      "type": "m.room.message",
      "delay_ms": 5500,
      "scheduled_at": 1721732853284,
      "content": {
        "msgtype": "m.text",
        "body": "I am now offline"
      }
    },
    ...
  ]
}
```

#### Retention of finalised delayed events

The amount of finalised events that stay on the homeserver can be chosen by the homeserver.
The recommended strategy is to retain finalised events for up to 7 days or 1000 events per user,
whichever occurs first.

There is no guarantee for a client that events will be available
if they exceed the limits of their homeserver.
Additionally, a homeserver MAY discard finalised delayed events that have been returned by a
`GET /_matrix/client/v1/delayed_events/{delay_id}` response.

### Additional homeserver behaviour

> #### `delay_id` in `unsigned` event data

Reviewer 11: Does the server also include the `transaction_id` property in `unsigned`, like it does with events sent from the other event-sending endpoints?

Reviewer 2: No, because the `transaction_id` only distinguishes requests, not the delayed events themselves.

Reviewer 11: Can you add some words to the MSC to say this?

Reviewer 2: Sure: fadd1f88

Reviewer: (I think "MUST NOT" might be slightly too strong here. We need to explain why the MSC doesn't require it but there's no need to explicitly prohibit it? Maybe just replacing "MUST NOT" with "does not" would be sufficient?)

Reviewer 2: How about `SHOULD NOT`?  Reason being, there is no benefit for servers to include the transaction ID field, because it is redundant & therefore wasteful.

Reviewer: Ok, yeah sounds fine.

Reviewer 2: Done: 256957ad

endpoint. It also makes use of [transaction identifiers](
https://spec.matrix.org/v1.18/client-server-api/#transaction-identifiers),
and supports [timestamp massaging](https://spec.matrix.org/v1.18/application-service-api/#timestamp-massaging)
when called by an application service.

The body for requests to this endpoint is a JSON object containing the following fields:

- `delay_ms` - Required. A positive non-zero number of milliseconds the homeserver should wait before sending the event.
- `state_key` - The state key for the event to be sent, if it is to be a state event; absent otherwise.
- `content` - Required. The content of the event to be sent.

The homeserver schedules the event to be sent with the specified delay and responds with an
opaque `delay_id` field (omitting the `event_id` as it is not available):

```http
200 OK
Content-Type: application/json

{
  "delay_id": "1234567890"
}
```

The `delay_id` is an [opaque identifier](https://spec.matrix.org/v1.18/appendices/#opaque-identifiers)
generated by the homeserver.

The homeserver MUST NOT send the event before the scheduled time.

> To support batch sending, homeservers MAY add up to 30 seconds to the scheduled send time.

Reviewer 12: This doesn't really match up with the 'tea timer' use case suggested in the opening section. Probably the right fix here is to just accept that's not really a use case.

Reviewer 2: Sure.  The only way to properly support a "tea timer" would be to reduce the allowed delay of scheduled send times, but that would put extra load / obligations on servers that might not always be satisfiable in practice.  So it is safer to simply not suggest that scheduled sent times are accurate enough for timer usecases.

Removed by e5f656b3

> Note: clients might find that their events are delayed further due to server load and similar conditions.

Reviewer 12: Here might be a good place to also mention that the same rate limits will apply as if the client sent the events in the standard way at the time.

Reviewer 2: ca9f55a5

The homeserver MUST persist scheduled delayed events such that they will not be lost by the homeserver being restarted.
Moreover, when a homeserver restarts, it MUST scan for all scheduled delayed events whose send time has already passed
(i.e. delayed events that were scheduled to be sent while the homeserver was offline),

> and send those delayed events as soon as possible, in chronological order of their scheduled send times.

Reviewer 12: These are implementation details for the homeserver ( and I don't think they're relevant for the spec. I think the spec should just say that HS's job is to deliver the events at the right time and it's the job of the HS implementer to make that happen. They can go in as additional / non-normative info if we think it's helpful.

Reviewer 9: Agreed, this section is very odd to include here.

Reviewer 13: Removed in fc9077af5216f3d31c4a32b551716714a5180a6d

The homeserver MAY enforce a maximum allowed delay for delayed events. This limit is
communicated to the client in a capability (described later in this proposal).
If a requested delay exceeds this maximum, the homeserver will respond with HTTP 403
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)

> with an `errcode` of `M_FORBIDDEN`.

Reviewer 12: Not critically important but 403 seems slightly odd in this case as implies the user might be able to do it if they had more rights, which probably isn't going to be the case.

Reviewer 2: I thought so too at first, but it turns out 403 can be used for more than auth errors / insufficient privileges: https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3500003217

[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threat.

The homeserver SHOULD enforce a limit of how many delayed events a user may have scheduled at once
to provide mitigation against both the
[High Volume of Messages](https://spec.matrix.org/v1.18/appendices/#threat-high-volume-of-messages) and
[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threats.
If a user's request to schedule a delayed event would exceed this limit, the homeserver will respond with HTTP 429,
a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
with an `errcode` of `M_LIMIT_EXCEEDED`, and a `Retry-After` header whose value is set to the time of/until
the scheduled send time of the next of the user's delayed events to be sent,
rounded up to the nearest second.

```http
429 Too Many Requests
Content-Type: application/json
Retry-After: 1200

{
  "errcode": "M_LIMIT_EXCEEDED",
  "error": "The maximum number of delayed events has been reached.",
}
```

As a special case, if this limit has been set to 0 such that scheduling delayed events is disallowed entirely,

> the homeserver will instead respond with HTTP 403 and a standard error response with an `errcode` of `M_FORBIDDEN`.

Reviewer 12: Again, not sure 403 feels right here, and likewise 429, whilst technically correct since the client can wait until the next event is sent, feels a bit weird.

Reviewer 9: Why special case this at all?

Reviewer 14: > likewise 429, whilst technically correct since the client can wait until the next event is sent, feels a bit weird.

That's exactly what we do for `/send` and co: it seems fine to me?

Reviewer 13: Removed the special case in 6c3911d309c33cf573867cc1b08f86c49d5fea79. It looks like we don't have a better suggestion for the error code than 429, so we would like to stick with that. Prior art for using 429 as a response to the client having too many operations ongoing in parallel is given by the media creation endpoint, I believe: https://spec.matrix.org/unstable/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid

Please reopen the thread if anyone has a better suggestion for the error code.

Reviewer 2: The aim of the special case was to avoid returning a rate-limit response when the limit cannot be bypassed by waiting; i.e. when delayed events have been disabled by server config (by setting the allowed max to 0) and no amount of waiting will help.

That special case also helps if we don't want to return `M_DELAY_TOO_LARGE` for an impossible-to-satisfy max allowed delay of 0, as discussed here: https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3483103991  Similar to 429 suggesting that waiting will help, I'd expect `M_DELAY_TOO_LARGE` to suggest that there exist delay values that are _not_ too large, which is not the case for a max of 0.  Plus, the special case would ensure that the same response would be given regardless of which config the server applied to disallow delayed events (either max delay of 0, or max number of allowed delayed events to 0).

With that said, does the special case feel okay to use here?

Reviewer: I suppose it feels sensible. But also the client could always just verify whether delayed events are disabled entirely through the capability. So I think I'm leaning slightly more towards skipping the special case to keep things simpler.

Reviewer 2: The situation I want to avoid is servers having to decide whether to return `M_DELAY_TOO_LARGE` or `M_LIMIT_EXCEEDED` whenever either (or both) of the server-side configs has effectively disabled delayed events.  So in a way, the special case provides simplicity through complexity :)

Reviewer: Ok, I see. @clokep what are your thoughts on reintroducing the special case to fix the error code?

Reviewer 9: It feels like setting the delay to 0 to disable the feature is the wrong thing to do. Maybe having a max allowed delayed events to 0 gives a more reasonable behavior?

Reviewer 2: Even a max allowed delayed event limit of 0 would benefit from a special error response, because the usual response would be a `M_LIMIT_EXCEEDED` / 429 for a rate limit that can never be satisfied.  Using a different response would make it clear that a rate limit isn't the issue in that case.

Reviewer 14: I think it's reasonable to define a specific error code to be returned to mean "this feature isn't supported on this server".

I think the problem here is couching it in terms of the limits set. What if my server impl considers any configured max delay below 10 to be effectively disabled? Or requires the admin to set some boolean? Or perform a little dance? These seem like valid implementation decisions.

In other words, remove the references to max delay and max events, and just something like:

```
If the feature is disabled on the server via configuration, the homeserver SHOULD instead respond with HTTP 403 and a standard error response with an `errcode` of `M_FORBIDDEN`.
```

That's all that a client or server implementer needs here.

Reviewer 2: That's a great way to phrase it.  Added by ec7cbfd8

Leaving this thread unresolved in case there are any objections to it.

Reviewer 9: Agreed that sounds better!

Reviewer 2: Since there's been no further objections, I'll consider this settled.

#### Delayed event limits as a capability

The values of both the maximum allowed delay and the maximum allowed number of scheduled events are advertised as a
[capability](https://spec.matrix.org/v1.18/client-server-api/#capabilities-negotiation) named `m.delayed_events`, via
the values of fields named `max_delay_ms` and `max_scheduled` respectively.

> If the server doesn't enforce one of these limits, its representative field MUST be absent from the capability.

Reviewer 12: Seems odd that it only specifies this MUST one way around, ie. it must be absent if not enforced but implies that it may be enforced but not present, or is the capability intentionally optional?

Reviewer 2: The intention is for the fields to be present if and only if their respective limit is enforced.

15770e69 tries to make that explicit.

If the server enforces none of these limits, the capability MAY be omitted entirely instead of having an empty body.

For example, the following specifies a maximum allowed delay of 24 hours and a per-user limit of 10 delayed events:

```json
{
  "capabilities": {
    "m.delayed_events": {
      "max_delay_ms": 86400000,
      "max_scheduled": 10
    }
  }
}
```

### Managing scheduled delayed events

A set of new authenticated Client-Server API endpoints at
`POST /_matrix/client/v1/delayed_events/{delay_id}/{action}` allows clients to
manage the sending of scheduled delayed events.

The `action` specifies the management action to take on the scheduled delayed event with the specified `delay_id`.
The supported `action`s are the following:

- `send` - Send the delayed event immediately instead of waiting for its scheduled send time.
- `cancel` - Cancel the delayed event so that it is never sent.

> - `restart` - Reset the delayed event's scheduled send time to be the current time + its original `delay_ms`.

Reviewer 12: This ties in with another thread but why not just have a way to update the delay on an event, then you could update it back to what it originally was (equivalent to restart) or update it to 0 (equivalent to send, except it wouldn't make sense to report sending errors back).

Also, there are idempotency issues here I think, either with updating the event's scheduled time or the plain 'restart': if a request is re-sent, does it restart the timer again or is it ignored for idempotency? This may need some kind of sequence number to differentiate separate requests from retries.

Reviewer 14: > if a request is re-sent, does it restart the timer again 

I think I can answer this from memory: the timer is restarted again. That is correct behaviour for the intended usecase of VoIP call keepalives: "I am still here, please don't auto-send the delayed hangup event yet".

I'm in favour of keeping it simple unless/until we have a different usecase that needs fancier semantics.

Reviewer 5: VoIP team is also in favor of keeping it simple.

- `delayed_events` - An array of objects describing delayed events owned by the requesting user.
  These objects contain the same fields as the object returned by
  [the single-item lookup](#getting-a-single-delayed-event),
  except for the fields exclusive to finalised delayed events (`error`, `event_id`, and `finalised_ts`).

```http
200 OK
Content-Type: application/json

{
  "delayed_events": [
    {
      "delay_id": "...",
      "room_id": "!roomid:example.com",
      "type": "m.room.message",
      "delay_ms": 5500,
      "scheduled_at": 1721732853284,
      "content": {
        "msgtype": "m.text",
        "body": "I am now offline"
      }
    },
    ...
  ]
}
```

#### Retention of finalised delayed events

The amount of finalised events that stay on the homeserver can be chosen by the homeserver.
The recommended strategy is to retain finalised events for up to 7 days or 1000 events per user,
whichever occurs first.

There is no guarantee for a client that events will be available
if they exceed the limits of their homeserver.
Additionally, a homeserver MAY discard finalised delayed events that have been returned by a

> `GET /_matrix/client/v1/delayed_events/{delay_id}` response.

Reviewer 8: This sounds kind of dangerous from a request retrying perspective since network errors can cause responses to be lost. I guess storing finalized events is best-effort anyway and clients shouldn't rely on any of them being stored?

Reviewer 2: > I guess storing finalized events is best-effort anyway and clients shouldn't rely on any of them being stored?

Correct.

Eventually, finalized delayed events should appear in `/sync` responses (and appservice transactions), which will reliably communicate them.

Reviewer: @tulir are you content with this or would you prefer some clarification in the proposal text?

Reviewer 2: Actually, there's no pressing need to discard finalised events on retrieval.  The retention rules should be enough to protect against resource bloat of keeping too many finalised events.  Removed this with e18c6ab4

Conversely, it's also possible to successfully schedule an event that the user has no permission to send at the time of
sending. If the power level situation has changed at the time the delay passes, the event can even reach the DAG.

#### Rate-limiting at the point of sending

Further to the rate limiting of the API endpoints, the homeserver SHOULD apply rate limiting to the sending
of delayed messages at the point that they are inserted into the DAG.

This is to provide mitigation against the [High Volume of Messages](
https://spec.matrix.org/v1.18/appendices/#threat-high-volume-of-messages) threat where a malicious actor
could schedule a large volume of events ahead of time without exceeding a rate limit on the initial `PUT` request,
but has specified a `delay_ms` that corresponds to a common point of time in the future.

If a delayed event fails to be sent at its scheduled send time due to a rate limit failure,
the homeserver SHOULD NOT retry sending the event. Instead, the event will be stored as a finalised delayed event
with its `error` field set, available to be retrieved by a client for the user who requested the event.

> It then becomes the user's responsibility to fetch this error and retry sending the event as appropriate.

Reviewer 12: This seems odd? Why? I think I would expect it to be retried, otherwise this makes the mechanism a chunk less reliable if it might also fail permanently due to transient rate limiting.

Reviewer 2: This has come up before: https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3315572163

I stand by my comment from then:

> I'd rather not offload more work to the server, especially for working around rate limits that are supposed to reduce server load. I can also imagine situations where a retry isn't desired, so it makes more sense for clients to be in control of the decision to retry or not.

Reviewer: This sounds sensible to me. We should probably add that reasoning into the proposal text.

Reviewer 2: 8de254bc

Reviewer: @dbkr / @clokep are you content with the added motivation?

Reviewer 9: Seems ok with me.

Reviewer 12: Fair enough, as long as it's explained (plus, an HS being unable to send an event should be quite an extreme edge case).

### Guest accounts

All delayed event related endpoints are available to guest accounts.
This allows guest accounts to participate in MatrixRTC sessions.

> ## Potential issues

Reviewer 12: A fairly significant potential issue in using this for VoIP hangups is that a client now needs to be able to make regular, timely HTTP requests to not get a call hung up. I frequently make calls in network conditions that are fine for getting UDP voice packets through but don't work for making a small HTTP request.

Reviewer 5: that's the reason why there is delegation of delayed events. In the VoIP example the media server is taking over control.

Reviewer 12: Oh right, of course - I think I was generally failing to understand that this also acts as a way for a client to effectively hand the SFU a signed and sealed envelope with a hangup event that it can send (or trigger to be sent) later. Could add a thing to clarify this maybe.

Reviewer 5: However, as it was decided to move the LiveKit Auth service into the homeserver, the initial idea of "delegation of delayed events" was cut down to be  homeserver internal (e.g., app services). It's subject to a new MSC once there is demand to delegate to an external service.

Reviewer: > Oh right, of course - I think I was generally failing to understand that this also acts as a way for a client to effectively hand the SFU a signed and sealed envelope with a hangup event that it can send (or trigger to be sent) later. Could add a thing to clarify this maybe.

We already seem to have this covered right below in the "Delegating scheduled delayed events" section. So I think this should be clear enough after reading further. Happy to take suggestions for improving the text of course.

> A fairly significant potential issue in using this for VoIP hangups is that a client now needs to be able to make regular, timely HTTP requests to not get a call hung up. I frequently make calls in network conditions that are fine for getting UDP voice packets through but don't work for making a small HTTP request.

I've captured this with a tiny change in the delegation alternative section (83d9e10f9aa6e7306843ea3260330fa938d51a8c).

proposal shouldn't affect how that verification works because those other proposals will
need to account for eventual consistency anyway, which may appear as a delayed event or
attached to a disjointed part of the DAG.

### Conflicting delayed state events

A delayed state event can overwrite other state events that were sent in between the delayed event
being scheduled and it being sent. Whether or not this is problematic strongly depends on the
use case, though. When the overwrite is undesired, a possible remedy could be to cancel the scheduled
delayed event when a conflicting new state event is sent into the room. Alternatively, it might be
possible to avoid the conflict in the first place by using separate `state_key`s or by not relying on
state events to begin with. Additionally, this type of race condition can also happen without delayed
events due to federation delay. Potentially addressing this situation is, therefore, left to a future
proposal.

### Inability to filter and paginate delayed events

`GET /_matrix/client/v1/delayed_events` lacks request parameters for filtering and pagination. It also

> doesn't allow querying finalised delayed events. This could be limiting in some cases. A future proposal

Reviewer 12: > It also doesn't allow querying finalised delayed events

Could you clarify this sentence - I'm not sure exactly what it's saying (unless it's fetching only finalised events, but isn't that the same as filtering?)

Reviewer: Yes, you're right. I believe this was just trying to call out getting finalized events as a particular example. But we can just collapse it into general filtering. Have made this small change in e5bbe4167851a1820569f95fe3e540ab61db3b9e.

such as [MSC4486] may extend the endpoint to support those use cases.

[MSC4486]: https://github.com/matrix-org/matrix-spec-proposals/pull/4486

## Alternatives

> 

Reviewer 8: The opening paragraph has multiple use cases, but there are also various limits that could interfere with clients actually using this for all those use cases. For example, if a server sets a low pending delayed event limit, then a client uses delayed events for self-destructing messages, would calls just fail because the user's limit is used up?

The 429 behavior is also mildly scary if clients aren't smart enough to ignore too long Retry-After fields (like if all the delayed events are for scheduled or self-destructing messages in 3 months, starting a call might just get stuck in a `sleep()`). That's mostly a client implementation concern though, the MSC is probably fine as is.

I'm not sure if it should be fixed in this MSC or just documented as a possible alternative/future MSC. The fix would probably be having the server tell the client what use case it's targeting so the server can have per-use-case limits

Reviewer 2: > For example, if a server sets a low pending delayed event limit, then a client uses delayed events for self-destructing messages, would calls just fail because the user's limit is used up?
>
> The 429 behavior is also mildly scary if clients aren't smart enough to ignore too long Retry-After fields (like if all the delayed events are for scheduled or self-destructing messages in 3 months, starting a call might just get stuck in a sleep()).

I suppose in cases like that, it's the client's responsibility to be proactive in checking server capabilities to verify that a desired usecase is feasible.  Then all that'd be required from this MSC is to ensure that enough capabilities are exposed for such checking, which IMO they already are.

Reviewer: Agreed. It feels like a client responsibility to verify that the server's published limits don't interfere with the features the client builds on top of them.

@tulir maybe it's sufficient if we just document this as a potential issue in implementations?

Reviewer 2: I've added something to the Potential Issues section to cover this.  @tulir Let me know if that's sufficient.  5398aa97

Reviewer 8: Yeah looks reasonable

Reviewer 2: Thanks. Marking this as resolved, then

that only allows to interact with the [delayed event management endpoints](#managing-scheduled-delayed-events)
on that specific `delay_id`.
This would require those endpoints to be unauthenticated, but for the `delay_id` to serve as their access token instead.

With this, an SFU that tracks the current client connection state could be given the power to control the delayed event.
The client would share the `delay_id` and other possibly required details (such as its homeserver's server name),
so that the SFU can call the `POST /delayed_events/{delay_id}/restart` endpoint while a user is connected
and can call the `POST /delayed_events/{delay_id}/send` endpoint once the user disconnects.
This way, the SFU can be used as the source of truth for the call membership events without knowing anything about
the Matrix call.

Since the SFU has a much lower chance of running into a network issue,
`POST /delayed_events/{delay_id}/restart` calls may be sent much more infrequently.
Instead of calling that endpoint every couple of seconds, a delayed event's
timeout can be set to be long (e.g. 6 hours), as the SFU can be expected to not forget sending the

> `POST /delayed_events/{delay_id}/send` requests when it detects a disconnecting client.

Reviewer 12: Honestly I suspect if we were to delegate this to the SFU, the sensible approach would to just let the SFU send the hangup sent in the normal way which would be much simpler: it ought to be reliable enough.

Reviewer 5: > Honestly I suspect if we were to delegate this to the SFU, the sensible approach would to just let the SFU send the hangup sent in the normal way which would be much simpler: it ought to be reliable enough.

wdym with in the normal way?

We are using encrypted sticky events for MatrixRTC membership. So a delayed event is the solution to sending the pre encrypted hangup event. The reason for the /reset is for a potentially long going MatrixRTC session a reset is required as the homeserver admin can restrict the max sticky event duration.

Reviewer 12: Oh I see yes, the SFU can't just send the hangup event since it needs to be encrypted.

Instead, the homeserver allocates a `delay_id` to the scheduled event which is used during the above API operations.

### Scheduling a delayed event

A new authenticated Client-Server API endpoint at
`PUT /_matrix/client/v3/rooms/{roomId}/delayed_event/{eventType}/{txnId}` allows clients to
schedule the sending of a message or state event.

The path parameters of this endpoint are the same as for the existing
[`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](
https://spec.matrix.org/v1.18/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
endpoint. It also makes use of [transaction identifiers](
https://spec.matrix.org/v1.18/client-server-api/#transaction-identifiers),
and supports [timestamp massaging](https://spec.matrix.org/v1.18/application-service-api/#timestamp-massaging)
when called by an application service.

The body for requests to this endpoint is a JSON object containing the following fields:

- `delay_ms` - Required. A positive non-zero number of milliseconds the homeserver should wait before sending the event.
- `state_key` - The state key for the event to be sent, if it is to be a state event; absent otherwise.

> - `content` - Required. The content of the event to be sent.

Reviewer 9: Having a very different body than `/send` but the same path is very confusing.

It seems like it would be better to mirror the existing `/send` and `/state` endpoints (or not even bother with `/state` if it isn't needed yet).

Reviewer 8: The MSC isn't using the same path? That was an older revision, but it was replaced with a dedicated /delayed_events path

Reviewer 9: > The MSC isn't using the same path?

Yes, but it purposefully mirrors the path, I think it'll be quite confusing to have completely different API shapes.

Reviewer 14: It's fine IMHO.

I guess if you wanted to mirror the existing endpoints, you'd have `PUT /_matrix/client/v3/rooms/{roomId}/delayed_send/{eventType}/{txnId}` and `PUT /_matrix/client/v3/rooms/{roomId}/delayed_state/{eventType}/{stateKey}`. I woudn't be dead against that, but I'm unconvinced it's worth changing.

Reviewer 2: Another point against having split delayed send/state endpoints is that the semantics of sending a delayed event are the same for both message & state events, unlike for non-delayed events where `/send` uses a transaction ID to distinguish requests while `/state` uses the state key.

Reviewer: @clokep are you alright with keeping the endpoint as is or would you like to keep this thread open for further discussion?

Reviewer 9: I'm not sure. I'm a bit worried that supporting state at all will be a huge footgun down the road.

Reviewer 2: What about delayed state events in particular is problematic?

Some usecases I can think of are:
- kicking a set users in a room after a set time, to implement an "expiring room"
- changing the name/topic of a room after a set time, to indicate the room's age
- sending an event to prevent making voice/video calls after a certain point, to prevent open calls for too long
- scheduling a room upgrade to prevent a room from staying on an old version for too long (this is already possible to do with some finesse)

Also note that if the eventual sending of a state event fails, the account that scheduled it can learn about what went wrong by looking up the associated finalised delayed event.

Reviewer 9: I guess those all sound reasonable, I didn't think we had any use cases for it though and I dislike adding features just because.

Reviewer: Maybe we could list one or two of the examples above in the introductory section of the proposal to motivate why delayed state events are included.

Reviewer 2: Yes, may as well.  Done by 4d395ec9

(Not re-opening this thread because the contention re: delayed state is still resolved)

The homeserver schedules the event to be sent with the specified delay and responds with an
opaque `delay_id` field (omitting the `event_id` as it is not available):

```http
200 OK
Content-Type: application/json

{
  "delay_id": "1234567890"
}
```

The `delay_id` is an [opaque identifier](https://spec.matrix.org/v1.18/appendices/#opaque-identifiers)
generated by the homeserver.

The homeserver MUST NOT send the event before the scheduled time.

> To support batch sending, homeservers MAY add up to 30 seconds to the scheduled send time.

Reviewer 9: Why does this need to batch sending? It seems very at odds with the feature.

Reviewer 8: If the user (or different users) schedule multiple events at timestamps close to each other, the homeserver may want to batch the sends together for efficiency reasons

Reviewer 9: OK, that seems like an implementation detail though. I think what this is trying to say is "don't expect your event to get send at _exactly_ the millisecond it is scheduled to, but I think that's kind of obvious?

> The homeserver MUST NOT send the event before the scheduled time, but
> MAY send it shortly after the scheduled send time.

Reviewer: This suggestion seems reasonable to me. We don't need to prescribe any further details for how such efficiency delays are implemented. @AndrewFerr WDYT?

Reviewer 2: Sure; the fact that the allowed lag can be used to implement batch sending is very much an implementation detail.

Also, the "batch sending" here may also be confused with the kind of batch sending discussed in one of the Alternatives sections, i.e. sending related events together.

We should at least define an upper bound on the allowed lag here though, lest "shortly after" is too ambiguous.

All done by 062e6491.

Note: clients might find that their events are delayed further due to server load and similar conditions.

The homeserver MUST persist scheduled delayed events such that they will not be lost by the homeserver being restarted.
Moreover, when a homeserver restarts, it MUST scan for all scheduled delayed events whose send time has already passed
(i.e. delayed events that were scheduled to be sent while the homeserver was offline),
and send those delayed events as soon as possible, in chronological order of their scheduled send times.

The homeserver MAY enforce a maximum allowed delay for delayed events. This limit is
communicated to the client in a capability (described later in this proposal).
If a requested delay exceeds this maximum, the homeserver will respond with HTTP 403
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)

> with an `errcode` of `M_FORBIDDEN`.

Reviewer 9: This seems like both the wrong status code and wrong error code. I'd expect 400, unsure about the error code.

Reviewer 11: Agreed.  I don't think any of the existing error codes quite match, so maybe a new one is needed for this.

Reviewer 5: addressed by https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/0c7193aaa07c9a2634dccd7cf3bd83e2733a806e

Reviewer 2: As long as we're adding a new error code for this, is there any interest in including the value of the max allowed delay in the error response body?

That is what used be done by `M_MAX_DELAY_EXCEEDED`, which was removed in favour of reusing existing error codes & expressing the limit in a capability instead: https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3257895622

Admittedly, there has been enough discomfort around contorting existing error codes to fit this endpoint that it makes sense to reinstate a dedicated error code.

This isn't to suggest removing the max delay capability, but to save clients the effort of having to look up the capability to know the allowed max when they hit this error response.

Reviewer: > As long as we're adding a new error code for this, is there any interest in including the value of the max allowed delay in the error response body?

I'm not opposed but given that we have (and probably want to keep) the capability anyway maybe that's sufficient and we don't need to bother with the extra complexity?

Reviewer 2: Alright, let's leave out the field then.  Besides, there's nothing stopping it from being added in a later MSC if it's really desired.

The homeserver SHOULD apply rate limiting to the scheduling of delayed events to provide mitigation against the
[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threat.

The homeserver SHOULD enforce a limit of how many delayed events a user may have scheduled at once
to provide mitigation against both the
[High Volume of Messages](https://spec.matrix.org/v1.18/appendices/#threat-high-volume-of-messages) and
[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threats.
If a user's request to schedule a delayed event would exceed this limit, the homeserver will respond with HTTP 429,
a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
with an `errcode` of `M_LIMIT_EXCEEDED`, and a `Retry-After` header whose value is set to the time of/until
the scheduled send time of the next of the user's delayed events to be sent,

> rounded up to the nearest second.

Reviewer 9: `M_USER_LIMIT_EXCEEDED` seems to the the standard error code?

This should be a 400.

Reviewer 14: I think `M_USER_LIMIT_EXCEEDED` was introduced to deal with premium accounts; the general implication is that it's a quota you can pay to increase. Accordingly, `M_LIMIT_EXCEEDED` is probably appropriate here. And normally we would use a 429.

Reviewer 2: `M_USER_LIMIT_EXCEEDED` was considered for this, but that is meant for quota limits that can be bypassed (such as with a "premium" subscription / payment-gated feature), whereas this is a server-imposed limit for the sake of appropriate resource allocation (akin to something like a maximum file upload size limit).

See: https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r2878970224

Reviewer 2: The discussion around using `M_LIMIT_EXCEEDED` / HTTP 429 here has settled in https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3658009434 , so resolving this thread.

(Also, I hadn't seen Rich's comment on `M_USER_LIMIT_EXCEEDED` until after I had posted mine, hence the duplication 😅)

The supported `action`s are the following:

- `send` - Send the delayed event immediately instead of waiting for its scheduled send time.
- `cancel` - Cancel the delayed event so that it is never sent.
- `restart` - Reset the delayed event's scheduled send time to be the current time + its original `delay_ms`.

For example, the following would send the delayed event with `delay_id` `1234567890` immediately:

```http
POST /_matrix/client/v1/delayed_events/1234567890/send
Content-Type: application/json

{
}
```

Where the `action` is `send`, the homeserver SHOULD apply rate limiting to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.18/appendices/#threat-high-volume-of-messages) threat.

For all `action`s, the homeserver SHOULD apply rate limiting to provide mitigation against the
[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threat.

If no delayed event with the specified `delay_id` can be found,
the homeserver will respond with HTTP 404
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)

> with an `errcode` of `M_NOT_FOUND`.

Reviewer 9: It is implied by other areas, but not explicit here that these are user scoped.

Reviewer 5: addressed by https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/2d42dafcf8d6f50c965f84ba35ac4107b6e2267b

On success, the homeserver will respond with HTTP 200
and a response body of an empty object. A future MSC may define additional keys, such as returning
the event ID for an `action` of `send` or the new expected send time for an `action` of `restart`.

To allow safely retrying requests, the homeserver will respond with success
if the target delayed event is already finalised with an outcome that matches the `action`, i.e.
if the `action` is `send` and the delayed event has already been sent,
or if the `action` is `cancel` and the delayed event has already been cancelled
(either due to user action or an error).

If the target delayed event is already finalised with an outcome that conflicts with the `action`, i.e.
if the `action` is `send` or `restart` and the delayed event has already been cancelled,
or if the `action` is `cancel` and the delayed event has already been sent,
the homeserver will respond with HTTP 409
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)

> with an `errcode` of `M_UNKNOWN`.

Reviewer 14: I'm not a big fan of us deliberately specifying M_UNNOWN here. Maybe we can define an M_CONFLICT or something.

Reviewer 2: There is precedent in the spec for using `409` without a more specific errcode: [PUT /_matrix/client/v3/directory/room/{roomAlias}](https://spec.matrix.org/latest/client-server-api/#put_matrixclientv3directoryroomroomalias) uses it when trying to create a room alias with a name that already exists.

What else the spec does sometimes is not enforce any particular `errcode` in the error response for a given HTTP code (though example response bodies always include one, typically `M_UNKNOWN`).

So to keep things simple, how about we simply leave the errcode unsaid here?

Reviewer 14: [I'd still like to see this changed, ftr]

Reviewer 2: e2636d62 just leaves the errcode unspecified.

IMO there is no value in defining a more specific errcode here since it wouldn't provide any extra information.  Only if there were multiple possible reasons for a 409 would it help.

But if you'd still rather have a new errcode, I will oblige.

Reviewer 14: Oops, I missed your previous comment in this thread, sorry.

The spec is abitshit about defining error codes in general, so I wouldn't put too much store by what is already there. The 409 in /directory dates from prehistory: https://github.com/matrix-org/matrix-doc/pull/178.

I'd kinda prefer for there to be a defined Matrix error code, but I don't feel that strongly about it, and as you say the client gets enough info purely from the HTTP error code. I'm happy as it is.

Reviewer 2: Thanks for the tolerance.  I will leave this as-is then.

If the action is `send` and the delayed event is unable to be sent due to an error,
the homeserver will respond with that error (e.g. HTTP 403
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)

> with an `errcode` of `M_FORBIDDEN` if the user doesn't have permission to send the event at the time of sending,

Reviewer 9: This doesn't match what the `/send` endpoint would do.

Reviewer 14: Doesn't it?

Reviewer 2: Really?  That's what Synapse does, at least.

The spec's only error response for `/send` is 400 with a standard error response.  This section could be changed to mimic that.

Reviewer: +1 for changing to 400 then for consistency.

Reviewer 2: 57a24037 changes this error to HTTP 400, and uses `M_DUPLICATE_ANNOTATION` as the example error code to match the example given in [`/send`](https://spec.matrix.org/latest/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)'s 400 response.

or HTTP 429 if the user has exceeded rate limits for sending room events at that time),
as if the request had been to send the event as a non-delayed event with either the `/send` or `/state` endpoint.
The homeserver SHOULD keep the delayed event scheduled, to account for the fact that the cause of the error
may resolve by the time of the delayed event's scheduled send time,
and to allow retries of the `send` action until then.

### Getting delayed events

A set of new authenticated Client-Server API endpoints allows clients to look up

> both scheduled and finalised delayed events owned by the requesting user.

Reviewer 9: No rationale is provided for why you need finalized events from these APIs.

Reviewer 13: I can't recall immediately what the rationale is. (@AndrewFerr, your input please?) If there is none at this time, would you prefer that this behavior is removed from the MSC?

Reviewer 9: I think it would heavily simplify the API and remove some ambiguous pieces (ie how long finalized operations are kept around).

I don't feel super strongly if there's a use case for it though.

Reviewer 2: The main purpose of finalized events is to make it possible to discover delayed events that failed to be sent due to an error.  Otherwise, failed delayed events would fail silently (at least until they're included in the `/sync` response).

Reviewer: This paragraph has been eliminated since but maybe we could add a small explanation of why we retain finalized events under the "#### Retention of finalised delayed events" section?

Reviewer 2: Sure.  Added by 5d3a1ee2

[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threat.
They most likely require (dependent on the implementation) serialization steps
and can be used to slow down the homeserver.

#### Getting a single delayed event

A new authenticated Client-Server API endpoint at
`GET /_matrix/client/v1/delayed_events/{delay_id}` responds with
details on the delayed event with the specified `delay_id`.

If no delayed event with the specified `delay_id` exists,
or it does exist but had been scheduled by a user other than the one requesting this endpoint,
the homeserver will respond with HTTP 404
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
with an `errcode` of `M_NOT_FOUND`.

On success, the homeserver will respond with HTTP 200 and a JSON object containing the following fields:

- `delay_id` - Required. The ID of the delayed event.
- `room_id` - Required. The ID of the room that the delayed event was scheduled to be sent in.
- `type` - Required. The event type of the delayed event.
- `state_key` - The state key of the delayed event if it is a state event; absent otherwise.

> - `delay_ms` - Required. The delay in milliseconds after the point of scheduling that the event is/was to be sent at.

Reviewer 9: Is this useful or do clients actually want to know how many milliseconds until the event is sent? Time drift between servers and clients is a real thing.

Reviewer 5: This enables other devices in your account to gather all required data to tell the user when something will be sent.

Reviewer: Yeah, this feels like quite essential information to me if we want to enable management of delayed events. It's true that time drift will make this somewhat fuzzy. But then again, delayed events are not guaranteed to be sent at the exact instant specified anyway. So I think this is fine?

Reviewer 2: Resolving this on account of Patrick's 👍.

- `scheduled_at` - Required. The timestamp (as Unix time in milliseconds) when the delayed event was scheduled or

>   last restarted.

Reviewer 9: `scheduled_ts` would fit existing fields better.

Reviewer 2: IMO the `_ts` suffix could make it seem like this field is for the scheduled send time instead of when the delayed event had been scheduled/restarted.  This is called out here: https://github.com/matrix-org/matrix-spec-proposals/blob/15770e6988f307499519b0cb294b95ed494171b6/proposals/4140-delayed-events-futures.md?plain=1#L564

Reviewer 9: That's exactly how I read 'scheduled_at` to mean! Maybe `created_ts`?

Reviewer 5: I think created_ts would not be precise in case of a reset of a delayed event. So if we want to have the ts it should be scheduled_at_ts

@AndrewFerr wdyt?

Reviewer 2: `_at_ts` feels redundant.  For precedent, the spec uses either `_at` or `_ts` for timestamp values, but not both together.

It also looks like the choice of `_at` or `_ts` hasn't helped with the clarity of this field name.

How about `scheduled_since_ts`?
- `scheduled_since` makes it clear that the time is in the past, as opposed to the scheduled send time
- `_ts` makes it clear that it's a timestamp

Reviewer 2: Last suggestion: `delayed_since_ts`, done by b9efebe2.  Reasoning is given in the reworked Alternatives section (which also removes names we've long since stopped considering).

Reviewer: This seems fair to me. "since" makes it very clear that it refers to the start rather than the end of the interval. I'll consider this resolved.

  The [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
  of the error that prevented the delayed event from being sent.
- `event_id` - The `event_id` this event got in case it was sent.
- `finalised_ts` - The timestamp (as Unix time in milliseconds) when the event was finalised;
  absent if it is still scheduled.
  Using [timestamp massaging](https://spec.matrix.org/v1.18/application-service-api/#timestamp-massaging)
  does not affect the value of this field.

Whether a delayed event is still scheduled, or has been sent, failed to be sent due to an error, or was cancelled
can be determined by examining which of the optional fields are present in the response object:

- If `finalised_ts` is absent, then the delayed event is still scheduled.
- Otherwise, if `event_id` is present, then the delayed event has been sent.
  - If `finalised_ts` < `scheduled_at` + `delay_ms`, then the event was sent manually by
    [the `/send` endpoint](#managing-scheduled-delayed-events); otherwise, it was sent on its scheduled send time.
- Otherwise, if `error` is present, then the delayed event failed to be sent (and was descheduled) due to an error.

> - Otherwise, the delayed event was cancelled by [the `/cancel` endpoint](#managing-scheduled-delayed-events).

Reviewer 9: This seems like a really hard response to fit into an OpenAPI definition. There's a lot of optional fields, etc.

Reviewer 14: Optional fields are easy to represent in an OpenAPI definition.

We could bikeshed about more structured ways of representing this data (a `state` enum`, maybe?) but I'm not that bothered.

Reviewer 9: It isn't optional fields that are the issue, but sets of fields that are all optional or all must be included.

Reviewer 13: Would you prefer that there were a single optional `finalised` object containing both the `event_id` and the `finalised_ts`?

Reviewer 2: If a `finalised` field were to be introduced, what would end up changing is:
- `finalised` would be optional
- `finalised.finalised_ts` would be required
- `finalised.error` & `finalised.event_id` would be optional

That would guard against `error` and `event_id` from ever being set for non-finalised events.

Reviewer 2: 0ffd6731 adds a `finalised` field as described above.  It also removes the normative section on what the different combinations of fields mean now that there are fewer such combinations.

```http
200 OK
Content-Type: application/json

{
  "delayed_events": [
    {
      "delay_id": "...",
      "room_id": "!roomid:example.com",
      "type": "m.room.message",
      "delay_ms": 5500,
      "scheduled_at": 1721732853284,
      "content": {
        "msgtype": "m.text",
        "body": "I am now offline"
      }
    },
    ...
  ]
}
```

#### Retention of finalised delayed events

The amount of finalised events that stay on the homeserver can be chosen by the homeserver.
The recommended strategy is to retain finalised events for up to 7 days or 1000 events per user,
whichever occurs first.

There is no guarantee for a client that events will be available
if they exceed the limits of their homeserver.
Additionally, a homeserver MAY discard finalised delayed events that have been returned by a
`GET /_matrix/client/v1/delayed_events/{delay_id}` response.

### Additional homeserver behaviour

#### `delay_id` in `unsigned` event data
The `delay_id` of a sent delayed event MUST be included in the resulting room event's `unsigned` data
if, and only if, the client being given the event (e.g. over `/sync` or `/messages`) is authenticated as the event's sender.

> #### Power levels are evaluated at the point of sending

Reviewer 9: What about all the other ways events can fail validation?

Reviewer 2: Yes, this part should call out how malformed events are handled.

Synapse validates the event at the time of scheduling, as opposed to at the scheduled send time, because unlike power levels, event validation rules are unlikely to change over time.  So IMO this section should at enforce at least a schedule-time validation check.

Reviewer 2: Done by ae669936

Power levels are evaluated for each event only once the delay has occurred and it will be distributed/inserted into the
DAG. This implies a delayed event can fail if it violates power levels at the time the delay passes.

Conversely, it's also possible to successfully schedule an event that the user has no permission to send at the time of

> sending. If the power level situation has changed at the time the delay passes, the event can even reach the DAG.

Reviewer 9: Why? What's the use case for this?

Reviewer 13: It's just simpler to evaluate the power levels once at sending time. The alternative, automatically rejecting/canceling delayed events whenever the current room state looks unfavorable for them eventually getting sent, seems more complex.

Reviewer: Considering this resolved given the thumbs up reaction.

#### Rate-limiting at the point of sending

Further to the rate limiting of the API endpoints, the homeserver SHOULD apply rate limiting to the sending

> of delayed messages at the point that they are inserted into the DAG.

Reviewer 9: "further to" is confusing:

```suggestion
In addition to the rate limiting of the API endpoints, the homeserver SHOULD apply rate limiting to the sending
of delayed messages at the point that they are inserted into the DAG.
```

but has specified a `delay_ms` that corresponds to a common point of time in the future.

If a delayed event fails to be sent at its scheduled send time due to a rate limit failure,
the homeserver SHOULD NOT retry sending the event. Instead, the event will be stored as a finalised delayed event
with its `error` field set, available to be retrieved by a client for the user who requested the event.
It then becomes the user's responsibility to fetch this error and retry sending the event as appropriate.

### Guest accounts

All delayed event related endpoints are available to guest accounts.
This allows guest accounts to participate in MatrixRTC sessions.

## Potential issues

### Compatibility with Cryptographic Identities

Proposals like [MSC4080](https://github.com/matrix-org/matrix-spec-proposals/pull/4080)
enable a client to verify that an event was *actually* sent by the sender's devices. This
proposal shouldn't affect how that verification works because those other proposals will
need to account for eventual consistency anyway, which may appear as a delayed event or
attached to a disjointed part of the DAG.

> ### Conflicting delayed state events

Reviewer 9: Do we even need this feature for state events today? This seems like a big foot gun without a use case.

Reviewer 2: There is indeed no longer a planned usecase for delayed state events, or at least one that relies on any particular behaviour of them.  f0ad1898 removes this section.

A delayed state event can overwrite other state events that were sent in between the delayed event
being scheduled and it being sent. Whether or not this is problematic strongly depends on the
use case, though. When the overwrite is undesired, a possible remedy could be to cancel the scheduled
delayed event when a conflicting new state event is sent into the room. Alternatively, it might be
possible to avoid the conflict in the first place by using separate `state_key`s or by not relying on
state events to begin with. Additionally, this type of race condition can also happen without delayed
events due to federation delay. Potentially addressing this situation is, therefore, left to a future
proposal.

> ### Inability to filter and paginate delayed events

Reviewer 9: Leaving pagination to a future proposal is not acceptable IMO.

@mscbot concern List endpoint is unbounded and needs pagination.

Reviewer 14: The easiest solution to this may be to get rid of the list endpoint, if we don't have a usecase for it?

Reviewer 14: In any case, I'm not sure I agree. This MSC is big enough as it is, and pagination was broken out to MSC4486 deliberately. Let's just land this MSC without adding more clutter.

Reviewer 9: Agreed this is lengthy.

Moving the entire list endpoint to a separate MSC that provides all the needed filtering/pagination probably makes the most sense.

But also if there's no use-case, why include it?

Reviewer 2: The use-case is to give clients the ability to check if any delayed events failed to be sent.

Though admittedly, if a client really cares about the status of any particular delayed event, it would very likely have its `delay_id` on hand & use the single-item lookup to find it.

Still, it feels incomplete to not be able to simply list all available delayed events, especially now that it's contained in its own MSC.

Reviewer 2: To avoid leaving this endpoint partially defined, I've moved it out of this MSC entirely & put it into MSC4486.  No use case currently relies on bulk lookup, so there is no reason to rush it in now.

b1eb9b96 for the changes in this MSC.

- `org.matrix.msc4140.delayed_events` should be used instead of the `m.delayed_events` capability name.

Additionally, the feature is to be advertised as an unstable feature in the `GET /_matrix/client/versions` response,
with the key `org.matrix.msc4140` set to `true`. So, the response could then look as follows:

```json
{
    "versions": ["..."],
    "unstable_features": {
        "org.matrix.msc4140": true
    }
}
```

Once the MSC is accepted, but before the homeserver advertises the spec version that includes the MSC, the homeserver
should advertise `org.matrix.msc4140.stable` as an unstable feature flag to let clients know that they can use the
stable endpoints for sending and managing delayed events.

## Dependencies

None.

[^eventId]: An event's ID is computed from its [reference hash](https://spec.matrix.org/v1.18/rooms/v11/#event-ids)
which is obtained by combining several event properties including `origin_server_ts`. The latter is
only available once the event has actually been sent, however. Since scheduled delayed events may be
cancelled or re-scheduled, the `origin_server_ts` and, thus, the event ID cannot be determined ahead

> of time.

Reviewer 9: Is this the only field stopping event ID calculation?

Reviewer 8: prev_events and auth_events are also not known ahead of time and we don't want to mess up the DAG

Reviewer 9: Right, those are much stronger rationale IMO. I could see it being reasonable to use the original schedule time as `origin_server_ts` or something like that if it was the only issue.

Reviewer 2: Resolving as the concern appears to be quelled.

Reviewer 2: Oh right, those other fields should be mentioned here.  Done by 1f7822f0

- If any homeservers missed the federated "heartbeat"/restart message, then they might decide that the event is visible
  to clients whereas other homeservers might have received it and come to a different conclusion.
  If the event was later cancelled, then resolving the inconsistency feels more complex than if the event was never sent
  in the first place.

[MSC3277: Scheduled messages](https://github.com/matrix-org/matrix-spec-proposals/pull/3277) proposes a similar feature
and there is an extensive analysis of the pros and cons of this MSC vs MSC3277
[in this discussion](https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r1653083566).

If modifying scheduled delayed events is not required, there is a benefit in federating them. It increases
resilience because the sender's homeserver can disconnect and the delayed event will still be delivered to receiving clients by their own homeservers.

However, for the MatrixRTC use case it's required to be able to modify the event after it has been scheduled. As such,
this approach has been discounted.

The considerations above apply irrespective of whether delayed events are federated directly

> or though other means such as by (ab)using typing notification EDUs.

Reviewer: Small typo.

```suggestion
or through other means such as by (ab)using typing notification EDUs.
```

However, it is limited to application services and focuses on historic
data. Since the additional capability to use a template `event_id` parameter is also needed,
this probably is not a good fit.

### Reusing the `send`/`state` endpoints

Instead of creating a new endpoint for scheduling a delayed event,
the `/send` and `/state` endpoints could support sending events with a delay,
via an optional query parameter for specifying the desired delay:

`PUT /_matrix/client/v1/rooms/{roomId}/send/{eventType}/{txnId}?delay={delay_ms}`

`PUT /_matrix/client/v1/rooms/{roomId}/state/{eventType}/{stateKey}?delay={delay_ms}`

This would require the endpoints to give a different response body when sending a delayed event,
as the usual response of the event ID is unavailable at the time of scheduling a delayed event.

The main benefit of reusing the endpoints as described is the potential for code reuse in homeserver implementations.
However, this could be negated by the complexity of having to support multiple response body formats.

This approach also requires using a query parameter for a purpose other than filtering,

> which defies the usual semantics of a URI query.

Reviewer: Addressing comment from https://github.com/matrix-org/matrix-spec-proposals/pull/4140/changes#r3483208622.

```suggestion
```

[MQTT]: https://mqtt.org/

The advantage of this approach is that it doesn't require a new system for emitting heartbeats. An API
for letting clients store the leave event to be sent would still be required due to encryption, however.

On the downside, the common timeouts for syncs are much larger than what would be desirable for
call membership (30s vs. 5s). Overall this alternative feels overly implicit and less flexible than the
solution proposed above.

### Federated delayed events

Delayed events could be sent over federation immediately to then have the receiving homeservers send them down to clients at the appropriate time.

Downsides of this approach that have been considered are that:

- restarts ("heartbeats") would need to be distributed via federation, meaning more traffic and processing
- if any homeservers missed the federated "heartbeat"/restart message, then they might decide that the event is visible
  to clients whereas other homeservers might have received it and come to a different conclusion.
  If the event was later cancelled, then resolving the inconsistency feels more complex than if the event was never sent

>   in the first place.

Reviewer: Nit: Let's make these proper sentences.

```suggestion
- Restarts ("heartbeats") would need to be distributed via federation, meaning more traffic and processing.
- If any homeservers missed the federated "heartbeat"/restart message, then they might decide that the event is visible
  to clients whereas other homeservers might have received it and come to a different conclusion.
  If the event was later cancelled, then resolving the inconsistency feels more complex than if the event was never sent
  in the first place.
```

[MSC3277: Scheduled messages](https://github.com/matrix-org/matrix-spec-proposals/pull/3277) proposes a similar feature
and there is an extensive analysis of the pros and cons of this MSC vs MSC3277
[in this discussion](https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r1653083566).

If it's not needed to allow modification of a delayed event after it has been scheduled, there is a benefit in
federating the scheduled event (adding it to the DAG immediately). It increases resilience: the sender's homeserver can

> disconnect and the delayed message still will enter non-soft-failed state (will be sent).

Reviewer: I think we can shorten this slightly.

```suggestion
If modifying scheduled delayed events is not required, there is a benefit in federating them. It increases
resilience because the sender's homeserver can disconnect and the delayed event will still be delivered to receiving clients by their own homeservers.
```

It used the client's sync loop as an indicator to determine if the call membership event is expired.
Conceptually, this is similar to [MQTT]'s "Will Message" that is published by the server when a client
disconnects.

[MSC4018: Reliable call membership]: https://github.com/matrix-org/matrix-spec-proposals/pull/4018
[MQTT]: https://mqtt.org/

The advantage of this approach is that it doesn't require a new system for emitting heartbeats. An API
for letting clients store the leave event to be sent would still be required due to encryption, however.

On the downside, the common timeouts for syncs are much larger than what would be desirable for
call membership (30s vs. 5s). Overall this alternative feels overly implicit and less flexible than the
solution proposed above.

### Federated delayed events

Delayed events could be sent over federation immediately to then have the receiving homeservers send them down to clients at the appropriate time.

Downsides of this approach that have been considered are that:

- individual "heartbeats"/restarts would need to distributed via federation, meaning more traffic and processing

>   to be done.

Reviewer: I think there's a "be" missing here.

```suggestion
- restarts ("heartbeats") would need to be distributed via federation, meaning more traffic and processing
```

https://spec.matrix.org/v1.18/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
endpoint. It also makes use of [transaction identifiers](
https://spec.matrix.org/v1.18/client-server-api/#transaction-identifiers),
and supports [timestamp massaging](https://spec.matrix.org/v1.18/application-service-api/#timestamp-massaging)
when called by an application service.

The body for requests to this endpoint is a JSON object containing the following fields:

- `delay_ms` - Required. A positive non-zero number of milliseconds the homeserver should wait before sending the event.
- `state_key` - The state key for the event to be sent, if it is to be a state event; absent otherwise.
- `content` - Required. The content of the event to be sent.

The homeserver schedules the event to be sent with the specified delay and responds with an
opaque `delay_id` field (omitting the `event_id` as it is not available):

```http
200 OK
Content-Type: application/json

{
  "delay_id": "1234567890"
}
```

The `delay_id` is an [opaque identifier](https://spec.matrix.org/v1.18/appendices/#opaque-identifiers)
generated by the homeserver.

> It MUST be globally unique and SHOULD be cryptographically secure (in the sense that it is infeasible to predict).

Reviewer 15: There's a mismatch here with the wording in the `Security considerations – Authentication` section below, which says
> As such, generated `delay_id`s MUST be cryptographically random such that they are difficult to guess.

1. "cryptographically _random_" vs. "cryptographically _secure_" – I think a more accurate and unambiguous wording in both places would be something like
    > generated using a CSPRNG (Cryptographically Secure Pseudorandom Number Generator) and has sufficient entropy

2. MUST vs. SHOULD – I would change the SHOULD here to MUST, because that's what the authentication for these endpoints through knowledge of `delay_id` relies on.

Reviewer 15: A bit later, "the *cryptographic security* of the `delay_id`" is mentioned again, which reads a bit weird to me. Alternatives that would sound better to me are for example
- the CSPRNG-generated `delay_id`, or
- the cryptographic randomness of the `delay_id`, or
- the unpredictability of the `delay_id`.

Reviewer: > A bit later, "the _cryptographic security_ of the `delay_id`" is mentioned again, which reads a bit weird to me. Alternatives that would sound better to me are for example
> 
>     * the CSPRNG-generated `delay_id`, or
> 
>     * the cryptographic randomness of the `delay_id`, or
> 
>     * the unpredictability of the `delay_id`.

This part was resolved in https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/5dac49c7d5f4bb7de5f4dce8733c576e9f5df291. The original comment above is still open though.

Reviewer: This is obsolete now that delegation has been descoped from the MSC (https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/cd878d211fc855951e79cbb940bd68c304eecc87).

> It MUST be globally unique and SHOULD be cryptographically secure (in the sense that it is infeasible to predict).

Reviewer 15: What is "globally" here? Depending on the answer, I'm not sure how the homeserver is able to ensure this uniqueness?

Reviewer: Yeah, I can see how this might be slightly ambiguous. In my understanding, "globally" refers to the realm of the home server here. That is to say, when generating delay IDs, the server must take care to not produce collisions with other IDs for delayed events that it has currently stored. There is no requirement to keep delay IDs unique across home servers.

Does this make it more clear?

```suggestion
It MUST be unique with respect to the server's stored delayed events and SHOULD be cryptographically secure (in the sense that it is infeasible to predict).
```

Reviewer 15: This makes sense and makes it clearer :+1:

This way, the `delay_id` alone may be provided to approved external services to control the delayed event through the
unauthenticated management endpoints introduced [below](#managing-scheduled-delayed-events),

> with the cryptographic security of the `delay_id` preventing unapproved access to the delayed event.

Reviewer 15: > may be provided to approved external services

> preventing unapproved access

This does not make clear: un/approved by whom?

I guess the approval is supposed to be that of the user who sent the delayed event – but in practice, anyone that got hold of the `delay_id` once can forward it to others, including e.g. the homeserver. So I would argue that "approved external services" is a bit confusing, and "preventing unapproved access" overstates the security guarantees the `delay_id` can provide.

Reviewer: Yes, it's meant to be the client's choice what services it deems "approved" for delegating delayed event management to. Also agree on the potential overstating of the security guarantees.

Does this make it clearer?

```suggestion
This way, clients MAY share the `delay_id` with external services to allow delegated management of
their delayed events through the unauthenticated management endpoints introduced
[below](#managing-scheduled-delayed-events). The cryptographic security of the `delay_id` mitigates
unauthorized access to the delayed event.
```

Reviewer 15: Yes, this conveys better what's actually possible in my mind.

The only thing I'm not too sure about is the "The *cryptographic security* of the `delay_id`" part, it reads a bit weird (relating to [my comment](https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3529519943) above). But this fits better into the other discussion, I'm adding a comment there.

Reviewer: Good point. Out of the options in the other thread, "cryptographic randomness" sounds best to me here because it avoids us having to spell out the acronym again.

```suggestion
This way, clients MAY share the `delay_id` with external services to allow delegated management of
their delayed events through the unauthenticated management endpoints introduced
[below](#managing-scheduled-delayed-events). The cryptographic randomness of the `delay_id` mitigates
unauthorized access to the delayed event.
```

with an `errcode` of `M_FORBIDDEN` if the user doesn't have permission to send the event at the time of sending,
or HTTP 429 if the user has exceeded rate limits for sending room events at that time),
as if the request had been to send the event as a non-delayed event with either the `/send` or `/state` endpoint.
The homeserver SHOULD keep the delayed event scheduled, to account for the fact that the cause of the error
may resolve by the time of the delayed event's scheduled send time,
and to allow retries of the `send` action until then.

#### Delegating scheduled delayed events

It is useful for external services to also interact with scheduled delayed events.
If a client disconnects, an external service can be the best source to send the delayed event/"last will".

To permit this, the `delay_id` that uniquely identifies a delayed event also behaves as a scoped access token
that only allows to interact with the `POST /delayed_events/{delay_id}/{action}` endpoints on that specific `delay_id`.

With this, an SFU that tracks the current client connection state could be given the power to control the delayed event.

> The client would share the `delay_id` and the required details, so that the SFU can call the

Reviewer 15: What are the "required details" other that the `delay_id`, purely from the perspective of being able to call the necessary endpoint?

If there indeed are such additional required details, it could be good to spell them out here; conversely, if these details are only technical, SFU-specific ones not really to do with the delayed event management itself, the "and the required details" part could probably be omitted to avoid confusion.

Reviewer: Depending on how the external service is set up, other required details could for instance be the server name of the client's homeserver. The service needs to know this in order to determine what server to call the endpoints on.

```suggestion
The client would share the `delay_id` and other possibly required details (such as its homeserver's server name), so that the SFU can call the
```

> `POST /delayed_events/{delay_id}/refresh` endpoint while a user is connected

Reviewer 15: I suppose `/refresh` is from an older version of this MSC, and it should really be `/restart` here?

Reviewer: Indeed. Good catch!

```suggestion
`POST /delayed_events/{delay_id}/restart` endpoint while a user is connected
```

the Matrix call.

Since the SFU has a much lower chance of running into a network issue,
`POST /delayed_events/{delay_id}/restart` calls may be sent much more infrequently.
Instead of calling that endpoint every couple of seconds, a delayed event's
timeout can be set to be long (e.g. 6 hours), as the SFU can be expected to not forget sending the
`POST /delayed_events/{delay_id}/send` requests when it detects a disconnecting client.

### Getting delayed events

A set of new authenticated Client-Server API endpoints allows clients to look up
both scheduled and finalised delayed events owned by the requesting user.

The homeserver SHOULD apply rate limiting to these endpoints to provide mitigation against the
[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threat.
They most likely require (dependent on the implementation) serialization steps
and can be used to slow down the homeserver.

#### Getting a single delayed event

A new authenticated Client-Server API endpoint at
`GET /_matrix/client/v1/delayed_events/{delay_id}` responds with

> details on the delayed event with the specified `delay_id` owned by the requesting user.

Reviewer 15: There's a sort of duplicate authentication here: That of the requesting user via the standard authenticated endpoint mechanism, and through knowledge of the `delay_id` as explained in this MSC. While the management endpoints above only check for the latter, this `GET` endpoint should also check that the requesting, authenticated user owns the specified `delay_id`. This is somewhat suggested by the wording "the specified `delay_id` owned by the requesting user", but not entirely clear. I think it should be made explicit what happens if the given `delay_id` exists, but is not owned by the requesting user.

Reviewer 2: The current wording does intend to convey that requests for a `delay_id` belonging to another user will respond with 404.

b75b9fc7 rewords this to try to make it more clear.

This approach also requires using a query parameter for a purpose other than filtering,
which defies the usual semantics of a URI query.

### Sync-loop based heartbeats

[MSC4018: Reliable call membership] was an earlier attempt at making call membership more reliable.
It used the client's sync loop as an indicator to determine if the call membership event is expired.
Conceptually, this is similar to [MQTT]'s "Will Message" that is published by the server when a client
disconnects.

[MSC4018: Reliable call membership]: https://github.com/matrix-org/matrix-spec-proposals/pull/4018
[MQTT]: https://mqtt.org/

The advantage of this approach is that it doesn't require a new system for emitting heartbeats. An API
for letting clients store the leave event to be sent would still be required due to encryption, however.

On the downside, the common timeouts for syncs are much larger than what would be desirable for
call membership (30s vs. 5s). Overall this alternative feels overly implicit and less flexible than the
solution proposed above.

### Federated delayed events

Delayed events could be sent over federation immediately and then have the receiving homeservers process

> (sent down to clients) them at the appropriate time.

Reviewer: I think we can shorten the formulation here slightly.

```suggestion
Delayed events could be sent over federation immediately to then have the receiving homeservers send them down to clients at the appropriate time.
```

- `delaying_since` - `since` might be a clearer suffix than `from`
- `last_restart` - but this feels less clear than other names for a delayed event that hasn't been restarted

An alternative field altogether is `send_ts`, with a value of `delay_ms` + the start time of the timer.
However, explicitly returning the scheduled send time suggests a strong guarantee of exactly when a delayed event will
be sent, despite this proposal allowing homeservers to adjust the scheduled send time to support batch sending.

### Syncing failed delayed events

Currently, clients have to fetch the delayed event info after the timeout to find an error in case the event failed,
or if another client belonging to the same user had cancelled a scheduled delayed event.
We could instead define a new method to push failed & cancelled delayed events down `/sync` to the sender.
For application services, this information would need to be pushed via transactions.
However, this is not strictly necessary for delayed events to be usable, and may thus be discussed in a separate MSC
in the interest of keeping this MSC focused on the core functionality of delayed events.

> ## Security considerations

Reviewer 15: An aspect I'm missing here is the fact that the `POST /delayed_events/{delay_id}/{action}` and `GET /_matrix/client/v1/delayed_events/{delay_id}` endpoints are including the `delay_id`, which is aptly described to "behave as a scoped access token". This risks a `delay_id`, i.e. an access token, getting leaked to various logs.

Can this risk be avoided / reduced by putting the `delay_id` in a header field or the request body instead? If there's technical reasons to not do that, this should at least be mentioned in the Security Considerations.

Reviewer: I suspect the main reason is that in the path in makes for a very natural REST API shape.

I think putting it into the body would require changing `GET /_matrix/client/v1/delayed_events/{delay_id}` to POST. It's not as nice an API but would probably work.

As a header, we could leave the HTTP method unchanged. The API shape strikes me as equally odd but technically it should also be possible.

@AndrewFerr curious what you think? Have I missed any reasons why this needs to be in the path?

Reviewer 2: Yes, the main reason is to be able to use the `delay_id` as a typical "identifier" token.

Besides, there's limited effectiveness in trying to hide a `delay_id` by moving it into request headers / body / elsewhere, because it will nevertheless appear in client logs via `/sync` responses once its associated event gets sent, as per [`delay_id` in `unsigned` event data](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/expiring-events-keep-alive/proposals/4140-delayed-events-futures.md#delay_id-in-unsigned-event-data).

Reviewer 15: > it will nevertheless appear in client logs via /sync responses once its associated event gets sent

But once the event gets sent, is there even a need to keep the `delay_id` secret any longer? Aren't all of the `delay_id`'s access token capabilities void once the respective event is finalised?

See this part of the MSC:
> If the target delayed event is already finalised with an outcome that conflicts with the action, i.e. if the action is send or restart and the delayed event has already been cancelled, or if the action is cancel and the delayed event has already been sent, the homeserver will respond with HTTP 409 and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response) with an errcode of M_UNKNOWN.

Reviewer 2: > But once the event gets sent, is there even a need to keep the delay_id secret any longer? Aren't all of the delay_id's access token capabilities void once the respective event is finalised?

Good point, that's correct.

In the interest of unblocking this, I'll split off the delegation feature into its own MSC. For now, I've moved it to an alternative: cd878d21

---

With that out of the way, I'll say that I'm still hesitant to protect `delay_id`s as much as access tokens, for a few reasons:
- It would preclude future endpoints from being able to use `delay_id`s in ways that would "leak" them by design.  A realistic possible addition is for `/sync` to include information about newly-scheduled delayed events (so that clients other than the one that scheduled an event would be notified of it, instead of having to manually hit the lookup endpoint to discover it).
- It may add friction against migrating to using OAuth 2.0 scopes for the management endpoints.

Reviewer 15: With the delegation removed from this MSC, and the management endpoints being authenticated, I'm obviously fine with the security considerations in this MSC :+1: 

I'd be interested to see the additional delegation MSC when it's ready though – especially how it plans to deal with the authentication for the management endpoints prescribed in this MSC – and I think it would be good the mention the concerns form your comment about the handling of the `delay_id` there.

with the cryptographic security of the `delay_id` preventing unapproved access to the delayed event.

The homeserver MUST NOT send the event before the scheduled time.
To support batch sending, homeservers MAY add up to 30 seconds to the scheduled send time.
Note: clients might find that their events are delayed further due to server load and similar conditions.

The homeserver MUST persist scheduled delayed events such that they will not be lost by the homeserver being restarted.
Moreover, when a homeserver restarts, it MUST scan for all scheduled delayed events whose send time has already passed
(i.e. delayed events that were scheduled to be sent while the homeserver was offline),
and send those delayed events as soon as possible, in chronological order of their scheduled send times.

The homeserver MAY enforce a maximum allowed delay for delayed events. This limit is
communicated to the client in a capability (described later in this proposal).
If a requested delay exceeds this maximum, the homeserver will respond with HTTP 400
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)

> with an `errcode` of `M_INVALID_PARAM`.

Reviewer 2: To go even further with https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3483103991, how about returning `M_FORBIDDEN` & HTTP 403 even for this case of requesting a delay longer than allowed?

This would prevent needing a special case error response for delayed events being entirely disallowed, as both that case & this one would then both give the same response.

It also looks like the spec uses `M_INVALID_PARAM` only for parameter values that are _always_ invalid (like a malformed room alias or MXID), as opposed to values that are disallowed by server config that may change later.

Reviewer 2: Erm, actually, the special case would still be needed for when the maximum per-user amount of delayed events is 0, lest the response would be `M_LIMIT_EXCEEDED` & HTTP 429 for a limit that can never be satisfied.

But the point stands about `M_FORBIDDEN` & HTTP 403 potentially being more appropriate than `M_INVALID_PARAM` / HTTP 400.

Reviewer 2: In the interest of progressing with the Synapse implementation of MSC4140 error codes ([PR](https://github.com/element-hq/synapse/pull/19539)), I've committed this change as 9447ee01.

I've also updated the PR description accordingly.

Reviewer: > But the point stands about `M_FORBIDDEN` & HTTP 403 potentially being more appropriate than `M_INVALID_PARAM` / HTTP 400.

This sounds sensible to me.

Reviewer 8: `M_FORBIDDEN` for a parameter exceeding a limit sounds weird, maybe it should just be a new dedicated error code instead?

Reviewer 2: A dedicated error code was the original plan, but was decided against after opting to express the max-allowed delay in a capability: https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3257895622, https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3259516213.

There is also precedent of the spec using `M_FORBIDDEN` for disallowed parameter values, like an invalid password or login token for the [user-interactive login API](https://spec.matrix.org/latest/client-server-api/#user-interactive-api-in-the-rest-api), and an unsupported media type for [the media upload endpoint](https://spec.matrix.org/latest/client-server-api/#put_matrixmediav3uploadservernamemediaid).

Reviewer 8: Personally I'd err towards adding new error codes even if they're not technically necessary, rather than reusing existing codes in a potentially confusing way. I don't feel too strongly about that though, so either way is fine

> an unsupported media type for [the media upload endpoint](https://spec.matrix.org/latest/client-server-api/#put_matrixmediav3uploadservernamemediaid)

not related to this MSC, but ???? HTTP 415 exists

Reviewer 2: This has gone back to using a dedicated error code: 0c7193aa

  These objects contain the same fields as the object returned by
  [the single-item lookup](#getting-a-single-delayed-event),
  except for the fields exclusive to finalised delayed events (`error`, `event_id`, and `finalised_ts`).

```http
200 OK
Content-Type: application/json

{
  "delayed_events": [
    {
      "delay_id": "...",
      "room_id": "!roomid:example.com",
      "type": "m.room.message",
      "delay": 5500,
      "running_since": 1721732853284,
      "content": {
        "msgtype": "m.text",
        "body": "I am now offline"
      }
    },
    ...
  ]
}
```

#### Retention of finalised delayed events

The amount of finalised events that stay on the homeserver can be chosen by the homeserver.
The recommended strategy is to retain finalised events for up to 7 days or 1000 events per user,
whichever occurs first.

There is no guarantee for a client that events will be available
if they exceed the limits of their homeserver.
Additionally, a homeserver MAY discard finalised delayed events that have been returned by a
`GET /_matrix/client/v1/delayed_events/{delay_id}` response.

> ### Additional homeserver behaviour

Reviewer 16: in my reading, it is implied that when a client sends a delayed event, the server should effectively persist it in some table of delayed events and eventually as the delay runs out persist it from there like a regular event "on the client's behalf". we don't usually have the following problem in matrix, but with this new mechanic it is possible that a server is offline at the time it should send the delayed event. perhaps this MSC should clarify explicitly how servers should handle this case. E.g. the temporary persistence while the event is still delayed should not be ephemeral storage such as RAM, but in a DB table proper. Upon startup a server should then check this table for "missed" delays and do something with the events contained.

Reviewer 2: > it is possible that a server is offline at the time it should send the delayed event. perhaps this MSC should clarify explicitly how servers should handle this case.

Good idea.  The proposal shouldn't be too strict in saying how this should be done, but the bare minimum is for servers to not lose scheduled delayed events.

> Upon startup a server should then check this table for "missed" delays and do something with the events contained.

FTR this is precisely what Synapse does.

Reviewer 2: Added by fcffc2db

#### `delay_id` in `unsigned` event data
The `delay_id` of a sent delayed event MUST be included in the resulting room event's `unsigned` data

> if, and only if, the client being given the event is authenticated as the event's sender.

Reviewer 17: Could you help me understand this, please ? The unsigned data is part of the PDU sent to the room, correct ? Its content does not depend on the client receiving it.

Reviewer 2: > The unsigned data is part of the PDU sent to the room, correct ? Its content does not depend on the client receiving it.

Not quite; servers can restrict which fields of a persisted event that a client may receive.  Prior art is how various CS-API endpoints withhold an event's transaction ID from all but the user who sent the event.  This is codified in the spec as responses that include a `transaction_id` field "if the client being given the event is the same one which sent it."

Reviewer 17: I see. It would be helpful if the spec (and also this MSC) were a bit more clear on how this is done. As it's currently specified, this result could be achieved by putting the field in the PDU and then having the homeserver filter it out if needed when sending it to the client; or by *not* putting it in the PDU, and having the homeserver add it if needed when sending it to the client.

If I understand you correctly, it should be the first one ?

Reviewer 2: If I'm not mistaken, PDUs are solely for server-server communication, and thus they don't need to carry `delay_id`s at all due to delayed events not being federated.

> As it's currently specified, this result could be achieved by putting the field in the PDU and then having the homeserver filter it out if needed when sending it to the client

This is pretty much how it's done in Synapse now, except the event is stored as ad-hoc internal representation instead of a "formal" PDU.  When formatting the event in responses, the server decides which fields to include depending on the user requesting it.

Reviewer 15: I'm also quite confused by this paragraph. I might be missing some general background knowledge, but I think what might be missing is a mention of the context that "the client being given the event" happens in. It's also not made explicit that the homeserver (I guess?) is doing the "giving" here.

Reviewer: In my understanding, this only applies when a homeserver hands out the event to its local users, e.g. via `/sync` or `/messages`. Maybe it's sufficient to mention these endpoints as examples?

```suggestion
if, and only if, the client being given the event (e.g. over `/sync` or `/messages`) is authenticated as the event's sender.
```

Reviewer 15: yes that makes it clearer to me already 👍

Reviewer 2: > Maybe it's sufficient to mention these endpoints as examples?

2f3b817c now includes those examples.

a list of details about scheduled delayed events owned by the requesting user.

Delayed events are returned in chronological order of their intended send time, which is `running_since` + `delay`.

On success, the response is HTTP 200 and a JSON object containing the following fields:

- `delayed_events` - An array of objects describing delayed events owned by the requesting user.
  These objects contain the same fields as the object returned by
  [the single-item lookup](#getting-a-single-delayed-event),
  except for the fields exclusive to finalised delayed events (`error`, `event_id`, and `finalised_ts`).

```http
200 OK
Content-Type: application/json

{
  "delayed_events": [
    {
      "delay_id": "...",
      "room_id": "!roomid:example.com",
      "type": "m.room.message",
      "delay": 5500,
      "running_since": 1721732853284,
      "content": {
        "msgtype": "m.text",
        "body": "I am now offline"
      }
    },
    ...
  ]
}
```

#### Retention of finalised delayed events

The batch size and the amount of finalised events that stay on the homeserver can be chosen by the homeserver.
The recommended values are:

> - batch size: 10

Reviewer 17: Batch size has been removed from this MSC.

Reviewer 2: Good catch, thanks.  Removed by 0ae35c4d

scheduled to be sent later), or other ephemeral events such as temporary power level changes.

[Teams]: https://support.microsoft.com/en-us/office/schedule-chat-messages-in-microsoft-teams-2fc5ea77-7bb4-4511-8f59-e62bac1c0f6a
[Telegram]: https://telegram.org/blog/scheduled-reminders-themes

Another possible use case are reliable "hang up" events in VoIP calls. [MSC4143], for instance,
communicates call membership through timeline events. Leaving a call requires sending a new
event. The leaving client may not be able to send this event on its own, however, as it may lose
connectivity due to network issues or because the application was shutdown unexpectedly.
In this situation it would be helpful, if the client could schedule its "hang up" event to be
sent by the server at a defined later time. While the client is still connected and in the call,
it could repeatedly push the scheduled time forward as a kind of "heartbeat" mechanism. If the
client then loses connectivity, the server would emit the "hang up" event at the scheduled time
resulting in reliable call membership status for other participants.

> 

Reviewer: Looks like a link is missing here.

```suggestion

[MSC4143]: https://github.com/matrix-org/matrix-spec-proposals/pull/4143

```

- Send a scheduled delayed event immediately
- Cancel a scheduled delayed event so that it is never sent

At the point of an event being scheduled, the homeserver is unable to allocate the event ID[^eventId].
Instead, the homeserver allocates a `delay_id` to the scheduled event which is used during the above API operations.

### Scheduling a delayed event

A new authenticated Client-Server API endpoint at
`PUT /_matrix/client/v3/rooms/{roomId}/delayed_event/{eventType}/{txnId}` allows clients to
schedule the sending of a message or state event.

The path parameters of this endpoint are the same as for the existing
[`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](
https://spec.matrix.org/v1.18/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
endpoint. It also makes use of [transaction identifiers](
https://spec.matrix.org/v1.18/client-server-api/#transaction-identifiers),
and supports [timestamp massaging](https://spec.matrix.org/latest/application-service-api/#timestamp-massaging)
when called by an application service.

The body for requests to this endpoint is a JSON object containing the following fields:

> - `delay` - Required. A positive non-zero number of milliseconds the homeserver should wait before sending the event.

Reviewer 2: Is there any interest in renaming this to `delay_ms`, to make it crystal clear that the value is in milliseconds?

If this were to be renamed, then so too should the `delay` field in responses to the `GET` endpoints.

Additionally, the `max_delay` in the [new capability](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/expiring-events-keep-alive/proposals/4140-delayed-events-futures.md#delayed-event-limits-as-a-capability) would then benefit from being renamed to `max_delay_ms`.

For comparison, the names of most other specced millisecond-valued fields/parameters end with `_ms`, with the following exceptions:
- the `unsigned.age` field in events ([ref](https://spec.matrix.org/v1.18/client-server-api/#room-event-format))
- the `timeout` query parameter for `/sync` ([ref](https://spec.matrix.org/v1.18/client-server-api/#syncing))
  - Amusingly, there is also a parameter named `timeout_ms` used by the Content Repository module ([ref](https://spec.matrix.org/v1.18/client-server-api/#content-repo-client-behaviour))
- the `info.duration` field in `m.audio` & `m.video` events ([ref](https://spec.matrix.org/v1.18/client-server-api/#mroommessage-msgtypes))
- the `lifetime` field in `m.call.invite` and `m.call.negotiate` events ([ref](https://spec.matrix.org/v1.18/client-server-api/#events-4))
- the `timeout` body parameter in `/typing/{userId}` ([ref](https://spec.matrix.org/v1.18/client-server-api/#client-behaviour-4))
- The `last_active_ago` field in `m.presence` events ([ref](https://spec.matrix.org/v1.18/client-server-api/#events-9))

Reviewer 2: Done by fd27ef00.

Reviewer: I've documented this in the known implementation differences in the PR description (and hopefully did so correctly).

  would normally `await` a future in various programming languages.
- Postponed Events: Similar to "delayed events" but longer.
- Last Will: Seems unfitting because the feature can be used regardless of whether the client goes offline
  or not. The term also feels somewhat dark.

### Don't provide a `send` action

Instead of providing a `send` management action for scheduled delayed events,
the client could cancel the scheduled delayed event and send a new non-delayed event instead.

This would simplify the API, but it's less efficient since the client would have to send two requests instead of one.

### Use `DELETE` HTTP method for `cancel` action

Instead of providing a `cancel` management action for scheduled delayed events,
the client could send a `DELETE` request to an endpoint representing a target delayed event.

This feels more elegant, but it doesn't feel like a good suggestion for how the other actions are mapped.
Also, `DELETE` suggests that the target resource will be truly deleted, but this is at odds with how
cancelling a delayed event has it retained as a finalised event for later lookup.

### Alternative to `running_since` field

> Some alternatives for the `running_since` field on the `GET` response are:

Reviewer 2: At the risk of bikeshedding, how does `scheduled_at` sound?  Or at least something suffixed with `_at`, which the spec uses for a few other timestamp-valued fields.

I also feel that `scheduled_at` is less likely than `scheduled_ts` to be misinterpreted as the intended send time of the event, and thus is a good opportunity to put `scheduled` in the name, which is an existing keyword used elsewhere in this module, instead of having to introduce a new term (like "running" or "created").

Reviewer 2: Made the leap to prefer `scheduled_at` with f1845796.

Reviewer: I've documented this in the known implementation differences in the PR description (and hopefully did so correctly).

Reviewer 2: There is still contention with `scheduled_at`, which is being discussed here: https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3708744335

On success, the response is HTTP 200 and a JSON object containing the following fields:

- `delayed_events` - An array of objects describing delayed events owned by the requesting user.
  These objects contain the same fields as the object returned by
  [the single-item lookup](#getting-a-single-delayed-event).

```http
200 OK
Content-Type: application/json

{
  "delayed_events": [
    {
      "delay_id": "...",
      "room_id": "!roomid:example.com",
      "type": "m.room.message",
      "delay": 5500,
      "running_since": 1721732853284,
      "content": {
        "msgtype": "m.text",
        "body": "I am now offline"
      }
    },
    ...
  ]
}
```

#### Retention of finalised delayed events

The batch size and the amount of finalised events that stay on the homeserver can be chosen by the homeserver.
The recommended values are:

- batch size: 10
- `finalised` retention: up to 7 days or 1000 events per user, whichever occurs first

There is no guarantee for a client that all events will be available in the
finalised events list if they exceed the limits of their homeserver.
Additionally, a homeserver MAY discard finalised delayed events that have been returned by a

> `GET /_matrix/client/v1/delayed_events(/{delay_id})?` response.

Reviewer: I think we should generalize this a little now that the listing API extensions live in https://github.com/matrix-org/matrix-spec-proposals/pull/4486.

```suggestion
There is no guarantee for a client that events will be available
if they exceed the limits of their homeserver.
Additionally, a homeserver MAY discard finalised delayed events that have been returned by a
`GET /_matrix/client/v1/delayed_events(/{delay_id})?` response.
```

Conceptually, this is similar to [MQTT](https://mqtt.org/)'s "Will Message" that is published by the server
when a client disconnects.

The advantage is that this does not require introducing a new ping system
(as is proposed here by using the `/delayed_events/{delay_id}/restart` endpoint).
Though with cryptographic identities, the client needs to create the leave event.

The timeout for syncs are much slower than what would be desirable (30s vs 5s).

With a widget implementation for calls, it cannot be guaranteed that the widget is running during the sync loop.
So one either has to move the hangup logic to the hosting client or let the widget run all the time.

A dedicated ping (independent to the sync loop) is more flexible and allows for the widget to
execute the timer restart.
If the widget dies, the call membership will disconnect.

Additionally, the specification should not include specific
custom server rules if possible.
Sending an event on behalf of a user based on the client sync loop if there is an event with a specific type and
specific content is quite a server-specific behaviour, and also would not work well with encrypted state events and
cryptographic identities.

> This proposal is a general behaviour valid for all event types.

Reviewer: Given the overall length of the proposal, this section still feels slightly too wordy to me. It also refers to widgets which are secondary to the MSC and, I think, should not be mentioned because they are themselves unspecced.

```suggestion
### Sync-loop based heartbeats

[MSC4018: Reliable call membership] was an earlier attempt at making call membership more reliable.
It used the client's sync loop as an indicator to determine if the call membership event is expired.
Conceptually, this is similar to [MQTT]'s "Will Message" that is published by the server when a client
disconnects.

[MSC4018: Reliable call membership]: https://github.com/matrix-org/matrix-spec-proposals/pull/4018
[MQTT]: https://mqtt.org/

The advantage of this approach is that it doesn't require a new system for emitting heartbeats. An API
for letting clients store the leave event to be sent would still be required due to encryption, however.

On the downside, the common timeouts for syncs are much larger than what would be desirable for
call membership (30s vs. 5s). Overall this alternative feels overly implicit and less flexible than the
solution proposed above.
```

- Retrieve details on scheduled and finalised delayed events
    - A delayed event is said to be "finalised" when it has been sent,
      or has been cancelled due to user action or an error on the attempt to send the event.
- Restart the timer of a scheduled delayed event
- Send a scheduled delayed event immediately
- Cancel a scheduled delayed event so that it is never sent

At the point of an event being scheduled, the homeserver is unable to allocate the event ID[^eventId].
Instead, the homeserver allocates a `delay_id` to the scheduled event which is used during the above API operations.

### Scheduling a delayed event

A new authenticated Client-Server API endpoint at
`PUT /_matrix/client/v3/rooms/{roomId}/delayed_event/{eventType}/{txnId}` allows clients to
schedule the sending of a message or state event.

The path parameters of this endpoint are the same as for the existing
[`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](
https://spec.matrix.org/v1.18/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
endpoint. It also makes use of [transaction identifiers](
https://spec.matrix.org/v1.18/client-server-api/#transaction-identifiers),

> and supports [timestamp massaging](https://spec.matrix.org/latest/application-service-api/#timestamp-massaging)

Reviewer 6: ```suggestion
and supports [timestamp massaging](https://spec.matrix.org/v1.18/application-service-api/#timestamp-massaging)
```

"latest" can change without notice and break historical links - we prefer that MSCs use a tagged spec release instead.

unauthenticated management endpoints introduced [below](#managing-scheduled-delayed-events),
with the cryptographic security of the `delay_id` preventing unapproved access to the delayed event.

The homeserver MUST NOT send the event before the scheduled time.
To support batch sending, homeservers MAY add up to 30 seconds to the scheduled send time.
Note: clients might find that their events are delayed further due to server load and similar conditions.

The homeserver MUST persist scheduled delayed events such that they will not be lost by the homeserver being restarted.
Moreover, when a homeserver restarts, it MUST scan for all scheduled delayed events whose send time has already passed
(i.e. delayed events that were scheduled to be sent while the homeserver was offline),
and send those delayed events as soon as possible, in chronological order of their scheduled send times.

The homeserver MAY enforce a maximum allowed delay for delayed events.
If a requested delay exceeds this maximum, the homeserver will respond with HTTP 400
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)

> with an `errcode` of `M_INVALID_PARAM`.

Reviewer 6: ```suggestion
The homeserver MAY enforce a maximum allowed delay for delayed events. This limit is
communicated to the client in a capability (described later in this proposal).
If a requested delay exceeds this maximum, the homeserver will respond with HTTP 400
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
with an `errcode` of `M_INVALID_PARAM`.
```

to clarify that the MSC doesn't really suggest a limit, but the limit is known to clients so they aren't left completely in the dark about what is "acceptable".

The homeserver SHOULD apply rate limiting to the scheduling of delayed events to provide mitigation against the
[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threat.

The homeserver SHOULD enforce a limit of how many delayed events a user may have scheduled at once
to provide mitigation against both the
[High Volume of Messages](https://spec.matrix.org/v1.18/appendices/#threat-high-volume-of-messages) and
[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threats.
If a user's request to schedule a delayed event would exceed this limit, the homeserver will respond with HTTP 429,

> a [standard error response](https://spec.matrix.org/latest/client-server-api/#standard-error-response)

Reviewer 6: ```suggestion
a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
```

with an `errcode` of `M_LIMIT_EXCEEDED`, and a `Retry-After` header whose value is set to the time of/until
the scheduled send time of the next of the user's delayed events to be sent,
rounded up to the nearest second.

```http
429 Too Many Requests
Content-Type: application/json
Retry-After: 1200

{
  "errcode": "M_LIMIT_EXCEEDED",
  "error": "The maximum number of delayed events has been reached.",
}
```

As a special case, if the homeserver has set either of these limits such that scheduling delayed events is disallowed
(i.e. it sets the maximum allowed delay to 0 seconds, or a limit of 0 scheduled delayed events per user), it may respond

> to event scheduling requests with HTTP 400 and a standard error response with an `errcode` of `M_UNKNOWN`.

Reviewer 6: Shouldn't this be `M_FORBIDDEN` as the maximum delay condition is impossible to satisfy?

Reviewer 2: Sure.  I was under the impression that `M_FORBIDDEN` was to be used only for insufficient permissions.

On that note, I thought the same was true for HTTP 403 as well, but it turns out that it can be used for "reasons unrelated to the credentials [of the request]" ([reference](https://www.rfc-editor.org/info/rfc9110/#status.403)).

So it may be most appropriate for this error response to use both `M_FORBIDDEN` and HTTP 403.  That change is made by d91a9321.

Reviewer 2: Further discussion on this special case is here: https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3729390953

Resolving this thread to avoid duplication.

- `cancel` - Cancel the delayed event so that it is never sent.
- `restart` - Reset the delayed event's scheduled send time to be the current time + its original `delay_ms`.

For example, the following would send the delayed event with `delay_id` `1234567890` immediately:

```http
POST /_matrix/client/v1/delayed_events/1234567890/send
Content-Type: application/json

{
}
```

These endpoints are unauthenticated so that control over a particular scheduled delayed event may be
[delegated to an external service](#delegating-scheduled-delayed-events)
by sharing the target delayed event's `delay_id` with the service.

Where the `action` is `send`, the homeserver SHOULD apply rate limiting to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.18/appendices/#threat-high-volume-of-messages) threat.

For all `action`s, the homeserver SHOULD apply rate limiting to provide mitigation against the
[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threat.

If no delayed event with the specified `delay_id` can be found,
the homeserver will respond with HTTP 404

> and a [standard error response](https://spec.matrix.org/latest/client-server-api/#standard-error-response)

Reviewer 6: ```suggestion
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
```

with an `errcode` of `M_NOT_FOUND`.

On success, the homeserver will respond with HTTP 200
and a response body of an empty object. A future MSC may define additional keys, such as returning
the event ID for an `action` of `send` or the new expected send time for an `action` of `restart`.

To allow safely retrying requests, the homeserver will respond with success
if the target delayed event is already finalised with an outcome that matches the `action`, i.e.
if the `action` is `send` and the delayed event has already been sent,
or if the `action` is `cancel` and the delayed event has already been cancelled
(either due to user action or an error).

If the target delayed event is already finalised with an outcome that conflicts with the `action`, i.e.
if the `action` is `send` or `restart` and the delayed event has already been cancelled,
or if the `action` is `cancel` and the delayed event has already been sent,
the homeserver will respond with HTTP 409

> and a [standard error response](https://spec.matrix.org/latest/client-server-api/#standard-error-response)

Reviewer 6: ```suggestion
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
```

with an `errcode` of `M_UNKNOWN`.

If the action is `send` and the delayed event is unable to be sent due to an error,
the homeserver will respond with that error (e.g. HTTP 403

> and a [standard error response](https://spec.matrix.org/latest/client-server-api/#standard-error-response)

Reviewer 6: ```suggestion
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
```

`POST /delayed_events/{delay_id}/restart` calls may be sent much more infrequently.
Instead of calling that endpoint every couple of seconds, a delayed event's
timeout can be set to be long (e.g. 6 hours), as the SFU can be expected to not forget sending the
`POST /delayed_events/{delay_id}/send` requests when it detects a disconnecting client.

### Getting delayed events

A set of new authenticated Client-Server API endpoints allows clients to look up
both scheduled and finalised delayed events owned by the requesting user.

The homeserver SHOULD apply rate limiting to these endpoints to provide mitigation against the
[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threat.
They most likely require (dependent on the implementation) serialization steps
and can be used to slow down the homeserver.

#### Getting a single delayed event

A new authenticated Client-Server API endpoint at
`GET /_matrix/client/v1/delayed_events/{delay_id}` responds with
details on the delayed event with the specified `delay_id` owned by the requesting user.

If no such delayed event can be found, the homeserver will respond with HTTP 404

> and a [standard error response](https://spec.matrix.org/latest/client-server-api/#standard-error-response)

Reviewer 6: ```suggestion
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
```

- `delay_id` - Required. The ID of the delayed event.
- `room_id` - Required. The ID of the room that the delayed event was scheduled to be sent in.
- `type` - Required. The event type of the delayed event.
- `state_key` - The state key of the delayed event if it is a state event; absent otherwise.
- `delay_ms` - Required. The delay in milliseconds after the point of scheduling that the event is/was to be sent at.
- `scheduled_at` - Required. The timestamp (as Unix time in milliseconds) when the delayed event was scheduled or
  last restarted.
- `content` - Required. The content of the delayed event.
  This is the body of the original `PUT` request, not a preview of the full event after sending.
- `error` - Present only for finalised events that were cancelled due to an error.
  The [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
  of the error that prevented the delayed event from being sent.
- `event_id` - The `event_id` this event got in case it was sent.
- `finalised_ts` - The timestamp (as Unix time in milliseconds) when the event was finalised;
  absent if it is still scheduled.

>   Using [timestamp massaging](https://spec.matrix.org/latest/application-service-api/#timestamp-massaging)

Reviewer 6: ```suggestion
  Using [timestamp massaging](https://spec.matrix.org/v1.18/application-service-api/#timestamp-massaging)
```

by a malicious homeserver.

In the context of this proposal, the delayed events should be signed with the same cryptographic identity as the client
that scheduled them.

This means that the content of the original scheduled event must be sent "as is" without modification by the homeserver.
The consequence is an implementation detail that client developers must be aware of: if the content of the delayed
event contains a timestamp, then it would be the timestamp of when the event was originally scheduled rather than
anything later.

However, the `origin_server_ts` of the delayed event should be the time that the event is actually sent
by the homeserver.

This is a general problem that arises with the introduction
of [Cryptographic Identities](https://github.com/matrix-org/matrix-spec-proposals/pull/4080).
A user can intentionally, or caused by network conditions, delay the signing and sending of an event.
A possible solution would be the introduction of a `signing_ts` (in the signed section) and keep the `origin_server_ts`
in the unsigned section.
Both are reasonable data points that clients might want to use.
This would solve issues related to delayed events since

> it would make it transparent to clients, when an event was scheduled and when it was distributed over federation.

Reviewer 6: It's a little surprising to see this section here because the referenced MSC is a draft and in a much earlier stage of the process. Typically, this would mean it's MSC4080's problem to be compatible with this MSC rather than the other way around, though considerations in this MSC are certainly welcome where they make sense/are easy to do. I would suggest reducing the amount of page space we give to it however. 

```suggestion
### Compatibility with Cryptographic Identities

Proposals like [MSC4080](https://github.com/matrix-org/matrix-spec-proposals/pull/4080)
enable a client to verify that an event was *actually* sent by the sender's devices. This
proposal shouldn't affect how that verification works because those other proposals will
need to account for eventual consistency anyway, which may appear as a delayed event or
attached to a disjointed part of the DAG.
```

[MSC2716: Incrementally importing history into existing rooms](
https://github.com/matrix-org/matrix-spec-proposals/pull/2716) already proposes a `batch_send` endpoint.
However, it is limited to application services and focuses on historic
data. Since the additional capability to use a template `event_id` parameter is also needed,
this probably is not a good fit.

### Reusing the `send`/`state` endpoints

Instead of creating a new endpoint for scheduling a delayed event,
the `/send` and `/state` endpoints could support sending events with a delay,
via an optional query parameter for specifying the desired delay:

`PUT /_matrix/client/v1/rooms/{roomId}/send/{eventType}/{txnId}?delay={delay_ms}`

`PUT /_matrix/client/v1/rooms/{roomId}/state/{eventType}/{stateKey}?delay={delay_ms}`

This would require the endpoints to give a different response body when sending a delayed event,
as the usual response of the event ID is unavailable at the time of scheduling a delayed event.

The main benefit of reusing the endpoints as described is the potential for code reuse in homeserver implementations.

> However, this could be negated by the complexity of having to support multiple response body formats.

Reviewer 6: Do we have evidence of this being attempted? It seems like a great alternative (we aren't that worried about different response formats, generally speaking).

Reviewer: The currently merged implementations (https://github.com/element-hq/synapse/pull/17326 & https://github.com/matrix-org/matrix-js-sdk/pull/4294) use query strings. There are follow-up unmerged PRs that switch to dedicated endpoints (https://github.com/element-hq/synapse/pull/19354 & https://github.com/matrix-org/matrix-js-sdk/pull/5133).

Reviewer 6: hmm, I'm worried that there's seemingly viable implementations for both the MSC and the alternative. Where was the motivation to switch to a dedicated endpoint discussed? (links please)

Reviewer 6: Mentioned on a call: there was a discussion thread a few months ago which requested the split. On the call, multiple teams nodded in approval to the sentiments of having a different endpoint.

The ask is to incorporate a link into the MSC for that discussion, and include context in case links break/change (again).

Reviewer 2: 955ecf5d adds a link to the criticisms of the shared endpoint that motivated the move to a dedicated endpoint, and expands upon it lest the deep link gets lost & so that the MSC alone may be sufficient in communicating what the criticisms were.

Reviewer: Resolving as the request from https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3499401628 appears to be addressed now.

This approach also requires using a query parameter for a purpose other than filtering,

> which defies the usual semantics of a URI query.

Reviewer 6: We already use query strings for non-filtering purposes, so this doesn't hold up. If it's expected to stay in the proposal, a citation would be needed. (I'd suggest just removing the sentence.)

### Sync-loop based heartbeats

[MSC4018: Reliable call membership] was an earlier attempt at making call membership more reliable.
It used the client's sync loop as an indicator to determine if the call membership event is expired.
Conceptually, this is similar to [MQTT]'s "Will Message" that is published by the server when a client
disconnects.

[MSC4018: Reliable call membership]: https://github.com/matrix-org/matrix-spec-proposals/pull/4018

> [MQTT]: https://mqtt.org/

Reviewer 6: ```suggestion
Conceptually, this is similar to [MQTT's "Will Message"] that is published by the server when a client
disconnects.

[MSC4018: Reliable call membership]: https://github.com/matrix-org/matrix-spec-proposals/pull/4018
[MQTT's "Will Message"]: https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901040
```

Linking to spec is always better than not :)

- Cancel a delayed event so that it is never sent

At the point of an event being scheduled the homeserver is [unable to allocate the event ID](#allocating-the-event-id-at-the-point-of-scheduling-the-send).
Instead, the homeserver allocates a _delay ID_ to the scheduled event which is used during the above API operations.

### Scheduling a delayed event

An optional `delay` query parameter is added to the existing
[`PUT /_matrix/client/v3/rooms/{roomId}/state/{eventType}/{stateKey}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
and
[`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidstateeventtypestatekey)
endpoints.

The new query parameter is used to configure the event scheduling:

- `delay` - Optional number of milliseconds the homeserver should wait before sending the event. If no `delay` is provided,
the event is sent immediately as normal.

The body of the request is the same as currently.

If a `delay` is provided, the homeserver schedules the event to be sent with the specified delay and returns the _delay ID_

> in the `delay_id` field (omitting the `event_id` as it is not available):

Reviewer 7: It does not seem like a good idea to use the same version of the endpoints for this feature because the `event_id` in the response is currently required. Making it optional is not backwards-compatible.

Reviewer 7: Actually, there will always be an `event_id` for clients that don't support this feature, so it should be fine. Sorry for the noise.

Reviewer 4: Exactly. It is backwards compatible by virtue of the client having to opt-in.

  - `delay` - Required. The delay in milliseconds before the event is sent.
  - `running_since` - Required. The timestamp (as unix time in milliseconds) when the delayed event was scheduled or
    last restarted.
  - `content` - Required. The content of the delayed event.
- `next_batch` - Optional. A token that can be used to paginate the list of delayed events.

For example:

```http
200 OK
Content-Type: application/json

{
  "delayed_events": [
    {
      "delay_id": "1234567890",
      "room_id": "!roomid:example.com",
      "type": "m.room.message",
      "delay": 15000,
      "running_since": 1721732853284,
      "content":{
        "msgtype": "m.text",
        "body": "I am now offline"
      }
    },
    {
      "delay_id": "abcdefgh",
      "room_id": "!roomid:example.com",
      "type": "m.call.member",
      "state_key": "@user:example.com_DEVICEID",
      "delay": 5000,
      "running_since": 1721732853284,
      "content":{
        "memberships": []
      }
    }
  ],
  "next_batch": "b12345"
}
```

`running_since` is the timestamp (as unix time in milliseconds) when the delayed event was scheduled or last refreshed.

> So, unless the delayed event is updated beforehand, the event will be sent after `running_since` + `delay`.

Author: I first understood `running_since` as the time it was initially sent.
`last_restart: timestamp` is what we could use.

Reviewer 4: The problem with `last_restart` as a name is that it doesn't work in the case that the delayed events is never restarted and is sent after the initial delay.

Does `delaying_from` feel any better?

Reviewer 4: Alternative names are noted in [`8523ed4`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/8523ed416fbe3db09cf3031cacf4feb3f1be2ee1)

The server MAY round the delay up to a maximum of 30 seconds away from the request.

If the requested delay exceeds the maximum delay defined by the homeserver, the server will respond with HTTP 400,
a new `M_MAX_DELAY_EXCEEDED` error code, and the maximum allowed delay in milliseconds in a new `max_delay` field.

For example, the following specifies a maximum delay of 24 hours:

```http
400 Bad Request
Content-Type: application/json

{
  "errcode": "M_MAX_DELAY_EXCEEDED",
  "error": "The requested delay exceeds the allowed maximum.",
  "max_delay": 86400000
}
```

The homeserver **should** apply rate limiting to the scheduling of delayed events to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.16/appendices/#threat-high-volume-of-messages) threat.

If the user has too many outstanding delayed events, the server will respond with HTTP 400 and a new
`M_MAX_DELAYED_EVENTS_EXCEEDED` error code:

```http
400 Bad Request
Content-Type: application/json

{
  "errcode": "M_MAX_DELAYED_EVENTS_EXCEEDED",
  "error": "The maximum number of delayed events has been reached.",
}
```

### Managing delayed events

> A new authenticated client-server API endpoint at `POST /_matrix/client/v1/delayed_events/{delay_id}` allows scheduled events

Reviewer 18: To throw this on its head a bit, could we make this *unauthenticated* and instead make the delay_id a more secure random string. I have a fairly strong use case:

If I am building a RTC application and I want a backend service to be in control of whether one of my events expires, then I can pass my delayed event to the server and it can control the disconnect event on my behalf. Because the `delay_id` is scoped to one event that the user has already consented to sending, the auth doesn't really add much if the delay_id isn't trivally guessable.

The alternative to this is the backend service would need to get either my full token or a scoped access token (via as an as-of-yet unspec'd MSC) which adds some overhead.

Reviewer 18: We discussed this out of band and decided it was a good idea, right @fkwp ?

  - `event_id` - Optional. The `event_id` this event got in case it was sent.
  - `origin_server_ts` - Required. The timestamp of when the event was finalised.
- `next_batch` - Optional. A token that can be used to paginate the list of delayed events.

The batch size and the amount of terminated events that stay on the homeserver can be chosen, by the homeserver.
The recommended values are:

- batch size: 10
- `finalised` retention: 7 days
- `finalised` max cached events per user: 1000

There is no guarantee for a client that all events will be available in the
finalised events list if they exceed the limits of their homeserver.
Additionally, a homeserver may discard finalised delayed events that have been returned by a
`GET /_matrix/client/v1/delayed_events?status=finalised` response.

The homeserver **should** apply rate limiting to the `finalised` and `scheduled` delayed events `GET` endpoints.
Both most likely require (dependent on the implementation) serialization steps and can be used to slow down the server.

An example for a response to the `GET /_matrix/client/v1/delayed_events?status=scheduled` endpoint:

```http
200 OK
Content-Type: application/json

{

>   "scheduled": [

Reviewer 2: e2af939e

That deviation (among others) is being addressed by https://github.com/element-hq/synapse/pull/19038.

{
  "application": "m.call",
  "call_id": "",
  ...
}
```

Before sending the join event, it also schedules a delayed "hangup" state event with `delay` of around 5-20 seconds that
marks the end of its participation:

```http
PUT /_matrix/client/v1/rooms/!wherever:example.com/state/m.rtc.member/@someone:example.com?delay=10000
Content-Type: application/json

{
  "application": "m.call",
  "call_id": "",
  ...
}
```

Let's say the homeserver returns a `delay_id` of `1234567890`.

The client then periodically sends a "heartbeat" in the form of a "restart" of the delayed "hangup" state event to keep
the call membership "alive".

For example it could make the request every 5 seconds (or some other period less than the `delay`):

```http
POST /_matrix/client/v1/delayed_events/1234567890
Content-Type: application/json

{
  "action": "restart"
}
```

This would have the effect that if the homeserver does not receive a "heartbeat" from the client for 10 seconds, then
it will automatically send the "hangup" state event for the client.

Since the delayed event is sent first, a client can guarantee (at the time they are sending
the join event) that it will eventually leave.

> ### Self-destructing messages

Reviewer 19: This is a really good use case, but I see a fundamental problem in this MSC: Delayed events are not passed down the sync. So even if one device sends a self destruction, other devices of the same account would not be notified about it in the sync. Therefore other devices would always need to poll `delayed_events`, which breaks the concept of having the Matrix sync and would unnecessary flood the homeserver with requests.

  It might be important to have the guarantee that the delayed redact is received
  by the server at the time where the original message is sent.
- In the case of a state event, a user might want to set the state to `A` and after a
  timeout change it back to `{}`. By using two separate requests, sending `A` could work,
  but the event with content `{}` could fail. The state would not automatically
  reset to `{}`.

For this use case, batch sending of multiple delayed events would be desired.

Batch sending is not included in the proposal of this MSC however since batch sending should
become a generic Matrix concept as proposed with `/send_pdus`. (see: [MSC4080: Cryptographic Identities](https://github.com/matrix-org/matrix-spec-proposals/pull/4080))

[MSC2716: Incrementally importing history into existing rooms](https://github.com/matrix-org/matrix-spec-proposals/pull/2716)
already proposes a `batch_send` endpoint. However, it is limited to application services and focuses on historic
data. Since the additional capability to use a template `event_id` parameter is also needed,
this probably is not a good fit.

> ### Not reusing the `send`/`state` endpoint

Author: Its about the send body bein equivalent in both requests.
Sticky events also take the exact same appraoch.

I think the opinion on this very much depends on your mental model of delayed events
 - they are normal event sending actions but with a configurable increase in latency (it will take the hs a little bit to send the event anyways, you can just further delay this manually)
 - This endpoint schedules sth that is not a matrix event yet. Its a new entity that eveutally becomes a matrix event.
 
 I like the first view on it since it makes it easier to justify how this is compatible with matrix and why this probably wont change anything fundamental (matrix already needs to be capable dealing with differen network delays)

> ### Not reusing the `send`/`state` endpoint

Reviewer 19: As an SDK developer, I prefer a type safe API. It would've been helpful to have this natively in Matrix instead making a workaround at the SDK level to get type-safety. At the end, I will provide a type safe API to the user anyway.

> ### Not reusing the `send`/`state` endpoint

Reviewer 2: Another point against reusing the `/send` & `/state` endpoints is that having the delay as a query parameter conflicts with the convention of query parameters being filters, not action options.

It also interferes with the transaction identifier, for [which the spec says](https://spec.matrix.org/v1.16/client-server-api/#transaction-identifiers):
> The homeserver should identify a request as a retransmission if the transaction ID is the same as a previous request, and the path of the HTTP request is the same.

If I'm reading that rule correctly, it means that a request with the same transaction ID as an earlier request, but with a different query string, is still treated as a retransmission.  That would make a `/send` or `/state` request with a `?delay=` potentially be treated as a retransmission of an earlier such request without a `?delay=`.

Using a dedicated endpoint could avoid this by specifying the delay as a top-level field of the request body, with the content of the delayed event moved down one level into an object:

```
PUT /_matrix/client/v1/rooms/{roomId}/send_delayed_event/{eventType}/{txnId}
{
  "delay": <delay_ms>
  "content": {
    "body": "hello",
    "msgtype": "m.text"
  }
}
```

- `action` - The action to take on the delayed event. Must be one of:
  - `send` - Send the delayed event immediately.
  - `cancel` - Cancel the delayed event so that it is never sent.
  - `restart` - Restart the timeout of the delayed event.

For example, the following would send the delayed event with delay ID `1234567890` immediately:

```http
POST /_matrix/client/v1/delayed_events/1234567890
Content-Type: application/json

{
  "action": "send"
}
```

Where the `action` is `send`, the homeserver SHOULD apply rate limiting to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.11/appendices/#threat-high-volume-of-messages) threat.

### Getting delayed events

A new authenticated client-server API endpoint `GET /_matrix/client/v1/delayed_events` allows clients to get a list of
all the delayed events owned by the requesting user that have been scheduled to send.

The endpoint accepts a query parameter `from` which is a token that can be used to paginate the list of delayed events as
per the [pagination convention](https://spec.matrix.org/v1.11/appendices/#pagination). The homeserver can choose a suitable

> page size.

Reviewer 2: This should also define how the response should order the returned delayed events, lest the paginated results are jumbled & difficult to navigate through.

I propose sorting in order of increasing send time, i.e. `running_since` + `delay`.

      "content":{
        "msgtype": "m.text",
        "body": "I am now offline"
      }
    },
    {
      "delay_id": "abcdefgh",
      "room_id": "!roomid:example.com",
      "type": "m.call.member",
      "state_key": "@user:example.com_DEVICEID",
      "delay": 5000,
      "running_since": 1721732853284,
      "content":{
        "memberships": []
      }
    }
  ],
  "next_batch": "b12345"
}
```

`running_since` is the timestamp (as unix time in milliseconds) when the delayed event was scheduled or last restarted.
So, unless the delayed event is updated beforehand, the event will be sent after `running_since` + `delay`.

This can be used by clients to display events that have been scheduled to be sent in the future.

For use cases where the existence of a delayed event is also of interest for other room members,
(e.g. self-destructing messages) it is recommended to include this information in the original/affected event itself.

### Homeserver implementation details

#### Power levels are evaluated at the point of sending

Power levels are evaluated for each event only once the delay has occurred and it will be distributed/inserted into the

> DAG. This implies a delayed event can fail if it violates power levels at the time the delay passes.

Reviewer 2: Should users be notified of delayed events that hit power level failures (or any 400-class error) upon being sent?

If so, perhaps delayed event send failures can be included in `/sync` responses, for clients to handle however they like.

If not, then this spec should make it explicit that delayed event send failures are effectively unhandled.

Author: That is a really good comment.
Adding it to the sync response Sounds as if we would need to introduce a place for error messages there. Which  be a very new concept I think.

What about adding it to the get delayed events endpoint.
We don't known if the client will be online if the failiour occurs so it is a very async check for errors anyways so making it part of a request does not sound too bad.
I don't know how to remove errors from the get delayed events list. Maybe as soon as they are fetched once they get removed?

Is that an issue if there are more clients getting futures?

Author: I added this into the GET endpoint.

Reviewer 20: Would an event failure be something that a client needs to act upon? If so, the client would need to continuously poll this endpoint, which is starting to feel a lot like `/sync`...

Author: > Would an event failure be something that a client needs to act upon? If so, the client would need to continuously poll this endpoint, which is starting to feel a lot like /sync...

A good client implementation would also know about its delayed events. If it is important to know the outcome they can anticipate the time when the finalized is available via the get endpoint and can get the data only if required for an immediate response.

and a body with a Matrix error code `M_MAX_DELAY_EXCEEDED` and the maximum allowed delay (`max_delay` in milliseconds).

For example, the following specifies a maximum delay of 24 hours:

```http
400 Bad Request
Content-Type: application/json

{
  "errcode": "M_MAX_DELAY_EXCEEDED",
  "error": "The requested delay exceeds the allowed maximum.",
  "max_delay": 86400000
}
```

The homeserver **should** apply rate limiting to the scheduling of delayed events to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.11/appendices/#threat-high-volume-of-messages) threat.

The homeserver **may** apply a limit on the maximum number of outstanding delayed events in which case the Matrix error code
`M_MAX_DELAYED_EVENTS_EXCEEDED` can be returned:

```http
400 Bad Request
Content-Type: application/json

{
  "errcode": "M_MAX_DELAYED_EVENTS_EXCEEDED",
  "error": "The maximum number of delayed events has been reached.",
}
```

### Managing delayed events

A new authenticated client-server API endpoint at `POST /_matrix/client/v1/delayed_events/{delay_id}` allows scheduled events

> to be managed.

Author: solved in: 96f9063cf51598ad69e15cdecb7338d701eec1c3

> to be managed.

Reviewer 6: > > Should there be a transaction ID on this endpoint? The case being a client tries to send the event, has a network issue which prevents them seeing a 200 OK, retries, gets error for "already sent".
> > 
> > * An alternative may be to have the endpoint return information about what happened instead of making it a transaction. For example, when send is called, return 200 OK with the event_id always, even on subsequent calls. Similarly for cancel: just a 200 OK all the time. restart could return the new estimated send time so a client can inspect it to realize that the event would be sent too late and therefore needs to cancel and schedule a new event.
> 
> I do not understand the solution here. How can cahnging the response help in the case where a user does not see the response due to a network issue.

Transaction IDs are used on `/send` to ensure the caller gets the same response for the same input, performing the action exactly once. Given this is a high traffic endpoint, potentially, it's similarly possible for a caller to get a *different* answer given the same input in a way where they might not know what went wrong. Usually in these cases, a transaction ID can help.

- `action` - The action to take on the delayed event. Must be one of:
  - `send` - Send the delayed event immediately.
  - `cancel` - Cancel the delayed event so that it is never sent.
  - `restart` - Restart the timeout of the delayed event.

For example, the following would send the delayed event with delay ID `1234567890` immediately:

```http
POST /_matrix/client/v1/delayed_events/1234567890
Content-Type: application/json

{
  "action": "send"
}
```

Where the `action` is `send`, the homeserver SHOULD apply rate limiting to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.11/appendices/#threat-high-volume-of-messages) threat.

### Getting delayed events

A new authenticated client-server API endpoint `GET /_matrix/client/v1/delayed_events` allows clients to get a list of
all the delayed events that

> have been scheduled to send.

Author: We should be more explicit here that its only the delayed events that have been scheduled by this user.
Currently this is only mentioned implicitly here:
> For use cases where the existence of a delayed event is also of interest for other room members,
(e.g. self-destructing messages) it is recommended to include this information in the original/affected event itself.

Reviewer 2: Good point.  caece4d6

#### On demand

New authenticated client-server API endpoints `GET /_matrix/client/v1/delayed_events/scheduled` and
`GET /_matrix/client/v1/delayed_events/finalised` allows clients to get a list of

> all the delayed events owned by the requesting user that have been scheduled to send, have been sent, or failed to be sent.

Author: (Again) summarizing thoughts here first:

> Can these endpoints be rate limited?

Yes those should be rate limited. The response can be larger and I expect the server to not cache the json so a serialization step will happen with each request.

> Are these endpoints available to guest accounts?

Yes matrixRTC might be interesting to be available for guest accounts -> delayed events should also be

> all the delayed events owned by the requesting user that have been scheduled to send, have been sent, or failed to be sent.

Reviewer 6: unresolving: guest access requirements aren't declared.

      "running_since": 1721732853284,
      "content":{
        "memberships": []
      }
    }
  ],
  "next_batch": "b12345"
}
```

`running_since` is the timestamp (as unix time in milliseconds) when the delayed event was scheduled or last restarted.
So, unless the delayed event is updated beforehand, the event will be sent after `running_since` + `delay`.

This can be used by clients to display events that have been scheduled to be sent in the future.

For use cases where the existence of a delayed event is also of interest for other room members,
(e.g. self-destructing messages) it is recommended to include this information in the original/affected event itself.

### Homeserver implementation details

#### Power levels are evaluated at the point of sending

Power levels are evaluated for each event only once the delay has occurred and it will be distributed/inserted into the
DAG. This implies a delayed event can fail if it violates power levels at the time the delay passes.

Conversely, it's also possible to successfully schedule an event that the user has no permission to at the time of sending
if the power level situation has changed at the time the delay passes.

> #### Delayed state events are cancelled by a more recent state event
> 
> If a new state event is sent to the same room with the same (event type, state key) pair as a delayed event, the delayed
> event is cancelled.
> 
> There is no race condition here since a possible race between timeout and the _new state event_ will always converge to
> the _new state event_:
> 
> - timeout for _delayed event_ followed by _new state event_: the room state will be updated twice: once by the content of
>   the delayed event but later with the content of _new state event_.
> - _new state event_ followed by timeout for _delayed event_: the _new state event_ will cancel the outstanding _delayed event_.
> 
> Note that this behaviour does not apply to regular (non-state) events as there is no concept of a `(type, state_key)`
> pair that could be overwritten.

Author: This might not be the best way to treat this. Here are my reasons why:
 - It is very easy to emulate this behavior by just cancelling the delayed event before sending a new state event.
 - In the matrixRTC spec I would like to argue that the order should be: send the delayed leave event FIRST and afterwards the active membership event. This guarantees that only expiring call memberships will be sent.
 - I would like to replace this section with:

 ```suggestion
#### State which is effected by delayed events will be marked in the unsigned field

If a state event is sent for a state key where there is already a delayed event scheduled the homeserver will add a:
`has_delayed_overwrite: true` property to the unsigned section of the state event.

This is not required if the delayed event is received after a normal state event.

This allows client to check if a state event is temporal: `has_delayed_overwrite: true`.
`m.call.member` events for instance will only be valid if they are temporal (It allows the receiving clients to evaluate if the sending client did not forget to schedule the delayed leave event.)
```

Reviewer 2: > * It is very easy to emulate this behavior by just cancelling the delayed event before sending a new state event.

This isn't true for state that can be set by other users, who don't have the option to cancel (or even know about) delayed state from someone else.

> * In the matrixRTC spec I would like to argue that the order should be: send the delayed leave event FIRST and afterwards the active membership event. This guarantees that only expiring call memberships will be sent.

This is a good idea, but I believe it will work regardless of what the behaviour is for overwritten delayed state.

Author: We concluded that it is fine as it is:
 - There are examples where this behavior is confusing:
     - Pinned messages: I want to pin message a in a week. Now it would override all pins done until that point in time.
 - its hard to come up with good example why we don't want the delay to get canceled if the state (with same key/type) changes
 - The main reaosn this was brought up is to make it more similar to what we do with room events.
      - Once can think of state sending of "applying a diff" since the diff depends on the previous state scheduling a diff is not possible once the initial value of the diff changes.

However, the `origin_server_ts` of the delayed event should be the time that the event is actually sent by the homeserver.

This is a general problem that arises with the introduction
of [Cryptographic Identities](https://github.com/matrix-org/matrix-spec-proposals/pull/4080).
A user can intentionally, or caused by network conditions, delay the signing and sending of an event.
A possible solution would be the introduction of a `signing_ts` (in the signed section) and keep the `origin_server_ts`
in the unsigned section.
Both are reasonable data points that clients might want to use.
This would solve issues related to delayed events since
it would make it transparent to clients, when an event was scheduled and when it was distributed over federation.

## Alternatives

### Delegating delayed events

It is useful for external services to also interact with delayed events. If a client disconnects, an external service can
be the best source to send the delayed event/"last will".

This is not covered in this MSC but could be realized with scoped access tokens.
A scoped token that only allows to interact with the `delayed_events` endpoint and only with a subset of `delay_id`s

> would be used.

Reviewer 6: it'd probably be scoped to being allowed to send certain event types without delay, so when the SFU determines the user has dropped, it can send the event itself at that time.

Author: Both work. The advantage here is that a scoped token to only manage already send and signed delayed events is "very" save.
If we allow the SFU to send staate events (even if scoped to one type and room) is already a bit much. The SFU also does not need to be a trusted instance (The MatrixRTC spec is designed so it would break but it would not be a vounerablility if the SFU is malicious. It cannot even inject abusive content into the call since the source of truth is the room state.)

Additionally state events should be signed (or even better encrypted) to really make MatrixRTC as secure as possible. Then the SFU could not send a signed leave event but the delayed delegation still would work.

Do we want this discussion to be evaluated in the MSC text?

### Not reusing the `send`/`state` endpoint

Alternatively, new endpoints could be introduced to not overload the `send` and `state` endpoint.
Those endpoints could be called:

`PUT /_matrix/client/v1/rooms/{roomId}/send_delayed_event/{eventType}/{txnId}?delay={delay_ms}`

`PUT /_matrix/client/v1/rooms/{roomId}/state_delayed_event/{eventType}/{stateKey}?delay={delay_ms}`

This would allow the response for the `send` and `state` endpoints to remain as they are currently,
and to have a different return type for the new `send_delayed_event` and `state_delayed_event` endpoints.

### Allocating the event ID at the point of scheduling the send

This was considered, but when sending a delayed event the `event_id` is not yet available:

The Matrix spec says that the `event_id` must use the [reference hash](https://spec.matrix.org/v1.10/rooms/v11/#event-ids)
which is [calculated from the fields](https://spec.matrix.org/v1.10/server-server-api/#calculating-the-reference-hash-for-an-event)
of an event including the `origin_server_timestamp` as defined in [this list](https://spec.matrix.org/v1.10/rooms/v11/#client-considerations)

> Since the `origin_server_timestamp` should be the timestamp the event has when entering the DAG (required for call
> duration computation), the `event_id` cannot be computed when using the `send` endpoint before the delayed event has resolved.

Reviewer 6: ```suggestion
Since the `origin_server_ts` may change due to re-scheduling the event's send time, the event ID cannot be relied upon as it would also change.
```

say that it is no longer a member.

This works well when the client is running and can send the state events as needed. However, if the client is not able to
communicate with the homeserver (e.g. the user closes the app or loses connection) the call state is not updated to say
that the participant has left.

The motivation for this MSC is to allow updating call member state events after the user disconnected by allowing to
schedule/delay/timeout/expire events in a generic way.

The ["reliability requirements for the room state"](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/matrixRTC/proposals/4143-matrix-rtc.md#reliability-requirements-for-the-room-state)
section of [MSC4143: MatrixRTC](https://github.com/matrix-org/matrix-spec-proposals/pull/4143) has more details on the
use case.

There are numerous possible solution to solve the call member event expiration. They are covered in detail
in the [Use case specific considerations/MatrixRTC](#use-case-specific-considerations) section, because they are not part
of this proposal.

The proposal here allows a Matrix client to schedule a "hangup" state event to be sent after a specified time period.

> The client can then periodically restarts the timer whilst it is running. If the client is no longer running or able

Reviewer 21: ```suggestion
The client can then periodically restart the timer whilst it is running. If the client is no longer running or able
```

Content-Type: application/json

{
  "action": "send"
}
```

Where the `action` is `send`, the homeserver SHOULD apply rate limiting to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.11/appendices/#threat-high-volume-of-messages) threat.

### Getting delayed events

A new authenticated client-server API endpoint `GET /_matrix/client/v1/delayed_events` allows clients to get a list of
all the delayed events that
have been scheduled to send.

The endpoint accepts a query parameter `from` which is a token that can be used to paginate the list of delayed events as
per the [pagination convention](https://spec.matrix.org/v1.11/appendices/#pagination). The homeserver can choose a suitable
page size.

The response is a JSON object containing the following fields:

- `delayed_events` - Required. An array of delayed events that have been scheduled to send.
  - `delay_id` - Required. The ID of the delayed event.
  - `room_id` - Required. The room ID of the delayed event.
  - `type` - Required. The event type of the delayed event.
  - `state_key` - Optional. The state key of the delayed event if it is a state event.

>   - `transaction_id` - Optional. The transaction ID of the delayed event if it is a regular (non-state) event.

Reviewer 2: What's the value in returning the transaction ID?  (Pinging @hughns who wrote [the commit that added it](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/772590f6bac61bd026e8a60dc055b76637cce8f7))

If it's to help clients distinguish between multiple delayed events for the same room + event type + content + send time, delay IDs are already unique and can be used for that purpose instead.

If it's to help clients remember which transaction IDs they already used for delayed event requests, AIUI clients are already supposed to keep track of the transaction IDs they used for each CS-API endpoint, making it redundant to return here.

So unless there's some other purpose for it, I propose removing the transaction ID from this response.

Reviewer 2: Thanks! Reverted by 883e6b5d

  the delayed event but later with the content of _new state event_.
- _new state event_ followed by timeout for _delayed event_: the _new state event_ will cancel the outstanding _delayed event_.

Note that this behaviour does not apply to regular (non-state) events as there is no concept of a `(type, state_key)`
pair that could be overwritten.

#### Rate-limiting at the point of sending

Further to the rate limiting of the API endpoints, the homeserver SHOULD apply rate limiting to the sending of delayed messages
at the point that they are entered into the DAG.

This is to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.11/appendices/#threat-high-volume-of-messages) threat where a malicious
actor could schedule a large volume of events ahead of time without exceeding a rate limit on the initial `PUT` request,
but has specified a `delay` that corresponds to a common point of time in the future.

A limit on the maximum number of delayed events that can be outstanding at one time could also provide some mitigation against
this attack.

> ## Use case specific considerations

Reviewer 20: This feature could also be useful for temporary mutes/bans/kicks, temporary invites, etc.

Author: Added this to the "Use case specific considerations" section.

### MatrixRTC

In this section an overview is given how this MSC is used in [MSC4143: MatrixRTC](https://github.com/matrix-org/matrix-spec-proposals/pull/4143)
and alternative expiration systems are evaluated.

#### Background

MatrixRTC makes it necessary to have real time information about the current MatrixRTC session.
To properly display room tiles and header in the room list (or compute a list of ongoing calls) need to know:

- If there is a running session.
- What type that session has.
- Who and how many people are currently participating.

A particular delicate situation is that clients are not able to inform others if they lose connection.
There are numerous approaches to solve such a situation. They split into two categories:

- Polling based
  - Ask the users if they are still connected.
  - Ask an RTC backend (SFU) who is connected.
- Timeout based

  - Update the room state every x seconds.
    This allows clients to check how long an event has not been updated and ignore it if it's expired.

>   - Use Future events with a 10s timeout to send the disconnected from call

Reviewer 21: references to futures & /update_future are presumably stale now?

The Client-Server API currently has a  [Voice over IP module](https://spec.matrix.org/v1.11/client-server-api/#voice-over-ip)
that uses room messages to communicate the call state. However, it only allows for calls with two participants.

[MSC3401: Native Group VoIP Signalling](https://github.com/matrix-org/matrix-spec-proposals/pull/3401) proposes a scheme
that allows for more than two participants by using room state events.

In this arrangement each device signals its participant in a call by sending a state event that represents the device's
"membership" of a call. Once the device is no longer in the call, it sends a new state event to update the call state and
communicate that the device is no longer a member.

This works well when the client is running and can send the state events as needed. However, if the client is not able to
communicate with the homeserver (e.g. the user closes the app or loses connection) the call state is not updated to say
that the participant has left.

The motivation for this MSC is to allow updating call member state events after the user disconnected by allowing to
schedule/delay/timeout/expire events in a generic way.

> The ["reliability requirements for the room state"](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/matrixRTC/proposals/4143-matrix-rtc.md#reliability-requirements-for-the-room-state)

Reviewer 6: editorial: this link will break if MSC4143 gets accepted. Citing by section name, without link, and still continuing to link to the PR later in the sentence is recommended.

The homeserver will automatically send the "hangup" if it does not receive a "heartbeat".

## Proposal

The following operations are added to the client-server API:

- Schedule an event to be sent at a later time
- Get a list of delayed events
- Restart the timer of a delayed event
- Send the delayed event immediately
- Cancel a delayed event so that it is never sent

At the point of an event being scheduled the homeserver is [unable to allocate the event ID](#allocating-the-event-id-at-the-point-of-scheduling-the-send).
Instead, the homeserver allocates a `delay_id` to the scheduled event which is used during the above API operations.

### Scheduling a delayed event

An optional `delay` query parameter is added to the existing
[`PUT /_matrix/client/v3/rooms/{roomId}/state/{eventType}/{stateKey}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
and
[`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidstateeventtypestatekey)
endpoints.

The new query parameter is used to configure the event scheduling:

> - `delay` - Optional number of milliseconds the homeserver should wait before sending the event. If no `delay` is provided,
> the event is sent immediately as normal.

Reviewer 6: ```suggestion
- `delay` - Optional number of positive non-zero milliseconds the homeserver should wait before sending the event. If no `delay` is provided,
the event is sent immediately as normal.
```

> the event is sent immediately as normal.

Reviewer 6: Servers should also be able to round the delay to buckets, as otherwise the server needs to run timers with millisecond accuracy, which isn't possible at scale. 

Language-wise such a limitation would be something like "The server MAY round the delay up to a maximum of 30 seconds away from the request".

The body of the request is the same as it is currently.

If a `delay` is provided, the homeserver schedules the event to be sent with the specified delay and responds with an

> opaque `delay_id` field (omitting the `event_id` as it is not available):

Reviewer 2: Given that `delay_id`s aren't federated & are generated by the server, there doesn't need to be much restriction on their grammar.  This should be satisfactory: ba78cde1

> opaque `delay_id` field (omitting the `event_id` as it is not available):

Reviewer 8: The spec does have a standard opaque ID grammar too: https://spec.matrix.org/v1.16/appendices/#opaque-identifiers

> opaque `delay_id` field (omitting the `event_id` as it is not available):

Reviewer 2: Thanks, I somehow missed that. 6afd0f3a

and a body with a new Matrix error code `M_MAX_DELAY_EXCEEDED` and the maximum allowed delay (`max_delay` in milliseconds).

For example, the following specifies a maximum delay of 24 hours:

```http
400 Bad Request
Content-Type: application/json

{
  "errcode": "M_MAX_DELAY_EXCEEDED",
  "error": "The requested delay exceeds the allowed maximum.",
  "max_delay": 86400000
}
```

The homeserver **should** apply rate limiting to the scheduling of delayed events to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.11/appendices/#threat-high-volume-of-messages) threat.

The homeserver **may** apply a limit on the maximum number of
outstanding delayed events in which case a new Matrix error code
`M_MAX_DELAYED_EVENTS_EXCEEDED` can be returned:

```http
400 Bad Request
Content-Type: application/json

{
  "errcode": "M_MAX_DELAYED_EVENTS_EXCEEDED",
  "error": "The maximum number of delayed events has been reached.",
}
```

### Managing delayed events

A new authenticated client-server API endpoint at `POST /_matrix/client/v1/delayed_events/{delay_id}` allows scheduled events

> to be managed.

Reviewer 6: Guest access requirements are not declared in the MSC.

The body of the request is a JSON object containing the following fields:

- `action` - The action to take on the delayed event.\
Must be one of:
  - `send` - Send the delayed event immediately instead of waiting for the delay time to be reached.
  - `cancel` - Cancel the delayed event so that it is never sent.
  - `restart` - Reset the send time to be `now + original_delay`.

For example, the following would send the delayed event with delay ID `1234567890` immediately:

```http
POST /_matrix/client/v1/delayed_events/1234567890
Content-Type: application/json

{
  "action": "send"
}
```

Where the `action` is `send`, the homeserver **should** apply rate limiting to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.11/appendices/#threat-high-volume-of-messages) threat.

Once a delayed event has been sent or canceled, it is finalized and cannot be
accessed via this endpoint.

> The server will return a `M_NOT_FOUND` error.

Reviewer 6: ... in what circumstances?

Reviewer 8: I assume it means when the delay ID is unknown, but not 100% certain

Managing the delayed events does not require rate limits.

> The rate limits will be applied when evets are send after the delay times out (or when the event is send by calling `send`).

Reviewer 6: It feels important to allow this endpoint to be rate limited. The MSC doesn't have to (and probably shouldn't) specify what that rate limit is, but having the ability to prevent DoS attacks is important to a server.

If the user has scheduled a large amount of delayed events they can cancel all of them without rate limiting,

> because scheduling of the events itself is rate limited this is not causing issues.

Reviewer 2: This has since been replaced by:

> The primary point of rate limiting is event sending when the delay times out or the event is sent using the `send` action. However, servers can choose to rate limit the the management endpoints themselves as well if necessary.

i.e. the old text said to not apply rate limiting on bulk cancellations, whereas the new text says that even cancelling may be rate limited at the server's discretion.

Implementers should make cancelling delayed events similarly

> expensive as sending a rate limit error response.

Reviewer 6: This may not be possible in some server implementations, so we cannot specify it as a requirement.

Reviewer 8: I'm not even sure what that sentence is trying to say, so I just removed it while rewording the whole paragraph to allow ratelimiting

### Getting delayed events

> New authenticated client-server API endpoints `GET /_matrix/client/v1/delayed_events/scheduled` and

Reviewer 2: Good idea.  Since "state" is already a meaningful term in the spec, how about naming the query parameter "status"? 2afcebba

    - `delay_id` - Required. The ID of the delayed event.
    - `room_id` - Required. The room ID of the delayed event.
    - `type` - Required. The event type of the delayed event.
    - `state_key` - Optional. The state key of the delayed event if it is a state event.
    - `delay` - Required. The delay in milliseconds before the event is to be sent.
    - `running_since` - Required. The timestamp (as Unix time in milliseconds) when the delayed event was scheduled or
      last restarted.
    - `content` - Required. The content of the delayed event. This is the body of the original `PUT` request, not a preview
      of the full event after sending.
  - `next_batch` - Optional. A token that can be used to paginate the list of delayed events.

- For the `GET /_matrix/client/v1/delayed_events/finalised` endpoint:
  - `finalised` - Required. An array of finalised delayed events, that have either been sent or resulted in an error,
  sorted by `origin_server_ts` in decreasing order (latest finalised event first).
    - `delayed_event` - Required. Describes the original delayed event in the same format as the `delayed_events` array.

>     - `outcome`: `"send"|"cancel"`

Reviewer 2: 10361658 makes this use `"cancel"` for errors or intentional cancellations, with `reason` clarifying which of the two happened.

DAG. This implies a delayed event can fail if it violates power levels at the time the delay passes.

Conversely, it's also possible to successfully schedule an event that the user has no permission to send at the time of sending.
If the power level situation has changed at the time the delay passes, the event can even reach the DAG.

#### Delayed state events are cancelled by a more recent state event

> [!NOTE]
> Special rule for delayed state events:
> A delayed event `D` gets cancelled if:
>
> - `D` is a state event with key `k` and type `t` from sender `s`.
> - A new state event `N` with type `t` and key `k` is sent into the room.
> - The sender of `D` is different to the sender `N`.

If a new state event is sent to the same room at the same entry (`event_type`, `state_key` pair) as a delayed event by a
**different matrix user**, any delayed event for this entry (`event_type`, `state_key` pair) is cancelled.

This only happens if its a state update from a different user. If it is from the same user, the delayed event will not
get cancelled.
If the same user is updating the state which has associated delayed events, this user is in control of those delayed events.

> They can just cancel and check the events manually using the `/delayed_events` and the `/delayed_events/scheduled` endpoint.

Reviewer 14: > That would require the lookup endpoint to also return the device ID of the original request

Why so? Isn't it up to the server to decide whether or not to cancel the event, and it doesn't need to go through the lookup endpoint to do so.

> Besides, for state events that are associated with sending device, the sending device ID would (or should) be part of the state key, which clients could use to decide to cancel scheduled state events for the same/different device ID.

The concern is more about state events that are *not* associated with a specific device.

To be honest with you, this whole "Delayed state events are cancelled by a more recent state event" mechanism feels a bit like a solution looking for an application. It's not clear to me that we've given much thought to exactly what the semantics should be. I'd be in favour of removing this bit for now, until it's clearer what we want to happen.

> They can just cancel and check the events manually using the `/delayed_events` and the `/delayed_events/scheduled` endpoint.

Reviewer 2: 3efecea1

In the case where the delayed event gets cancelled due to a different user updating the same state, there
is no race condition here since a possible race between timeout and the _new state event_ will always converge to
the _new state event_:

- timeout for _delayed event_ followed by _new state event_: the room state will be updated twice: once by the content of
  the delayed event but later with the content of _new state event_.

> - _new state event_ followed by timeout for _delayed event_: the _new state event_ will cancel the outstanding _delayed event_.

Reviewer 14: Yes. So the _delayed event_ could end up "winning". It's unclear whether or not that is a problem in practice, but basically your assertion that "there is no race condition here" seems incorrect.

### Use `DELETE` HTTP method for `cancel` action

Instead of providing a `cancel` action for delayed events, the client could send a `DELETE` request to the same endpoint.

This feels more elegant, but it doesn't feel like a good suggestion for how the other actions are mapped.

### [Ab]use typing notifications

Some exploration of using typing notifications to indicate that a user is still connected to a call was done.

The idea of extending [MSC3038: Typed typing notifications](https://github.com/matrix-org/matrix-spec-proposals/pull/3038)
to allow for additional meta data (like device ID and call ID) was considered.

A perceived benefit was that if the delay events were federated, then the typing notification EDUs might provide an
efficient transport.

However, as the conclusion was to [not federate the delayed events](#federated-delayed-events), this approach was
discounted in favour of a dedicated endpoint.

### Alternative to `running_since` field

Some alternatives for the `running_since` field on the `GET` response are:

- `delaying_from`
- `delayed_since`
- `delaying_since`

> - `last_restart` - but this feels less clear than `running_since` for a delayed event that hasn't been restarted

Reviewer 2: Surprised that it's is absent.  Here it is now: 8c5b612f

> - `last_restart` - but this feels less clear than `running_since` for a delayed event that hasn't been restarted

Reviewer 2: There are no other specced fields ending with `_time`, but plenty with `_ts`.  8fee4cfe renames this to `send_ts`.

## Security considerations

All new endpoints are authenticated.

To mitigate the risk of users flooding the delayed events database, servers **must** make it possible to configure
the values for the maximum amount and the maximum timeout value for scheduled delayed events.

> (When the server returns the `M_MAX_DELAYED_EVENTS_EXCEEDED` and the `M_MAX_DELAY_EXCEEDED` error)

Reviewer 6: We generally shy away from mandating specific implementation like "config". Instead, we simply say that servers MUST impose maximum values - how they do that is their concern.

  - `cancel` - Cancel the delayed event so that it is never sent.
  - `restart` - Reset the send time to be `now + original_delay`.

For example, the following would send the delayed event with delay ID `1234567890` immediately:

```http
POST /_matrix/client/v1/delayed_events/1234567890
Content-Type: application/json

{
  "action": "send"
}
```

Where the `action` is `send`, the homeserver **should** apply rate limiting to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.11/appendices/#threat-high-volume-of-messages) threat.

The server will return a `M_NOT_FOUND` error if the `delay_id` is not recognized or was already canceled.

The success response body is always an empty object. A future MSC may define additional keys, such as returning
the event ID for `send` or the new expected send time for `restart`.

To allow safely retrying requests, the `send` action should return HTTP 200 even if the event was already sent
successfully by a previous request or by the delay timing out.

If sending the event via a `send` call fails, or if the event already failed previously, the endpoint returns the error

> that happened while sending the event (e.g. `M_FORBIDDEN` if the user doesn't have permission to send the event).

Reviewer 8: This part probably hasn't been implemented because I made it up just now (the error part might have been implemented, but send 200 on retry probably hasn't)

Reviewer 2: Should `cancel` always return 200 on retry as well?
https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r2382210026
> `cancel` can turn into a 404 which is also fine

What makes it fine, while it was not fine for `send`?

Reviewer 2: Actually, to further emphasize that retries are safe, how about using `PUT` for the `/send` and `/cancel` endpoints, instead of `POST`?

`/restart` would still have to use `POST` since it is not idempotent (as repeated requests push the send time further into the future).

Reviewer 2: This is now covered by 1e18fa2.

the client could send a `DELETE` request to an endpoint representing a target delayed event.

This feels more elegant, but it doesn't feel like a good suggestion for how the other actions are mapped.
Also, `DELETE` suggests that the target resource will be truly deleted, but this is at odds with how
cancelling a delayed event has it retained as a finalised event for later lookup.

### Alternative to `running_since` field

Some alternatives for the `running_since` field on the `GET` response are:

- `scheduled_ts` - clearly designates the field as a timestamp due to its suffix of `ts`,
  but might be interpreted as the scheduled send time
- `created_ts` - also clearly a timestamp, but no other part of this proposal uses the term "created"
- `delaying_from` - `delaying` might be clearer than `running`, and `from` might be clearer than `since`
- `delayed_since` - using past tense might better convey that this time is in the past
- `delaying_since` - `since` might be a clearer suffix than `from`
- `last_restart` - but this feels less clear than `running_since` for a delayed event that hasn't been restarted
- `send_time` - with a value of `delay` + the start time of the timer

> ### Syncing failed delayed events

Reviewer 19: Sorry to bring that topic up again.

What exactly is "complicated" about this? From an SDK perspecive re-using the _long_ polling based sync loop instead of a _short_ polling delayed events loop sounds easier and more efficient to do. Usually a sync loop already has all the fail-safety mechanisms implemented, which needs to be re-implemented with the current approach in the MSC making the non-sync approach more complicated. Not having it in a sync also sort of breaks the current concept of the spec, that a sync usually contains everything needed.

Additionally, from a non-MatrixRTC perspecitve ("a wide range of possible
applications such as tea timers, reminders, self-destructing messages,...") getting all (not only failed) delayed events via sync loop has way better UX in most of these use cases with a multi-device setup.

Reviewer 2: `/sync` would indeed be ideal, but the concern is that including it now would slow down the approval process & implementations of this MSC, especially since `/sync` support isn't strictly necessary for delayed events to be usable, and adding it later wouldn't conflict with any existing part of this MSC as currently written.  The intent is to eventually have a separate MSC for it.

I've put these thoughts in the MSC itself with e2a3dc34.

Reviewer 6: (+1 to pushing syncing failed delayed events out to its own MSC if/when needed rather than build it up front)

## Proposal

The following operations are added to the client-server API:

- Schedule an event to be sent at a later time
- Get a list of delayed events
- Restart the timer of a delayed event
- Send the delayed event immediately
- Cancel a delayed event so that it is never sent

At the point of an event being scheduled the homeserver is [unable to allocate the event ID](#allocating-the-event-id-at-the-point-of-scheduling-the-send).
Instead, the homeserver allocates a _delay ID_ to the scheduled event which is used during the above API operations.

### Scheduling a delayed event

An optional `delay` query parameter is added to the existing
[`PUT /_matrix/client/v3/rooms/{roomId}/state/{eventType}/{stateKey}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
and
[`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidstateeventtypestatekey)
endpoints.

The new query parameter is used to configure the event scheduling:

- `delay` - Optional number of milliseconds the homeserver should wait before sending the event. If no `delay` is provided,

> the event is sent immediately as normal.

Reviewer 22: Would it be possible to support setting the absolute target time rather than relative?
I was thinking of possibly using this as a "hack" to make sure a bunch of state event changes are batched in a single federation transaction by scheduling them at a single point in time ~5-10 seconds in the future. (eg. a moderation bot doing wildcard kicks/bans, or redacting tens if not hundreds of spam messages, this would allow for far higher event throughput and reduce federation overhead, where the bot may already be exempted from rate limits)

Just a thought, no intent to block this MSC.

Reviewer 2: The federation overhead should be the same, because the same amount of content (in your case, the moderation actions) will be sent in the request whether they're delayed or immediate events.

Unless you meant something else?

Instead, the homeserver allocates a `delay_id` to the scheduled event which is used during the above API operations.

### Scheduling a delayed event

An optional `delay` query parameter is added to the existing
[`PUT /_matrix/client/v3/rooms/{roomId}/state/{eventType}/{stateKey}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
and
[`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidstateeventtypestatekey)
endpoints.

The new query parameter is used to configure the event scheduling:

- `delay` - Optional number of positive non-zero milliseconds the homeserver should wait before sending the event. If
  no `delay` is provided, the event is sent immediately as normal.

The body of the request is the same as it is currently.

If a `delay` is provided, the homeserver schedules the event to be sent with the specified delay and responds with an
opaque `delay_id` field (omitting the `event_id` as it is not available):

```http
200 OK
Content-Type: application/json

{
  "delay_id": "1234567890"
}
```

> The `delay_id` is generated by the server. It SHOULD contain only valid non-surrogate Unicode codepoints except `:` and `NUL`, and MUST NOT exceed 255 bytes.

Reviewer 14: why excluding :, ooi?

255 bytes rather than characters, so emojis count as multiple bytes?

Honestly I'd suggest just using the opaque identifier grammar for consistency and to save thinking about these things

Reviewer 2: I just copied this part from the grammar for room aliases. The opaque identifier grammar is a much better fit, though. 6afd0f3a

The server will respond with HTTP 404 and an `M_NOT_FOUND` error if the `delay_id` is not recognized or was already canceled.

The success response body is always an empty object. A future MSC may define additional keys, such as returning
the event ID for `send` or the new expected send time for `restart`.

To allow safely retrying requests, the `send` action should return HTTP 200 even if the event was already sent
successfully by a previous request or by the delay timing out.

If sending the event via a `send` call fails, or if the event already failed previously, the endpoint returns the error
that happened while sending the event (e.g. HTTP 403 and `M_FORBIDDEN` if the user doesn't have permission to send the event).

The primary point of rate limiting is event sending when the delay times out or the event is sent using the `send`
action. However, servers can choose to rate limit the the management endpoints themselves as well if necessary.

### Getting delayed events

New authenticated client-server API endpoints `GET /_matrix/client/v1/delayed_events?status=scheduled` and

> `GET /_matrix/client/v1/delayed_events?status=finalised` allows clients to get a list of

Reviewer 2: Good point.  The default can be to return both scheduled & finalised events, and setting `status` can filter on one or the other.  This is also a good spot to filter on the `delay_id` to query for just a single event.

> `GET /_matrix/client/v1/delayed_events?status=finalised` allows clients to get a list of

Reviewer 2: 1359f2b6 & 59392f28

> `GET /_matrix/client/v1/delayed_events?status=finalised` allows clients to get a list of

Reviewer 2: > The default can be to return both scheduled & finalised events

On second thought, this complicates pagination, as it makes the response have to paginate two different item categories.

A possible solution is to take advantage of the fact that all scheduled delayed events have send times in the future, while finalised ones have send times in the past.  The first "page" of the response would include both scheduled & finalised delayed events whose send times are within an interval around when the lookup request was made, and would also return both a `next_batch` and `prev_batch` token for retrieving further scheduled or finalised delayed events, respectively.  Any "next" page would include only scheduled delayed events (since they are "next" in time) while any "previous" page would include only finalised ones (since they are "previous" in time).

If that is too much complexity, then there could simply be dedicated endpoints for looking up either scheduled delayed events or finalised ones.  This has the added benefit of each endpoint being able to support unique query parameters that are relevant only for their respective delayed event status (like filtering finalised delayed events by the value of their `outcome` property).

> [The combined endpoint] is also a good spot to filter on the `delay_id` to query for just a single event.

I'll walk back on this too, and will opt for having the `delay_id` as a path parameter [as was originally proposed](https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r2315944196) instead of a query parameter, because:
1. There is precedent in the CS-API for using a path parameter for the ID of a singular resource to look up.
2. Having a `delay_id` query parameter would suggest being able to specify it multiple times to look up multiple items, but this should have safeguards on how many items to allow looking up, to avoid absurdly long URLs or the need to paginate.  Getting that right is not really worth the effort, especially since a viable alternative to a multi-item lookup is to just do multiple single-item lookups.
3. Giving the `delay_id`-based lookup its own endpoint leaves room for it to be unauthenticated, just like the management endpoints now are, for the purpose of delegating control over specific delayed events to an external service.  If it shared the same endpoint as the status-based lookups, this would not be possible.

The only caveat is that if the status-based lookups were to be split into one endpoint per status, we'd be back [in the situation of needing to avoid a namespace clash](https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r2315944196) with a `delay_id`-based lookup endpoint at `GET /delayed_events/{delayId}`.  So, their paths could be something like `GET /delayed_events/by-status/[scheduled|finalised]` to satisfy that.

> `GET /_matrix/client/v1/delayed_events?status=finalised` allows clients to get a list of

Reviewer 2: cdc7a20 opts for having the multi-item lookup response put both scheduled & finalised events in the same array, which allows for straightforward pagination.

That commit also adds a single-item lookup, but doesn't make it unauthenticated as I had proposed in my last comment, as there is currently no use-case that requires external services to know the details (particularly the event content) of the delayed event(s) they are given control over the scheduling of.

Instead, the homeserver allocates a `delay_id` to the scheduled event which is used during the above API operations.

### Scheduling a delayed event

An optional `delay` query parameter is added to the existing
[`PUT /_matrix/client/v3/rooms/{roomId}/state/{eventType}/{stateKey}`](https://spec.matrix.org/v1.16/client-server-api/#put_matrixclientv3roomsroomidstateeventtypestatekey)
and
[`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](https://spec.matrix.org/v1.16/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
endpoints.

The new query parameter is used to configure the event scheduling:

- `delay` - Optional number of positive non-zero milliseconds the homeserver should wait before sending the event. If
  no `delay` is provided, the event is sent immediately as normal.

The body of the request is the same as it is currently.

If a `delay` is provided, the homeserver schedules the event to be sent with the specified delay and responds with an
opaque `delay_id` field (omitting the `event_id` as it is not available):

```http
200 OK
Content-Type: application/json

{
  "delay_id": "1234567890"
}
```

> The `delay_id` is an [opaque identifier](https://spec.matrix.org/v1.16/appendices/#opaque-identifiers) generated by the server.

Author: Resolved in: 49b200dcc11de286974925177b1e184cd905e6fa

per the [pagination convention](https://spec.matrix.org/v1.16/appendices/#pagination). The homeserver can choose a suitable
page size.

The response is a JSON object containing the following fields:

- `scheduled` - Required if the request has no `status` parameter, or sets it to `"scheduled"`.
An array of objects describing delayed events that have been scheduled to be sent,
sorted by `running_since + delay` in increasing order (event that will timeout soonest first).
These objects contain the following fields:
  - `delay_id` - Required. The ID of the delayed event.
  - `room_id` - Required. The room ID of the delayed event.
  - `type` - Required. The event type of the delayed event.
  - `state_key` - Optional. The state key of the delayed event if it is a state event.
  - `delay` - Required. The delay in milliseconds before the event is to be sent.
  - `running_since` - Required. The timestamp (as Unix time in milliseconds) when the delayed event was scheduled or
    last restarted.
  - `content` - Required. The content of the delayed event. This is the body of the original `PUT` request, not a preview
  of the full event after sending.

> - `finalised` - Required if the request has no `status` parameter, or sets it to `"finalised"`.

Reviewer 18: I think we agreed oob to drop finalised from this endpoint, so it only returns events due to be sent now. Perhaps for future proofing we can have `GET /_matrix/client/v1/delayed_events/scheduled`.

endpoint. It also makes use of [transaction identifiers](https://spec.matrix.org/v1.17/client-server-api/#transaction-identifiers).

The body for requests to this endpoint is a JSON object containing the following fields:

- `delay` - Required. A positive non-zero number of milliseconds the homeserver should wait before sending the event.
- `state_key` - Optional. The state key for the event to be sent, if it is to be a state event.
- `content` - Required. The content of the event to be sent.

The homeserver schedules the event to be sent with the specified delay and responds with an
opaque `delay_id` field (omitting the `event_id` as it is not available):

```http
200 OK
Content-Type: application/json

{
  "delay_id": "1234567890"
}
```

The `delay_id` is an [opaque identifier](https://spec.matrix.org/v1.17/appendices/#opaque-identifiers) generated by the server.
It MUST be globally unique and SHOULD be as difficult to guess/craft as an access token, so that knowledge of a `delay_id` may
securely grant control of it via [the management endpoint](#managing-delayed-events).

> The server MAY round the delay up to a maximum of 30 seconds away from the request.

Reviewer 18: I think this is saying that the server is allowed an accuracy of 30s?

This seems:

a) Difficult to enforce. What if the server is underpowered / unstable / has it's database fall out.
b) Probably next to useless if we're considering the RTC cases, as 30 seconds of delay is veeery slow for that kind of use case.

If the requested delay exceeds the maximum delay defined by the homeserver, the server will respond with HTTP 400,
a new `M_MAX_DELAY_EXCEEDED` error code, and the maximum allowed delay in milliseconds in a new `max_delay` field.

For example, the following specifies a maximum delay of 24 hours:

```http
400 Bad Request
Content-Type: application/json

{
  "errcode": "M_MAX_DELAY_EXCEEDED",
  "error": "The requested delay exceeds the allowed maximum.",
  "max_delay": 86400000
}
```

The homeserver **should** apply rate limiting to the scheduling of delayed events to provide mitigation against the
[High Volume of Messages](https://spec.matrix.org/v1.17/appendices/#threat-high-volume-of-messages) threat.

If the user has too many outstanding delayed events, the server will respond with HTTP 403 and the
`M_RESOURCE_LIMIT_EXCEEDED` error code:

```http
403 Forbidden
Content-Type: application/json

{

>   "errcode": "M_RESOURCE_LIMIT_EXCEEDED",

Reviewer 18: This is not the correct use of this status code. This is for when the homeserver has limits placed upon *it*.

As per https://spec.matrix.org/latest/client-server-api/#common-error-codes:

>The request cannot be completed because the homeserver has reached a resource limit imposed on it. For example, a homeserver held in a shared hosting environment may reach a resource limit if it starts using too much memory or disk space

This is not the same as the homeserver limiting individual users.

Naively can we not use a `M_LIMIT_EXCEEDED` with the delay being the time until the next event is sent?

Reviewer 18: Actually, great idea. It looks like this MSC is close to landing so we can probably rely on it.

If the target delayed event is already finalised with an outcome that conflicts with the `action`, i.e.
if the `action` is `send` or `restart` and the delayed event has already been cancelled,
or if the `action` is `cancel` and the delayed event has already been sent,
the homeserver will respond with HTTP 409
and a [standard error response](https://spec.matrix.org/latest/client-server-api/#standard-error-response)
with an `errcode` of `M_UNKNOWN`.

If the action is `send` and the delayed event is unable to be sent due to an error,
the homeserver will respond with that error (e.g. HTTP 403
and a [standard error response](https://spec.matrix.org/latest/client-server-api/#standard-error-response)
with an `errcode` of `M_FORBIDDEN` if the user doesn't have permission to send the event at the time of sending,
or HTTP 429 if the user has exceeded rate limits for sending room events at that time),
as if the request had been to send the event as a non-delayed event with either the `/send` or `/state` endpoint.
The homeserver SHOULD keep the delayed event scheduled, to account for the fact that the cause of the error
may resolve by the time of the delayed event's scheduled send time,

> and to allow retries of the `send` action until then.

Reviewer 17: I don't think the behavior of the `/send` endpoint should be different from the behavior of the event reaching it's scheduled send time. In case of error, the event should be finalized with an `error` outcome. Nothing else about the delayed event feature suggests that they should ever be retried in case of failure.

Additionally, with the behavior specified here, the client cannot know if the event is still scheduled or not after receiving an error response to a `send` request. The same error response could mean that the event has failed to be sent by the request itself (and is still scheduled) or has failed to send after it's timeout was reached (and is no longer scheduled).

Reviewer: Yeah, agreed. It would seem more consistent to me if we treated `/send` as an attempt to manually finalize the event – so that the event is either sent or cancelled afterwards.

Reviewer 2: The reason I added this section was to allow [external services](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/expiring-events-keep-alive/proposals/4140-delayed-events-futures.md#delegating-scheduled-delayed-events) to be able to retry `/send` attempts, especially for temporary errors like rate limits on messages sent into a room.

Otherwise, if the event that an external service tries to `/send` encounters an error, then the service can't simply retry sending/rescheduling that event, as the service won't have access to the event content to be re-sent (unless it was communicated with the service beforehand, or if `GET /delayed_events/{delay_id}` were to allow unauthed requests, which would sacrifice some privacy for convenience), and won't have permission to schedule/send a new event at all (unless it has an access token for doing so, in which case it would really just be a proper CS-API client and not an "external service" as this MSC defines it).

> Additionally, with the behavior specified here, the client cannot know if the event is still scheduled or not after receiving an error response to a `send` request. The same error response could mean that the event has failed to be sent by the request itself (and is still scheduled) or has failed to send after it's timeout was reached (and is no longer scheduled).

This is a good point.  A possible solution is to introduce a new error type for representing a `/send` request that was processed correctly by the server, but failed to send the event due to a CS-API error on the event itself; this error would wrap the standard error response of the underlying CS-API error in a dedicated field.  This would disambiguate from errors on the `/send` request itself, like a 404/409 for trying to send a delayed event that was not found / already cancelled / etc.

The error type would look something like this:
```json
{
  "errcode": "M_DELAYED_SEND_ERROR",
  "error": "Unable to send delayed event on-demand",
  "event_error": {
    "errcode": "M_LIMIT_EXCEEDED",
    "error": "Too many requests"
  } 
}
```

A limitation of this is that the wrapped error field can't wrap the HTTP status/headers of the CS-API error (such as [`Retry-After` for rate limits](https://spec.matrix.org/v1.18/client-server-api/#rate-limiting)), so I suppose the status/headers of `/send` would have to match those of the underlying CS-API error.

If all of this is too complicated, then an alternative is to propose that Matrix clients should have a communication channel with external services to be notified of when a service's `/send` attempt has failed & descheduled the target event, so it can possibly be retried.

Reviewer 17: If failing due to a temporary error is a concern, then shouldn't we also worry about retrying events which have reached their scheduled send time ?

To consider the initial use case. I have a scheduled event for disconnecting from a VOIP call, and I want to trust that it will get sent eventually after I stop refreshing it. If there exists a possibility that the event will fail due to a temporary error that calls for a retry (a rate-limit, for example), then we would like the event to get retried in that case, regardless of whether it gets sent by the `/send` endpoint or by reaching it's scheduled time. Maybe the server should re-schedule the scheduled event when it fails due to an error that calls for a retry ?

Conversely, if we decide that re-scheduling failed events isn't necessary when they get sent by timeout, then I would say that having the ability to send them again through the `/send` endpoint isn't necessary either.

Reviewer 2: > If failing due to a temporary error is a concern, then shouldn't we also worry about retrying events which have reached their scheduled send time ?

Yes, but the difference is that the scheduled send time of a delayed event is known (to a degree, considering restarts), but a `/send` request can be made at any time.

To check if a delayed event hit an error on its scheduled send time, a client can wait until that time and then call `GET /delayed_events/{delay_id}` to look for an `error` field on the response, and decide whether to retry sending/scheduling if there was an error.  In contrast, to check if a `/send` request descheduled its target event due to an error, a client must either proactively call the `GET` endpoint every so often, or instead wait until the event's original scheduled send time, at which point the event may have been cancelled for a long time already.

A way around that could be to notify clients (such as via a `/sync` payload) of failed `/send`s, so that they'll know right away when an event was unscheduled & can decide whether to retry sending it.  However, this would incur extra network traffic, and may not be available to "light" clients (like external services that only have access to `/send`, `/restart`, and `/cancel`, but not `/sync`, or even the ability to send/schedule a replacement event).

It also feels somewhat counter-intuitive for `/send` to behave like `/cancel` in the case of a temporary error.

> Maybe the server should re-schedule the scheduled event when it fails due to an error that calls for a retry ?

This could work, but I'd rather not offload more work to the server, especially for working around rate limits that are supposed to reduce server load.  I can also imagine situations where a retry isn't desired, so it makes more sense for clients to be in control of the decision to retry or not.

Reviewer 17: > To check if a delayed event hit an error on its scheduled send time, a client can wait until that time and then call GET /delayed_events/{delay_id} to look for an error field on the response, and decide whether to retry sending/scheduling if there was an error.

But isn't the use case for delayed events a situation where the client isn't available to send the event at the time ? Such as a client which has been disconnected and isn't there to hang-up a call. If we're counting on the client to be there after the event expired and check that it was sent successfully, then doesn't that defeat the whole purpose of a scheduled event in the first place ? If events that need to be retried are a concern, then we should also worry about the case of an event that needs to be retried after the client is no longer there to supervise it.

Reviewer 2: > But isn't the use case for delayed events a situation where the client isn't available to send the event at the time ? Such as a client which has been disconnected and isn't there to hang-up a call. If we're counting on the client to be there after the event expired and check that it was sent successfully, then doesn't that defeat the whole purpose of a scheduled event in the first place ?

This is a valid point.  The reality is that it is always possible for a scheduled event to become unsendable, even if the conditions at the time of its original scheduling would have allowed it to be sent.  This is what inspired the idea of finalised delayed events & allowing clients to check for failures after the fact; the only alternative is to let servers retry sending on failure.

> If events that need to be retried are a concern, then we should also worry about the case of an event that needs to be retried after the client is no longer there to supervise it.

My worry is that making the server retry failed scheduled events could be a futile effort, and may not be warranted depending on the kind of failure & the context in which it happened.  A valid case for retrying is rate limits, but a not-so-valid case is a permission issue, like the acting user no longer being in the room or having a power level that allows sending the event.  In those cases, automated retries would fail until the permission issue is resolved.

My view on this is, if a delayed event failed to be sent at its scheduled send time, nothing can fix how the event already missed its schedule, so clients can take their time (to a degree) to decide how to address the failure.  Meanwhile, if a `/send` hits a failure, that failure is known immediately & there is still time to address it before the event's scheduled send time, and should thus be as easy as possible to be addressed immediately.

The particular use-case I am considering is to allow Element Call to delegate management of delayed call-hangup events to the lk-jwt-service ([context](https://github.com/element-hq/lk-jwt-service/pull/171)).  When the service detects a dropped call, it may try `/send`ing the hangup event immediately instead of waiting for its scheduled send time ([context](https://github.com/element-hq/lk-jwt-service/pull/171/changes#diff-c48512963d0f5620e1d32f8c1e180dcc831e419f52ccea24a0d711130e65cc95R452-R455)).  What I want is for a failure on `/send` to not cancel the scheduled sending, because there could still be time for the scheduled send to succeed; and for the service to be able to retry the `/send` on failure, as the service has no other means of attempting a retry (as it can't query the homeserver for the event to be rescheduled/resent, nor can it hit any of the regular event-sending endpoints), and being able to retry is the easiest way for the service to send the hangup event as promptly as possible.  If all retried `/send`s fail & even the scheduled sending of the hangup event fails, then there is likely nothing that either the service or homeserver could do about it, and only then would a client have to clean up a lingering call state when it comes back online.

Reviewer 17: Okay I think I see the rationale. The event might be scheduled with a deliberately long timeout, counting on the external service to send it when it detects a hang-up. This gives the service plenty of time to retry it if needed.

Fair enough. It makes the API behavior a bit counterintuitive, but there's a good reason for it.

I still think that the need for retry when there isn't an external service to supervise the event is one worth worrying about. But I acknowledge that it doesn't necessarily have a good solution.

Reviewer: Resolving since – if I've followed correctly – it sounds like we acknowledged the current state of the proposal.

- `send_ts` - the intended scheduled send time (`running_since` + `delay`) of the delayed event.
  This is the ordering used when `order_by` is not specified.
- `origin_server_ts` - the time when the delayed event was finalised, or its scheduled send time if still scheduled.
- `running_since` - the time when the delayed event was scheduled or last restarted.

By default, delayed events are returned in increasing chronological order of the value specified by `order_by`.
To return results in reverse chronological order, the endpoint supports a query parameter of `dir=b`.
Providing a query parameter of `dir=f` uses the default of increasing chronological order.

The endpoint accepts a query parameter `from` which is a token that can be used to paginate the list of delayed events
as per the [pagination convention](https://spec.matrix.org/v1.18/appendices/#pagination).
The homeserver can choose a suitable page size.

By default, the first page of results starts from the time the request is made.
To have the first page start from a different point in time,
it may be specified in a query parameter of `since_ts`, expressed as a Unix timestamp in milliseconds,

> or as a special-case string of "end" to start from the latest possible results.

Reviewer 17: Shouldn't the option be "to start from the **earliest** possible results" ? If I understand correctly we're getting the events that come _after_ this timestamp.

It would also be good to clarify what timestamp is being looked to for the cutoff here. Is it the timestamp specified by the `order_by` parameter ?

Reviewer: > Shouldn't the option be "to start from the **earliest** possible results" ? If I understand correctly we're getting the events that come _after_ this timestamp.

Hm, I think it depends on whether you use `dir=f` or `dir=b`?

IIUC:

- `dir=f` and `since_ts=X` returns results whose scheduled time is `>=X` ordered by increasing scheduled time
- `dir=b` and `since_ts=X` returns results whose scheduled time is `<=X` ordered by decreasing scheduled time

So `dir=f&since_ts=end` gives you only the result(s) with the most distant scheduled time as you cannot go forward from there. But `dir=b&since_ts=end` returns all results starting from the most distant scheduled time and then ordered by decreasing scheduled time.

> It would also be good to clarify what timestamp is being looked to for the cutoff here. Is it the timestamp specified by the `order_by` parameter ?

Yeah, given that the results are ordered by what's specified in `order_by`, that's the only option that makes sense, I'd say.

Reviewer 17: Oh I see. If that's the intended meaning, it could maybe be clarified? With the way it's worded, I wasn't sure if `since_ts` indicates the start of the event in time (always return events after the timestamp) or the start of events in the list ordering (events selected depend on the order).

Reviewer: At least that's my interpretation of it. 🙈 But yeah, some clarification would help here.

On a tangential note, there's also maybe an [open question](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/changes#r3258882699) whether we really need all these query options. It looks like Synapse [doesn't currently implement any of them](https://github.com/element-hq/synapse/blob/4655b435ee6e83518cd9ea6941d403be1cdf3a63/synapse/rest/client/delayed_events.py#L149).

Reviewer 2: The interpretation in https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3301523622 is correct.  I'll try to reword this section for clarity.

Reviewer: Maybe this could work (contingent on https://github.com/matrix-org/matrix-spec-proposals/pull/4140/changes#r3333817315)?

```suggestion
or as a special-case string of "end" to start from the latest possible results. In other words:

- `dir=f&order_by=send_ts&since_ts=X` returns results whose scheduled time is `>= X` ordered by increasing scheduled time
- `dir=b&order_by=send_ts&since_ts=X` returns results whose scheduled time is `<= X` ordered by decreasing scheduled time

Thus, to paginate backwards from the latest available scheduled time, you would use `dir=b&order_by=send_ts&since_ts=end`.
```

Reviewer 2: IMO 88859bd does now make this endpoint simple enough & gives enough examples to not need extra clarification.

But for completeness, the changes made would alter your explanation to be the following:
- `dir=f&from_ts=X` returns results whose scheduled send time is `>= X` ordered by increasing scheduled time
- `dir=b&to_ts=X` returns results whose scheduled send time is `<= X` ordered by decreasing scheduled time
- To paginate backwards from the latest available scheduled time, you would use `dir=b` without `to_ts`.

Reviewer: Yes, agreed. This looks clear enough now to me. Thanks for the changes. 👍

A possible solution would be the introduction of a `signing_ts` (in the signed section) and keep the `origin_server_ts`
in the unsigned section.
Both are reasonable data points that clients might want to use.
This would solve issues related to delayed events since
it would make it transparent to clients, when an event was scheduled and when it was distributed over federation.

### Conflicting delayed state events

A delayed state event can overwrite other state events that were sent in between the delayed event
being scheduled and it being sent. Whether or not this is problematic strongly depends on the
use case, though. When the overwrite is undesired, a possible remedy could be to cancel the scheduled
delayed event when a conflicting new state event is sent into the room. Alternatively, it might be
possible to avoid the conflict in the first place by using separate `state_key`s or by not relying on
state events to begin with. Additionally, this type of race condition can also happen without delayed
events due to federation delay. Potentially addressing this situation is, therefore, left to a future
proposal.

> 

Reviewer: Capturing the finding from https://github.com/element-hq/element-call/issues/3985 as a potential issue.

```suggestion

### Rate-limiting for heartbeats

In office environments several clients might share the same public IP address. If the server
rate limits based on IP address and multiple clients use the heartbeat pattern (where a scheduled
delayed event is rescheduled recurringly), it is more likely for them to run into rate limit blocks.
This can result in their delayed event being triggered too early which can negatively affect application
logic. For instance, if the delayed event is a call disconnect event, triggering it too early will result
in unexpected disconnections from the call[^call-rate-limit].

[^call-rate-limit]: See also https://github.com/element-hq/element-call/issues/3985.

To mitigate this, the `/restart` endpoint SHOULD allow optional authentication so that the server
can rate limit per client when management of the delayed event is not delegated to an external service.

```

Author: I think this still collides with the delegation usecase. A JWT that handles 100dreds of calls will not use the authed endpoint and hence get rate limited easily.

I think the proper solution might be to rate limit per delay_token / delay_id. (looking at a delay_id like a scoped token, that fits the authed endpoint semantic where each token gets their own rate limit bucket.)
404 Not Found have to be a special case to mitigate denial of service attacks. If I send restarts for unknown delay events then we should fallback to IP based rate limiting. (In theory well behaving clients should never send restarts for unkown Id's)

Reviewer 2: > I think the proper solution might be to rate limit per delay_token / delay_id. (looking at a delay_id like a scoped token, that fits the authed endpoint semantic where each token gets their own rate limit bucket.)

This does feel like the best compromise currently available.  (Note that once OAuth 2.0 scopes are used for the management endpoints, the rate-limits may be applied against the scope tokens.)

> 404 Not Found have to be a special case to mitigate denial of service attacks. If I send restarts for unknown delay events then we should fallback to IP based rate limiting.

I'm not sure how this rate-limit could be applied.  By the time the server has done the work to find that a requested `delay_id`/`delay_token` is invalid, there is no more work to be done that a rate limit would protect against, so it doesn't really matter whether the error response is 404 or 429.  So I think the correct way to describe this rate-limiting strategy is to apply a bucket on each valid `delay_id`/`delay_token`, and no limit on invalid ones.

Either way, this leaves open the attack vector of a malicious client making a high number of restart/send/cancel requests for bogus `delay_id`/`delay_token`s.  However, servers should ideally index delayed events by their ID/token, so perhaps it wouldn't be terribly costly to handle such bogus requests.

Reviewer: How does this sound?

```suggestion

### Rate-limiting for heartbeats

In office environments several clients might share the same public IP address. If the server
rate limits based on IP address and multiple clients use the heartbeat pattern (where a scheduled
delayed event is rescheduled recurringly), it is more likely for them to run into rate limit blocks.
This can result in their delayed event being triggered too early which can negatively affect application
logic. For instance, if the delayed event is a call disconnect event, triggering it too early will result
in unexpected disconnections from the call[^call-rate-limit].

The same problem with rate limiting based on IP address also occurs when an external service
manages delayed events for a large number of users.

[^call-rate-limit]: See also https://github.com/element-hq/element-call/issues/3985.

To mitigate this, the server SHOULD rate limit the management endpoints based on the `delay_id`.
```

Reviewer 2: This is an accurate & approachable description of the issue.  I also like the idea of recommending a mitigation strategy with SHOULD instead of enforcing one with MUST, as any single approach has drawbacks (as we've seen), so servers should be able to choose whatever approach that works best for its usecase.  Plus, being flexible about it now will reduce friction against introducing an OAuth 2.0 scope for these endpoints, as it would allow enforcing a ratelimit on the scope token without having to replace a previously-specced ratelimit.

the scheduled send time of the next of the user's delayed events to be sent,
rounded up to the nearest second.

```http
429 Too Many Requests
Content-Type: application/json
Retry-After: 1200

{
  "errcode": "M_LIMIT_EXCEEDED",
  "error": "The maximum number of delayed events has been reached.",
}
```

#### Delayed event limits as a capability

The values of both the maximum allowed delay and the maximum allowed number of scheduled events are advertised as a
[capability](https://spec.matrix.org/v1.18/client-server-api/#capabilities-negotiation) named `m.delayed_events`, via
the values of fields named `max_delay` and `max_scheduled` respectively.
If the server doesn't enforce one of these limits, its representative field MUST be absent from the capability.
If the server enforces none of these limits, the capability MAY be omitted entirely instead of having an empty body.

For example, the following specifies a maximum allowed delay of 24 hours and a per-user limit of 10 delayed events:

```json
{
  "capabilities": {
    "m.delayed_events": {
      "max_delay": 86400000,
      "max_scheduled": 10

>     }

Reviewer: I _think_ the missing brace is what makes GitHub render the subsequent text all red.

```suggestion
    }
  }
```

- `content` - Required. The content of the event to be sent.

The homeserver schedules the event to be sent with the specified delay and responds with an
opaque `delay_id` field (omitting the `event_id` as it is not available):

```http
200 OK
Content-Type: application/json

{
  "delay_id": "1234567890"
}
```

The `delay_id` is an [opaque identifier](https://spec.matrix.org/v1.18/appendices/#opaque-identifiers)
generated by the homeserver.
It MUST be globally unique and SHOULD be cryptographically secure (in the sense that it is infeasible to predict).
This way, knowledge of the `delay_id` alone allows external services to control the delayed event through the
unauthenticated management endpoints introduced [below](#managing-delayed-events).

The homeserver MUST NOT send the event before the scheduled time.
To support batch sending, homeservers MAY add up to 30 seconds to the scheduled send time.
Note: clients might find that their events are delayed further due to server load and similar conditions.

The homeserver MAY enforce a maximum allowed delay for delayed events.
If a requested delay exceeds this maximum, the homeserver will respond with HTTP 400

> and a [standard error response](https://spec.matrix.org/latest/client-server-api/#standard-error-response)

Reviewer 6: `latest` can and does change, breaking historical links:
```suggestion
and a [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
```

#### Rate-limiting at the point of sending

Further to the rate limiting of the API endpoints, the homeserver SHOULD apply rate limiting to the sending
of delayed messages at the point that they are inserted into the DAG.

This is to provide mitigation against the [High Volume of Messages](
https://spec.matrix.org/v1.18/appendices/#threat-high-volume-of-messages) threat where a malicious actor
could schedule a large volume of events ahead of time without exceeding a rate limit on the initial `PUT` request,
but has specified a `delay` that corresponds to a common point of time in the future.

If a delayed event fails to be sent at its scheduled send time due to a rate limit failure,
the homeserver SHOULD NOT retry sending the event. Instead, the event will be stored as a finalised delayed event
with its `error` field set, available to be retrieved by a client for the user who requested the event.
It then becomes the user's responsibility to fetch this error and retry sending the event as appropriate.

### Guest accounts

All delayed event related endpoints are available to guest accounts.
This allows guest accounts to participate in MatrixRTC sessions.

> ## Use case specific considerations

Reviewer 6: I'd suggest removing this section. It has a few examples which look too close to actual proposals, so it makes this proposal even harder to review. At best, this should say "These MSCs use this functionality: [list]" and "Future MSCs, like self-destructing messages, might find benefit from this proposal" - anything more starts to risk review delays.

Reviewer: Attempting to address this with https://github.com/matrix-org/matrix-spec-proposals/pull/4478.

Reviewer: This landed.

A new authenticated Client-Server API endpoint is introduced for scheduling the sending of a message or state event:

`PUT /_matrix/client/v3/rooms/{roomId}/delayed_event/{eventType}/{txnId}`

The path parameters of this endpoint are the same as for the existing
[`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](
https://spec.matrix.org/v1.18/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
endpoint. It also makes use of [transaction identifiers](
https://spec.matrix.org/v1.18/client-server-api/#transaction-identifiers),
and supports [timestamp massaging](https://spec.matrix.org/latest/application-service-api/#timestamp-massaging)
when called by an application service.

The body for requests to this endpoint is a JSON object containing the following fields:

- `delay` - Required. A positive non-zero number of milliseconds the homeserver should wait before sending the event.
- `state_key` - The state key for the event to be sent, if it is to be a state event; absent otherwise.
- `content` - Required. The content of the event to be sent.

The homeserver schedules the event to be sent with the specified delay and responds with an

> opaque `delay_id` field (omitting the `event_id` as it is not available):

Reviewer: Given that the delay ID effectively acts as an access token for the management endpoints, we could consider changing the identifier to `delay_token`. This would more clearly call out to implementers that the ID / token is sensitive information. This will likely cause some compatibility hassle with the existing implementations, however.

Author: Can we evaluate in how many places this actually makes a difference (in the url for instance it does not make a difference since its not named there)

I think the main issue is the response to a `/send`/`/state`.
The default compatibility approach of sending both for some time seems very doable. The rust sdk and the js-sdk and synapse would be impacted but can be tackled one by one without coordination. (also LLM-able code imo. Should not introduce actual logic -> easy to review)

TLDR: The `cause some compatibility hassle` part seems to manageable and we can seriously consider this if we prefer it.

@AndrewFerr

Reviewer 2: I'm doubtful that the cost of changing the name across all clients/servers/SDKs is worth the effort, because:
- In the long run, access to delayed event endpoints / specific delayed events should be managed by OAuth 2.0 scopes, at which point a name of `delay_token` won't be as relevant
- Using `delay_id`s as an access token can't keep them as hidden as "real" access tokens (e.g. they appear in request URIs & `/sync` responses), so renaming them to `delay_token` suggests they adhere to a higher degree of security that they actually do

Reviewer: This sounds reasonable. 👍 Let's leave it as it is then.

```http
200 OK
Content-Type: application/json

{
  "delay_id": "1234567890"
}
```

The `delay_id` is an [opaque identifier](https://spec.matrix.org/v1.18/appendices/#opaque-identifiers)
generated by the homeserver.
It MUST be globally unique and SHOULD be as difficult to guess/craft as an access token, so that knowledge of a

> `delay_id` may securely grant control of it via [the management endpoints](#managing-delayed-events).

Reviewer: > [...] be as difficult to guess/craft as an access token [...]

I couldn't find any statements about the security of access tokens in the current spec. In fact, the spec seems to leave the format of access tokens [explicitly undefined](https://spec.matrix.org/v1.18/client-server-api/#client-authentication).

Maybe we could make this a little clearer by using the common term "cryptographically secure" rather than referring to "access tokens"?

```suggestion
It MUST be globally unique and SHOULD be cryptographically secure (in the sense that it is infeasible to predict).
This way, knowledge of the `delay_id` alone allows external services to control the delayed event through the
unauthenticated management endpoints introduced [below](#managing-delayed-events).
```

Author: I like this change. On top of this I would consider renaming `delay_id` -> `delay_token` as we have discussed briefly.

Reviewer: The suggestion has been applied separately and the potential renaming is being discussed in https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3264704128.

The homeserver MUST NOT send the event before the scheduled time.
To support batch sending, homeservers MAY add up to 30 seconds to the scheduled send time.
Note: clients might find that their events are delayed further due to server load and similar conditions.

> The homeserver MAY enforce a maximum allowed delay for delay events.

Reviewer: ```suggestion
The homeserver MAY enforce a maximum allowed delay for delayed events.
```

If a requested delay exceeds this maximum, the homeserver will respond with HTTP 400
and a [standard error response](https://spec.matrix.org/latest/client-server-api/#standard-error-response)

> with an `errcode` of `M_MAX_DELAY_EXCEEDED` and the maximum allowed delay in milliseconds in a `max_delay` field.

Reviewer: Instead of putting the maximum delay into the error response, maybe we could add a [capability](https://spec.matrix.org/v1.18/client-server-api/#capabilities-negotiation) or something akin to [` /_matrix/client/v1/media/config`](https://spec.matrix.org/v1.18/client-server-api/#get_matrixclientv1mediaconfig) (which coincidentally contains the maximum allowed upload size of media files)? This way clients would be able to discover the maximum delay (which might be zero if the server denies delayed events entirely) beforehand and adapt or disable certain parts of their UI.

The maximum allowed number of simultaneously scheduled events could possibly also go into such a capability / config.

Author: We need to consider if we want to ever allow a delay of 0.
It would break matrixRTC. (not entirely but the leave guarantee)
As this is like disabling the feature, we can either treat it as sliding sync (if a HS does not support it its incompatible) or like a "nice to have". If 0 is max allowed clients (and the jwt service) needs to work around it.

I am in favour of treating delayed events less as an optional "nice to have" but a necessity.
It should be the edgecase, that one exceeds the limit set by the HS. More like a rate limit than a file size limit.

But on the other hand an additional place to look up those configurations does not hurt in any way. Just not sure how important they are.

Reviewer: It's probably not terribly important. I think the main reason to do it now rather than later is that we could avoid the `max_delay` property in the error response and use the existing `M_INVALID_PARAM` instead.

Reviewer 2: Using a capability is a good idea.  An advantage it has over adding a new /config endpoint is that there's nothing so unique about delayed event limits that warrants being expressed with an additional endpoint.

It could take the form of:
```
"capabilities": {
  "m.delayed_events": {
    "max_delay": 10000,
    "max_scheduled": 100
  }
}
```

Reviewer 2: Done by 8f9f4544

timeout can be set to be long (e.g. 6 hours), as the SFU can be expected to not forget sending the
`POST /delayed_events/{delay_id}/send` requests when it detects a disconnecting client.

### Getting delayed events

A set of new authenticated Client-Server API endpoints allows clients to look up
both scheduled and finalised delayed events owned by the requesting user.

The homeserver SHOULD apply rate limiting to these endpoints to provide mitigation against the
[Resource Exhaustion](https://spec.matrix.org/v1.18/appendices/#threat-resource-exhaustion) threat.
They most likely require (dependent on the implementation) serialization steps
and can be used to slow down the homeserver.

#### Getting a single delayed event

The new authenticated Client-Server API endpoint `GET /_matrix/client/v1/delayed_events/{delay_id}` responds with
details on the delayed event with the specified `delay_id` owned by the requesting user.

If no such delayed event can be found, the homeserver will respond with HTTP 404
and a [standard error response](https://spec.matrix.org/latest/client-server-api/#standard-error-response)
with an `errcode` of `M_NOT_FOUND`.

> On success, the homeserver will respond with HTTP 200 and a JSON object containing the following fields:

Reviewer: I wonder if the status of the event could be too implicit here.

- To check for an error, you need to look at `error`.
- To check for a sent event, you probe `event_id`.
- A scheduled event is identified by the lack of both `error` and `event_id` and `running_since + delay` being in the future.
- A user-canceled event in turn can only be differentiated from a scheduled event once its trigger time has passed (so that `running_since + delay` is in the past).

Maybe this would become a bit easier to use if we added an explicit status property?

In fact, the listing endpoint introduced below has filter options for the status albeit in two separate parameters:

> - `status`: `"scheduled"|"finalised"` - Return only scheduled or finalised delayed events.
> [...]
> - `outcome`: `"send"|"cancel"|"error"` - Return only finalised delayed events that were either

How about we add a property à la `status = scheduled | sent | cancelled | failed`?

Author: In an earlier iteration we had the status (quiet recent change actually)
The argumentation was along the lines of: redundancy, save bits on the wire.

@AndrewFerr I think that new eyes stumbled over this is a good insight?
How often will this endpoint realistically be used? How much do we gain by leaving this field?

Reviewer: Convenience aside, I think we'll need something to differentiate manually cancelled events. Currently they'll look like scheduled events until their trigger time has passed. This could also be solved by an `is_canceled` property. If we need a new property any way, a more general `status` property might be more useful though?

Reviewer 2: Good catch on not being able to easily distinguish manually cancelled events vs scheduled ones.  Judging it by timestamps requires comparing client vs server times, which can get ugly.

Though I'm inclined against there being a `"status"` / `"outcome"` value for any status that can unambiguously determined by the presence of other fields on the response objects.  And as I see it, the only status that needs clarifying is a "cancelled" status, so a `"status"` enum field that has one of its possible values being `cancelled` really boils down to being a `"cancelled"` boolean field.

The reason why `status` and `outcome` remain as query parameters is because that felt like the best way to allow lookups to filter by the presence/absence of the fields they relate to, rather than by the values of those fields.

At the risk of being a tad esoteric, how about setting `"event_id"` to `null` for both cancelled & errored delayed events?  That is not ambiguous with an absent `"event_id"`, which would indicate a scheduled event.

Reviewer: > At the risk of being a tad esoteric, how about setting `"event_id"` to `null` for both cancelled & errored delayed events? That is not ambiguous with an absent `"event_id"`, which would indicate a scheduled event.

Hm, I'm a bit divided as this might be slightly non-intuitive and possibly a bit unergonomic in languages that don't differentiate `null` and `undefined`. I cannot think of another way to infer cancelled events without adding a new property though and I get the rationale behind avoiding that.

If we document what properties to check for which status in the MSC / spec, this variant will probably be fine, I suppose. I'd give it 6 out 10 strength.

Reviewer 17: I'm currently implementing this MSC for Ruma and Continuwuity. I think I'd prefer a field that clearly tells us what became of the event, rather than a set of rules for determining it.

Maybe a compromise is to leave out the `status` : `"scheduled"|"finalized"` field, as that can be inferred easily from available information. But to include the `outcome` : `"send"|"cancel"|"error"` field for finalized events.

Reviewer 2: Reducing the number of status-like fields from 2 to 1 is an improvement.  Still, it feels that even an `outcome` should be able to be inferred...

After thinking about it some more, there is indeed a "tell" for user-cancelled delayed events: the `origin_server_ts` / `finalised_ts` field.  If it is present and `event_id` & `error` are both absent, then the event was cancelled.

The full logic would then be:
- If `origin_server_ts` / `finalised_ts` is absent, then the delayed event is still scheduled.
- Otherwise, if `event_id` is present, then the delayed event has been sent (and was sent manually if `origin_server_ts`/`finalised_ts` < `running_since` + `delay`).
- Otherwise, if `error` is present, then the delayed event failed to be sent (and was descheduled) due to an error.
- Otherwise, the delayed event was cancelled.

This would also allow implementations to treat "falsey" values of any these fields (like the empty string or `undefined`/`null`/`nil`/etc) the same as the field being absent, since there would be no need to distinguish between the two.

Does that sound like it would be reasonable to implement?

Reviewer: I think I can live with that if we document it as explicitly as you did. 👍

Reviewer 2: Done by 5977f94e.

- `delay_id` - Required. The ID of the delayed event.
- `room_id` - Required. The ID of the room that the delayed event was scheduled to be sent in.
- `type` - Required. The event type of the delayed event.
- `state_key` - The state key of the delayed event if it is a state event; absent otherwise.
- `delay` - Required. The delay in milliseconds after the point of scheduling that the event is/was to be sent at.
- `running_since` - Required. The timestamp (as Unix time in milliseconds) when the delayed event was scheduled or
  last restarted.
- `content` - Required. The content of the delayed event.
  This is the body of the original `PUT` request, not a preview of the full event after sending.
- `error` - Present only for finalised events that were cancelled due to an error.
  The [standard error response](https://spec.matrix.org/v1.18/client-server-api/#standard-error-response)
  of the error that prevented the delayed event from being sent.
- `event_id` - The `event_id` this event got in case it was sent.

> - `origin_server_ts` - The timestamp (as Unix time in milliseconds) when the event was finalised;

Reviewer: Does this consciously include all finalization cases (sent, failed and cancelled)? I'm curious what the use case for the timestamp is when the event was _not_ sent.

It might be that my confusion just stems from the fact that `origin_server_ts` is only used for actually sent events in the current spec.

Reviewer 2: > Does this consciously include all finalization cases (sent, failed and cancelled)?

Yes, as the server will be aware of all such cases.

> I'm curious what the use case for the timestamp is when the event was _not_ sent.

This is to allow checking the time of when a delayed event was cancelled.

Another possibility is that a server doesn't allow retrying `/send` on a delayed event if the first attempt failed with an error (since keeping the event scheduled after an error has been marked as a SHOULD condition, not a MUST), in which case this timestamp would tell you when the send was attempted.  But this isn't terribly important.

> It might be that my confusion just stems from the fact that `origin_server_ts` is only used for actually sent events in the current spec.

True.  Maybe the name of this field could be changed to something more illustrative.  A straightforward option is `finalised_ts`.

Reviewer: > Another possibility is that a server doesn't allow retrying `/send` on a delayed event if the first attempt failed with an error (since keeping the event scheduled after an error has been marked as a SHOULD condition, not a MUST), in which case this timestamp would tell you when the send was attempted. But this isn't terribly important.

Oh, interesting. I actually wouldn't have expected the timestamp to be set in this case. When the server leaves the event scheduled after the failed initial `/send`, it's not actually finalized as per the definition at the beginning of the proposal (not sent and not cancelled). Given that the client has issued the `/send` itself, maybe we could omit setting the timestamp in that case?

> > It might be that my confusion just stems from the fact that `origin_server_ts` is only used for actually sent events in the current spec.
> 
> True. Maybe the name of this field could be changed to something more illustrative. A straightforward option is `finalised_ts`.

I like `finalised_ts` better as it feels a bit clearer. I'm only 6 out of 10 strength on it though given that we could also document the meaning of `origin_server_ts` properly.

Reviewer 2: > Oh, interesting. I actually wouldn't have expected the timestamp to be set in this case. When the server leaves the event scheduled after the failed initial `/send`, it's not actually finalized as per the definition at the beginning of the proposal (not sent and not cancelled). Given that the client has issued the `/send` itself, maybe we could omit setting the timestamp in that case?

To clarify, I didn't mean that a failed `/send` should set the `origin_server_ts` / `finalised_ts` timestamp even when the event remains scheduled; I meant that the server has the choice to deschedule an event if `/send` fails, _and only if it does so_ must the timestamp be set.  I.e. the timestamp must be set when & only when an event gets finalised/descheduled, and the server may choose to deschedule an event when `/send` fails.

> I like `finalised_ts` better as it feels a bit clearer. I'm only 6 out of 10 strength on it though given that we could also document the meaning of `origin_server_ts` properly.

Personally I've already confused myself with calling it `origin_server_ts`, as the field with the same name on sent events has different semantics.  So my vote is to rename it to make it clear that it has a distinct purpose.

Reviewer: > To clarify, I didn't mean that a failed `/send` should set the `origin_server_ts` / `finalised_ts` timestamp even when the event remains scheduled; I meant that the server has the choice to deschedule an event if `/send` fails, _and only if it does so_ must the timestamp be set. I.e. the timestamp must be set when & only when an event gets finalised/descheduled, and the server may choose to deschedule an event when `/send` fails.

Ah sorry, got it now. Thanks for clarifying. 👍 

> Personally I've already confused myself with calling it `origin_server_ts`, as the field with the same name on sent events has different semantics. So my vote is to rename it to make it clear that it has a distinct purpose.

Ok, then I'd say let's just do it. It definitely feels more intuitive.

Reviewer 2: Done by e557016e.

  absent if it is still scheduled.
  Using [timestamp massaging](https://spec.matrix.org/latest/application-service-api/#timestamp-massaging)
  does not affect the value of this field.

> #### Getting a list of delayed events

Reviewer: This API seems slightly more sophisticated than I expected. I'm sure there is thinking behind it but the proposal currently doesn't make the motivation very clear. Maybe we could add some examples of when and how a client would use this endpoint and its various query options?

Reviewer: @AndrewFerr given that Synapse [doesn't seem to implement](https://github.com/element-hq/synapse/blob/4655b435ee6e83518cd9ea6941d403be1cdf3a63/synapse/rest/client/delayed_events.py#L149) any of the query parameters yet, what's your take on descoping them to a future proposal?

Unless we assume extensive usage of this feature, we may be ok without pagination for a start. If really needed, we could always add it later in a backwards-compatible way by introducing a `/v2` of the endpoint.

Reviewer: Andrew and I discussed this out of band and we think a pagination API will be helpful but we'll want to strip the parameters down a little.

Reviewer 2: I'm hesitant to add a bulk retrieval endpoint that doesn't support pagination, even as a temporary measure.

What can be dropped/descoped is the complexity of the filters.  I'll have something for that later today.

(EDIT: this was written before seeing the above comment)

Reviewer 2: Done by 88859bdc.

It makes the following simplifications:
- Drop the custom sort options, and instead always sort by the scheduled send time
- Allow restricting results to fall within a specified time range
  - This works together with `dir` to allow dropping the `since_ts` option for having results start from a certain point in time
- Have `status` be the only "property" filter

It also adds some examples to demonstrate the purpose of the filters.

Reviewer: Thanks, I think this looks great now. 👍

The new authenticated Client-Server API endpoint `GET /_matrix/client/v1/delayed_events` responds with
a list of details of delayed events owned by the requesting user.

> Delayed events are returned in order of a time-based property which may specified by the `order_by` query parameter:

Reviewer: ```suggestion
Delayed events are returned in order of a time-based property which may be specified by the `order_by` query parameter:
```

- `reason`: `"action"|"delay"` - Return only finalised delayed events that were either
  sent or cancelled by [the /send or /cancel endpoint](#managing-delayed-events), or
  remained scheduled until their scheduled send time.

If any query parameter is set to an unsupported value,
the homeserver will respond with HTTP 400
and a [standard error response](https://spec.matrix.org/latest/client-server-api/#standard-error-response)
with an `errcode` of `M_INVALID_PARAM`.

On success, the response is HTTP 200 and a JSON object containing the following fields:

- `delayed_events` - An array of objects describing delayed events owned by the requesting user
  that match the filters provided in the request, in the order specified by the `order_by` and `dir` query parameters.
  These objects contain the same fields as the object returned by
  [the single-item lookup](#getting-a-single-delayed-event).
- `next_batch` - A token that be passed into a subsequent call to the endpoint to retrieve the next page of results.
  Absent if there are is no next page of results.
- `prev_batch` - A token that be passed into a subsequent call to the endpoint to retrieve the previous page of results.

>   Absent if there are is no previous page of results.

Reviewer: I assume you can only paginate one way at a time?

```suggestion
- `next_batch` - A token that can be passed into a subsequent call to the endpoint to retrieve the next page of results.
  Absent if paginating backwards or when there is no next page of results.
- `prev_batch` - A token that can be passed into a subsequent call to the endpoint to retrieve the previous page of results.
  Absent if paginating forward or when there is no previous page of results.
```

Reviewer 2: Good catch, both on the one-way pagination & the grammar :sweat_smile: 

Now that you mention it, I'm unaware of whether other bi-directional list lookups give only a single next/prev token for the 2nd page and onwards.  In any case, doing it like you mention here is reasonable.

      "state_key": "@user:example.com_DEVICEID",
      "delay": 5000,
      "running_since": 1721732853284,
      "content":{
        "application": "m.call",
        "call_id": "",
        ...
      }
    },
    {
      "delay_id": "zyxwvuts",
      "room_id": "!roomid:example.com",
      "type": "m.room.message",
      "delay": 2000,
      "running_since": 1721732853280,
      "content":{
        "body": "Boo!",
        "msgtype": "m.text"
      },
      "event_id": "$xyzyxyz"
    }
  ],
  "next_batch": "b12345"
}
```

Unless the delayed event is updated beforehand, the event will be sent after `running_since` + `delay`.

This can be used by clients to display events that have been scheduled to be sent in the future.

For use cases where the existence of a delayed event is also of interest for other room members
(e.g. self-destructing messages), it is recommended to include this information in the original/affected event itself.

#### Retention of finalised delayed events

The batch size and the amount of finalised events that stay on the homeserver can be chosen by the homeserver.
The recommended values are:

- batch size: 10
- `finalised` retention: 7 days

> - `finalised` max cached events per user: 1000

Reviewer: The term "cached" threw me off here at first. I think this means the number of retained finalized events? So the retention is effectively up to 7 days or 1000 events, depending on which strikes first?

```suggestion
- `finalised` retention: up to 7 days or 1000 events per user, whichever occurs first
```

There is no guarantee for a client that all events will be available in the
finalised events list if they exceed the limits of their homeserver.
Additionally, a homeserver MAY discard finalised delayed events that have been returned by a
`GET /_matrix/client/v1/delayed_events(/{delay_id})?` response.

> ### Homeserver implementation details

Reviewer: The term "implementation details" might be slightly misleading here because it usually refers to something that is consciously left unspecified or something that is not relevant under the current scope. This section contains actual specification for the server though.

Maybe something like this would be clearer?

```suggestion
### Additional homeserver behaviour
```

Author: needed to check the section again. Yes `Additional homeserver behaviour` fits much better.

#### Rate-limiting at the point of sending

Further to the rate limiting of the API endpoints, the homeserver SHOULD apply rate limiting to the sending
of delayed messages at the point that they are inserted into the DAG.

This is to provide mitigation against the [High Volume of Messages](
https://spec.matrix.org/v1.18/appendices/#threat-high-volume-of-messages) threat where a malicious actor
could schedule a large volume of events ahead of time without exceeding a rate limit on the initial `PUT` request,
but has specified a `delay` that corresponds to a common point of time in the future.

If a delayed event fails to be sent at its scheduled send time due to a rate limit failure,
the homeserver SHOULD NOT retry sending the event. Instead, the event will be stored as a finalised delayed event
with its `error` field set, available to be retrieved by a client for the user who requested the event.
It then becomes the user's responsibility to fetch this error and retry sending the event as appropriate.

A limit on the maximum number of delayed events that can be scheduled at a time could also provide some mitigation

> against this attack.

Reviewer: This seems to be part of the proposal text further up now if I understood correctly.

> The homeserver SHOULD enforce a limit of how many delayed events a user may have scheduled at once [...]

```suggestion
```

### Guest accounts

All delayed event related endpoints are available to guest accounts.
This allows guest accounts to participate in MatrixRTC sessions.

## Use case specific considerations

Delayed events can be used for many different features: tea timers, reminders, or ephemeral events could be implemented
using delayed events, where clients send room events with
intentional mentions or a redaction as a delayed event.

> It can even be used to send temporal power levels/mutes or bans.

Reviewer: This and some of the content from the "Self-destructing messages" section might be good to have at the beginning of the proposal. The current introduction exclusively touches upon the VoIP use case. I think it would give use a much stronger case if we also covered the other foreseen usages there. This could also help motivate some of the details and design choices that follow afterwards in the proposal.

Reviewer: Attempting to address this with https://github.com/matrix-org/matrix-spec-proposals/pull/4478.

Reviewer: This landed.

Both are reasonable data points that clients might want to use.
This would solve issues related to delayed events since
it would make it transparent to clients, when an event was scheduled and when it was distributed over federation.

## Alternatives

### OAuth 2.0 scope for management endpoints

Instead of the [delayed event management endpoints](#managing-delayed-events) being unauthenticated
to permit [delegation to an external service](#delegating-delayed-events),
those endpoints could be given an OAuth 2.0 scope and be restricted to sessions that have requested it.
The scope would be within the existing `urn:matrix:client:api:*` scope,
so that access to the entirety the Client-Server API would include access to these endpoints as well.

Using OAuth 2.0 to restrict access on these endpoints has many benefits over using a path parameter (`delay_id`) as an
access token, such as more fine-grained revocability on access to the endpoints, and
better identification of what entity is requesting these endpoints, which can be used to apply per-entity ratelimits.

The downsides of this approach are its requirement on the homeserver supporting the OAuth 2.0 login API,

> and the additional network/configuration overhead for external services to request access to this scope.

Reviewer: The OAuth APIs have officially been part of the spec for almost a year now. So the argument of having to support them should gradually get smaller and might not really be relevant here.

```suggestion
The downside of this approach is the additional network/configuration overhead for external services to
request access to this scope.
```

Author: I think the focus on this section is the required scoped token work (which I think it not yet there).
Maybe we can rephrase it so it becomes clear that this would then be blocked on figuring out the scoping implementation?

Reviewer: Oh, now I get it. How about something like this?

```suggestion
The downsides of this approach are the complexity of having  to implement scoped access tokens in homeservers
and the additional network/configuration overhead for external services to request access to this scope.
```

Reviewer 2: Instead of "complexity", maybe say "required work"; i.e. there is work to be done for servers to add more access scopes, but whether it's complex is a matter of debate (and suggesting that it is makes it seem more daunting than it might actually be)

Reviewer: Good idea. 👍 

```suggestion
The downsides of this approach are the required work of having to implement scoped access tokens in homeservers
and the additional network/configuration overhead for external services to request access to this scope.
```

If it's not needed to allow modification of a delayed event after it has been scheduled, there is a benefit in
federating the scheduled event (adding it to the DAG immediately). It increases resilience: the sender's homeserver can
disconnect and the delayed message still will enter non-soft-failed state (will be sent).

However, for the MatrixRTC use case it's required to be able to modify the event after it has been scheduled. As such,
this approach has been discounted.

### MQTT style Last Will

[MQTT](https://mqtt.org/) has the concept of a Will Message that is published by the server when a client disconnects.

The client can set a Will Message when it connects to the server.
If the client disconnects unexpectedly, the server will
publish the Will Message if the client is not back online within a specified time.

A similar concept could be applied to Matrix by having the client specify a set of "Last Will" events and have the
homeserver trigger them if the client (possibly identified by device ID) does not send an API request within a specified
time.

The main differentiator is that this type of approach might use the sync loop as the "heartbeat" equivalent similar to

> [MSC4018](https://github.com/matrix-org/matrix-spec-proposals/pull/4018).

Reviewer: When using the sync loop as the heartbeat, this scheme seems like a generalization of the solution discussed and discarded in the "MSC4018 (use client sync loop)" section above. Maybe we could just combine the two sections?

Reviewer: Attempting to address this with https://github.com/matrix-org/matrix-spec-proposals/pull/4479.

A benefit compared to this proposal is that theoretically there would be no additional network traffic overhead.

Some complications are:

- in order to avoid additional network traffic, the homeserver would need to proactively realise that a connection
  has dropped. Depending on the network/load balancer stack this might be problematic.
- as an alternative, the client could reduce the long poll timeout (from a typical 30s down to, say, 5s) which would
  result in a traffic increase.
- As syncing is a per-client concept, the MatrixRTC app has to either run in the same process as the client so that a
  MatrixRTC app failure triggers the client Last Will or the client has to observe the MatrixRTC app and simulate the
  Last Will if the MatrixRTC app fails.

### `M_INVALID_PARAM` instead of `M_MAX_DELAY_EXCEEDED`

> The existing `M_INVALID_PARAM` error code could be used instead of introducing a new error code `M_MAX_DELAY_EXCEEDED`.

Reviewer: I think we need to explain why we're discarding this alternative here. The only reason I can think of is that a separate error code lets the client know that it can look for the `max_delay` property? As discussed in https://github.com/matrix-org/matrix-spec-proposals/pull/4140/changes#r3257895622, however, we could possibly omit the `max_delay` property altogether.

Reviewer 2: Yes, the point of a new error code was to be able to have the `max_delay` field in the error object.  But this can now all be removed in favour of expressing the max delay in `/capabilities`.

Reviewer 2: Removed by 8f9f4544.

### `M_MAX_DELAYED_EVENTS_EXCEEDED` instead of `M_LIMIT_EXCEEDED`

A new error code `M_MAX_DELAYED_EVENTS_EXCEEDED` could be used instead of reusing the existing `M_LIMIT_EXCEEDED` error
code. The purpose of using a new error code would be to better distinguish delayed event scheduling limits from other

> resource limits.

Reviewer: This seems to contradict the design choice made in the proposal. We should either change this to motivate why we chose `M_LIMIT_EXCEEDED` or switch the proposal to use `M_MAX_DELAYED_EVENTS_EXCEEDED`.

Given that other resource limits are currently not differentiated via separate error codes in the API (see also https://github.com/matrix-org/matrix-spec-proposals/pull/4335), `M_LIMIT_EXCEEDED` seems like a reasonable choice to me.

```suggestion
A new error code `M_MAX_DELAYED_EVENTS_EXCEEDED` could be introduced instead of reusing the existing `M_LIMIT_EXCEEDED` error
code. This would allow clients to better distinguish delayed event scheduling limits from other resource limits.
Given that other resource limits are currently not differentiated via separate error codes in the API either,
reusing `M_LIMIT_EXCEEDED` seems reasonable though.
```

Reviewer 2: Good idea.  Adding the comment on how other resource limits don't get their own errcode justifies why this is alternative.

### Naming

The following alternative names for this concept are considered:

- Future
- DelayedEvents
- PostponedEvents

> - LastWill

Reviewer: It might be best to not open the naming can of worms ourselves. 🙈 😅

But if we really want to, I think we should add some reasoning for discarding the other terms. Here is what I can think of:

```suggestion
- Futures: Doesn't seem like a good match because the result of the API call cannot be "awaited" as you
  would normally `await` a future in various programming languages.
- Postponed Events: Similar to "delayed events" but longer.
- Last Will: Seems unfitting because the feature can be used regardless of whether the client goes offline
  or not. The term also feels somewhat dark.
```

Author: Lets add this.

Instead of providing a `cancel` management action for scheduled delayed events,
the client could send a `DELETE` request to an endpoint representing a target delayed event.

This feels more elegant, but it doesn't feel like a good suggestion for how the other actions are mapped.
Also, `DELETE` suggests that the target resource will be truly deleted, but this is at odds with how
cancelling a delayed event has it retained as a finalised event for later lookup.

### [Ab]use typing notifications

Some exploration of using typing notifications to indicate that a user is still connected to a call was done.

The idea of extending [MSC3038: Typed typing notifications](
https://github.com/matrix-org/matrix-spec-proposals/pull/3038) to allow for additional meta data
(like device ID and call ID) was considered.

A perceived benefit was that if the delay events were federated, then the typing notification EDUs might provide an
efficient transport.

However, as the conclusion was to [not federate the delayed events](#federated-delayed-events), this approach was
discounted in favour of a dedicated endpoint.

### Alternative to `running_since` field

> Some alternatives for the `running_since` field on the `GET` response are:

Reviewer: As another alternative, maybe we could pick something that ends with `..._ts`? The identifier `running_since` could erroneously be interpreted as a duration rather than a timestamp. This would be alleviated if we chose e.g. `scheduled_ts` or `created_ts`. This would also blend in slightly better with the other sorting options in GET /_matrix/client/v1/delayed_events: send_ts, origin_server_ts & ..._ts (rather than running_since)

In any case, if we list any alternatives, we need to motivate why we have discarded them.

Reviewer 2: Good idea.  IMO `scheduled_ts` is best because it reuses existing terminology introduced by this MSC.

Reviewer 2: Done by 06ab77af.

  "error":"The delayed event did not get sent because a different user updated the same state event.
  So the scheduled event might change it in an undesired way."
}
```

Whilst the MSC is in the proposal stage,
The `M_UNKNOWN` `errcode` should be used instead of `M_CANCELLED_BY_STATE_UPDATE` as follows:

```json
"outcome": "cancel",
"reason": "error",
"error": {
  "errcode": "M_UNKNOWN",
  "org.matrix.msc4140.errcode": "M_CANCELLED_BY_STATE_UPDATE",
  "error":"The delayed event did not get sent because a different user updated the same state event.
  So the scheduled event might change it in an undesired way."
}
```

Note that this behaviour does not apply to regular (non-state) events as there is no concept of a
(`event_type`, `state_key`) pair that could be overwritten.

This rule is proposed as alternative because there is presently no use case that relies on delayed state events.
An earlier revision of MatrixRTC used delayed state events for call membership and relied on this rule to prevent
race conditions, but it has since migrated to using [Sticky Events](
https://github.com/matrix-org/matrix-spec-proposals/pull/4354) instead of state events for this,

> and thus no longer needs this rule.

Reviewer: This seems slightly complicated and if we cannot foresee a use case for it, maybe we don't even need to list it as an alternative?

Author: I like the idea of dropping anything related to the autocancelling if a state event for the same key is sent.
At the end we do have power levels to figure out the a potential conflict. 
And matrix has this caveat (overwriting state) anyways. There is always delay in a federated world so app developers/spec designers should be expected to be aware of state having a risk to race and conflict and thing around this.

What about this:
We drop this as an alternative and make it a note along those lines:
The final spec text should mention, that a delayed event can overwrite changes that happened between sending the event and it getting finalised. (like with federation delay) Matrix based solutions that have to consider this and design around it. For example: pinned messages should have seperate keys or use relations instead of state events. State events that group independent information have to be challenged if they might be used with delayed events. If so one should consider splitting them into multiple events (or relation based solution where events represent a diff instead of a new state)

Reviewer: Yes, that sounds good. It could maybe go into the "Potential issues" section this way. 👍

Reviewer: Attempting to address this with https://github.com/matrix-org/matrix-spec-proposals/pull/4477.

- `org.matrix.msc4140.delay_id` should be used instead of `delay_id` as the key in `unsigned` event data.

Additionally, the feature is to be advertised as an unstable feature in the `GET /_matrix/client/versions` response,
with the key `org.matrix.msc4140` set to `true`. So, the response could then look as follows:

```json
{
    "versions": ["..."],
    "unstable_features": {
        "org.matrix.msc4140": true
    }
}
```

Once the MSC is accepted, but before the homeserver advertises the spec version that includes the MSC, the homeserver
should advertise `org.matrix.msc4140.stable` as an unstable feature flag to let clients know that they can use the
stable endpoints for sending and managing delayed events.

### Implementation differences

In the Synapse implementation of this MSC:
- `GET /_matrix/client/v1/delayed_events` does not yet support query parameters,
  and returns only scheduled delayed events.
- `GET /_matrix/client/v1/delayed_events/{delay_id}` is not yet implemented.
- [The management endpoints](#managing-delayed-events) use a single URL for all management actions,
  in the manner described [here](#management-endpoint-action-in-request-body).

> 

Reviewer: I think this shouldn't be part of the proposal text but instead go into the PR description (or a separate comment similar to https://github.com/matrix-org/matrix-spec-proposals/pull/4473#discussion_r3253382151).

```suggestion
```

Author: @AndrewFerr what is the current state of those anyways?

Reviewer 2: PRs for those points have been on hold but are still in progress.  I've posted them as an update to the top comment of this MSC instead of as a standalone comment.

# MSC4140: Delayed events (Futures)

This MSC proposes a mechanism by which a Matrix client can schedule an event (including a state event) to be sent into

> a room on behalf of a user at a later time.

Author: Since the user is still the one creating and in the future signing the event I am not sure "behalf of the user" is the right wording?
It is publishing an event sent by a user itself at a later point in time?

Reviewer 4: I'm not sure it really matters in this context... I've taken it out anyhow. 🤷

    - [Self-destructing messages](#self-destructing-messages)
  - [Potential issues](#potential-issues)
  - [Alternatives](#alternatives)
    - [Delegating futures](#delegating-futures)
    - [Batch sending](#batch-sending)
    - [Not reusing the send/state endpoint](#not-reusing-the-sendstate-endpoint)
    - [Batch sending futures with custom endpoint](#batch-sending-futures-with-custom-endpoint)
      - [Batch Response](#batch-response)
      - [Allocating event ID at the point of scheduling the send](#allocating-event-id-at-the-point-of-scheduling-the-send)
      - [EventId template variable](#eventid-template-variable)
    - [MSC4018 use client sync loop](#msc4018-use-client-sync-loop)
    - [Federated futures](#federated-futures)
    - [MQTT style Last Will](#mqtt-style-last-will)
    - [Naming](#naming)
  - [Security considerations](#security-considerations)
  - [Unstable prefix](#unstable-prefix)
  - [Dependencies](#dependencies)

<!-- /TOC -->

## Background and motivation

This proposal originates from the needs of VoIP signalling in Matrix:

The Client-Server API currently has a  [Voice over IP module](https://spec.matrix.org/v1.11/client-server-api/#voice-over-ip)

> that uses room messages to communicate the call state. However, it only allows for calls with two participants.

Author: @hughns those are marked as outdated but i think parts of it is still open?

> that uses room messages to communicate the call state. However, it only allows for calls with two participants.

Reviewer 4: > Do we need this context?

I think it is helpful so I'm going to leave it in.

> @hughns those are marked as outdated but i think parts of it is still open?

I have incorporated most unresolved conversations into the MSC including the outdated ones. There are still a few more to go.

[MSC3401: Native Group VoIP Signalling](https://github.com/matrix-org/matrix-spec-proposals/pull/3401) proposes a scheme that
allows for more than two-participants by using room state events.

In this arrangement each device signals its participant in a call by sending a state event that represents the device's
"membership" of a call. Once the device is no longer in the call, it sends a new state event to update the call state and say
that it is no longer a member.

This works well when the client is running and can send the state events as needed. However, if the client is not running and

> able to communicate with the homeserver (e.g. the user closes the app or loses connection) the call state is not updated to

Reviewer 4: I've incorporated this in [`a1b8121`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/a1b81213e2781758ca8097eb294ab663f6a2ea39).

say that the participant has left.

The motivation for this MSC is to allow updating call member state events after the user disconnected by allowing to
schedule/delay/timeout/expire events in a generic way.

The ["reliability requirements for the room state"](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/matrixRTC/proposals/4143-matrix-rtc.md#reliability-requirements-for-the-room-state)
section of [MSC4143: MatrixRTC](https://github.com/matrix-org/matrix-spec-proposals/pull/4143) has more details on the use case.

There are numerous possible solution to solve the call member event expiration. They are covered in detail
in the [Use case specific considerations/MatrixRTC](#use-case-specific-considerations) section, because they are not part
of this proposal.

The proposal here allows a Matrix client to schedule a "hangup" state event to be sent after a specified time period. The client

> can then periodically reset/restart the timer whilst it is running. If the client is no longer running or able to communicate then the timer would expire and the homeserver would send the "hangup" event on behalf of the client.

Reviewer 4: In [`9098fea`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/9098fea5e5bc31d900c5ad4db92c604f6293df86) I have rewritten the MatrixRTC use case of this MSC using a concrete example and the heartbeat terminology.

I've also added [`3a3a5b5`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/3a3a5b5fea265269068ccc423230fe817aea32ee).

## Proposal

We propose the following operations are added to the client-server API:

- Schedule an event to be sent at a later time
- Get a list of events that have been scheduled to send
- Refresh the timeout of a scheduled event
- Send the scheduled event immediately
- Cancel a scheduled event so that it is never sent

At the point of an event being scheduled the homeserver is [unable to allocate the event ID](#allocating-event-id-at-the-point-of-scheduling-the-send). Instead, the homeserver allocates a _future ID_ to the scheduled event which is used during the above API operations.

### Scheduling a delayed event

We propose extending the existing
[`PUT /_matrix/client/v3/rooms/{roomId}/state/{eventType}/{stateKey}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
and
[`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](https://spec.matrix.org/v1.11/client-server-api/#put_matrixclientv3roomsroomidstateeventtypestatekey)

> endpoints by adding an optional `future_timeout` query parameter.

Reviewer 4: As per discussion just now, we are going to stick with adding the query parameter to the existing API endpoints rather than having new API endpoints.

- `future_timeout` - Optional number of milliseconds the homeserver should wait before sending the event. If no `future_timeout` is provided, the event is sent immediately as normal.

The body of the request is the same as currently.

If a `future_timeout` is provided, the homeserver schedules the event to be sent with the specified delay and returns the _future ID_ in the `future_id` field (omitting the `event_id` as it is not available):

```http
200 OK
Content-Type: application/json

{
  "future_id": "1234567890"
}
```

The homeserver can optionally enforce a maximum timeout duration. If the requested timeout exceeds the maximum the homeserver
can respond with a [`400`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) and a Matrix error code `M_FUTURE_MAX_TIMEOUT_EXCEEDED` and the maximum allowed timeout (`timeout_duration` in milliseconds).

For example the following specifies a maximum delay of 24 hours:

```http
400 Bad Request
Content-Type: application/json

{
  "errcode": "M_FUTURE_MAX_TIMEOUT_EXCEEDED",
  "error": "The requested timeout exceeds the allowed maximum.",
  "timeout_duration": 86400000
}
```

### Getting delayed events

> We propose adding a new authenticated endpoint `GET /_matrix/client/v0/futures` to the client-server API to allow clients to get a list of all the events that have been scheduled to send in the future.

Reviewer 4: Hopefully [`b79e4e2`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/b79e4e2762648dd4fb8f4b4839f350cb27c12302) achieves this.

```http
HTTP 200 OK
Content-Type: application/json

[
  {
    "url":"/_matrix/client/v1/rooms/{roomId}/send_future/{eventType}/{txnId}?future_timeout={timeout_duration}",
    "body":{
      ...event_body
    },
    "response":{
      "future_id": "1234567890",
      "last_refresh": 1721732853284, // unix timestamp in milliseconds
      "timeout": 15000 // milliseconds
    }
  }

> ]

Author: Removing this or changing the formatting of the `GET` response is a good idea.
(As we discussed in chat)

Author: read through this commit by commit. This is already solved!

This translate to: _"only send the event when the client is not running the its program anymore (not sending the heartbeat anymore)"_
We call those delayed events `Futures`.

New endpoints are introduced:

`PUT /_matrix/client/v3/rooms/{roomId}/send_future/{eventType}/{txnId}?future_timeout={timeout_duration}&future_group_id={group_id}`

`PUT /_matrix/client/v3/rooms/{roomId}/state_future/{eventType}/{stateKey}?future_timeout={timeout_duration}&future_group_id={group_id}`

Those behave like the normal `send`/`state` endpoints except that that they allow
to define `future_timeout` and `future_group_id` in their query parameters.

- `future_timeout: number | "none"` is a required parameter that defines how long the homeserver will wait before sending
  the event into the room. Since the timeout can be refreshed and sending the future can be triggered via an endpoint (see: [Proposal/Delegating futures](#delegating-futures)) this value is not enough to predict the time this event will arrive in the room.
  - If set to `"none"` the future will never expire and can only be send by the [external delegation endpoint](#delegating-futures).
    We call such a future **action future**.

>   - If set to a `number` we call the future **timeout future**

Reviewer 5: What is the time unit here? seconds?

### Managing delayed events

A set of new unauthenticated client-server API endpoints at `POST /_matrix/client/v1/delayed_events/{delay_id}/[send|cancel|restart]` allows scheduled events
to be managed.

The final path component of these endpoints specifies the action to take on the delayed event:

- `send` - Send the delayed event immediately instead of waiting for the delay time to be reached.
- `cancel` - Cancel the delayed event so that it is never sent.
- `restart` - Reset the send time to be `now + original_delay`.

For example, the following would send the delayed event with delay ID `1234567890` immediately:

```http
POST /_matrix/client/v1/delayed_events/1234567890/send
Content-Type: application/json

{
}
```

> These endpoints are unauthenticated so that control over a particular delayed event may be

Reviewer 2: This is now merged: https://github.com/element-hq/synapse/pull/19152

### Getting delayed events

> We propose adding a new authenticated endpoint `GET /_matrix/client/v1/futures` to the client-server API to allow clients to get a list of all the events that have been scheduled to send in the future.

Author: is v1 what we are supposed to use? Do we not start at 0?

Reviewer 4: As far as I know, we start at v1.

    {
      "delay_id": "1234567890",
      "room_id": "!roomid:example.com",

>       "event_type": "m.room.message",

Author: This is sth we cannot change in this MSC.

### Managing delayed events

We propose adding a new authenticated client-server API endpoint at `POST /_matrix/client/v1/futures/{future_id}` to manage
delayed events that have already been scheduled.

The body of the request is a JSON object containing the following fields:

- `action` - The action to take on the delayed event. Must be one of:
  - `send` - Send the delayed event immediately.
  - `cancel` - Cancel the delayed event so that it is never sent.
  - `refresh` - Restart the timeout of the delayed event.

For example, the following would send the delayed event with future ID `1234567890` immediately:

```http

> POST /_matrix/client/v1/futures/1234567890

Author: I am wondering if this is a too generic endpoint name?
`update_future_event` might be better?
Otherwise this really could be anything

Reviewer 4: For consistency with how this MSC is now named and "positioned" I would suggest `delayed_events` so that you have:

`GET /_matrix/client/v1/delayed_events` and `POST /_matrix/client/v1/delayed_events/{future_id}`

Reviewer 4: This has been incorporated into the latest proposal.

If the target delayed event is already finalised with an outcome that conflicts with the `action`, i.e.
if the `action` is `send` or `restart` and the delayed event has already been cancelled,
or if the `action` is `cancel` and the delayed event has already been sent,
the homeserver will respond with HTTP 409.

If the action is `send` and the delayed event is unable to be sent due to an error,
the homeserver will respond with that error (e.g. HTTP 400
and a [standard error response](https://spec.matrix.org/v1.19/client-server-api/#standard-error-response)

> with an `errcode` of `M_DUPLICATE_ANNOTATION` if the event is an attempt to send a duplicate annotation,

Reviewer 14: I'm confused. Why is this about annotations now?

Reviewer 2: It's not; that's just an example of an errcode that could be responsible for the failed send.  I picked this because [the spec for non-delayed `/send`](https://spec.matrix.org/latest/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid) gives it as an example.  If it's distracting to have it here, though, I have no issues removing it.

Reviewer 2: I've gone ahead and reworded this with 05029e2b.  Instead of giving example responses, now it tries to more clearly say that the error response given is the same as for non-delayed event requests.  Let me know if this reads better.

With a widget implementation for calls we cannot guarantee that the widget is running during the sync loop.
So one either has to move the hangup logic to the hosting client or let the widget run all the time.

With a dedicated ping (independent to the sync loop) it is more flexible and allows us to let the widget
execute the refresh.
If the widget dies, the call membership will disconnect.

### Federated futures

The delayed/scheduled events could be sent over federation immediately and then have the receiving servers process them at the appropriate time.

> 

Author: If we mention it here this should also provide some of the context information we have evaluated.
 - heartbeats need to federate to.
 - other homeservers missing heartbeats would result in the event reaching the dag (fragile if there are lots of homeservers involved and all of them could cancel the session)
 
This comment: https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/fa461e21a39adcbbc676adf265fc74a20bf8488a#r1653083566, should be linked or summarized here.

Reviewer 4: I've added all of these to the alternatives section in [`4a2ca48`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/4a2ca4800d6e1f0debad654a8812d2e7ffc1bbd8).

### MQTT style Last Will

[MQTT](https://mqtt.org/) has the concept of a Will Message that is published by the server when a client disconnects.

The client can set a Will Message when it connects to the server. If the client disconnects unexpectedly the server will publish the Will Message if the client is not back online within a specified time.

A similar concept could be applied to Matrix by having the client specify a set of "Last Will" event(s) and have the homeserver trigger them if the client (possibly identified by device ID) does not send an API request within a specified time.

> 

Author: I think this should mention the main differentiator that this uses the sync loop as the "heartbeat" equivalent.
 - No network traffic overhead
 - Homeserver/client dependent timeouts
 - As syncing is a per client concept the rtc app has to either run in the same process so that a rtc app failure triggers the client last will or the client has to observe the rtc app and simulate the last will if the rtc app fails.

Reviewer 4: I've incorporated this in to [`37979cd`](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/commits/37979cd47c3a3e919c35cf4b22477478b9169290).
